// edgejs-parser/dist/main.js
var As = typeof global == "object" && global && global.Object === Object && global, Yi = typeof self == "object" && self && self.Object === Object && self, Ee = As || Yi || Function("return this")(), ee = Ee.Symbol, Rs = Object.prototype, Xi = Rs.hasOwnProperty, qi = Rs.toString, rt = ee ? ee.toStringTag : void 0;
function Zi(t) {
  var e = Xi.call(t, rt), n = t[rt];
  try {
    t[rt] = void 0;
    var r = !0;
  } catch {
  }
  var s = qi.call(t);
  return r && (e ? t[rt] = n : delete t[rt]), s;
}
var Qi = Object.prototype, Ji = Qi.toString;
function ea(t) {
  return Ji.call(t);
}
var ta = "[object Null]", na = "[object Undefined]", Tr = ee ? ee.toStringTag : void 0;
function be(t) {
  return t == null ? t === void 0 ? na : ta : Tr && Tr in Object(t) ? Zi(t) : ea(t);
}
function ce(t) {
  return t != null && typeof t == "object";
}
var ra = "[object Symbol]";
function Xt(t) {
  return typeof t == "symbol" || ce(t) && be(t) == ra;
}
function qt(t, e) {
  for (var n = -1, r = t == null ? 0 : t.length, s = Array(r); ++n < r; )
    s[n] = e(t[n], n, t);
  return s;
}
var N = Array.isArray, mr = ee ? ee.prototype : void 0, Ar = mr ? mr.toString : void 0;
function Is(t) {
  if (typeof t == "string")
    return t;
  if (N(t))
    return qt(t, Is) + "";
  if (Xt(t))
    return Ar ? Ar.call(t) : "";
  var e = t + "";
  return e == "0" && 1 / t == -1 / 0 ? "-0" : e;
}
var sa = /\s/;
function ia(t) {
  for (var e = t.length; e-- && sa.test(t.charAt(e)); )
    ;
  return e;
}
var aa = /^\s+/;
function oa(t) {
  return t && t.slice(0, ia(t) + 1).replace(aa, "");
}
function te(t) {
  var e = typeof t;
  return t != null && (e == "object" || e == "function");
}
var Rr = NaN, ca = /^[-+]0x[0-9a-f]+$/i, ua = /^0b[01]+$/i, la = /^0o[0-7]+$/i, ha = parseInt;
function fa(t) {
  if (typeof t == "number")
    return t;
  if (Xt(t))
    return Rr;
  if (te(t)) {
    var e = typeof t.valueOf == "function" ? t.valueOf() : t;
    t = te(e) ? e + "" : e;
  }
  if (typeof t != "string")
    return t === 0 ? t : +t;
  t = oa(t);
  var n = ua.test(t);
  return n || la.test(t) ? ha(t.slice(2), n ? 2 : 8) : ca.test(t) ? Rr : +t;
}
var Ir = 1 / 0, da = 17976931348623157e292;
function pa(t) {
  if (!t)
    return t === 0 ? t : 0;
  if (t = fa(t), t === Ir || t === -Ir) {
    var e = t < 0 ? -1 : 1;
    return e * da;
  }
  return t === t ? t : 0;
}
function Zt(t) {
  var e = pa(t), n = e % 1;
  return e === e ? n ? e - n : e : 0;
}
function gt(t) {
  return t;
}
var ga = "[object AsyncFunction]", Ea = "[object Function]", Ta = "[object GeneratorFunction]", ma = "[object Proxy]";
function Ce(t) {
  if (!te(t))
    return !1;
  var e = be(t);
  return e == Ea || e == Ta || e == ga || e == ma;
}
var En = Ee["__core-js_shared__"], _r = (function() {
  var t = /[^.]+$/.exec(En && En.keys && En.keys.IE_PROTO || "");
  return t ? "Symbol(src)_1." + t : "";
})();
function Aa(t) {
  return !!_r && _r in t;
}
var Ra = Function.prototype, Ia = Ra.toString;
function Ke(t) {
  if (t != null) {
    try {
      return Ia.call(t);
    } catch {
    }
    try {
      return t + "";
    } catch {
    }
  }
  return "";
}
var _a = /[\\^$.*+?()[\]{}|]/g, Sa = /^\[object .+?Constructor\]$/, Oa = Function.prototype, Na = Object.prototype, Ca = Oa.toString, La = Na.hasOwnProperty, ya = RegExp(
  "^" + Ca.call(La).replace(_a, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"
);
function va(t) {
  if (!te(t) || Aa(t))
    return !1;
  var e = Ce(t) ? ya : Sa;
  return e.test(Ke(t));
}
function Pa(t, e) {
  return t?.[e];
}
function We(t, e) {
  var n = Pa(t, e);
  return va(n) ? n : void 0;
}
var On = We(Ee, "WeakMap"), Sr = Object.create, Ma = /* @__PURE__ */ (function() {
  function t() {
  }
  return function(e) {
    if (!te(e))
      return {};
    if (Sr)
      return Sr(e);
    t.prototype = e;
    var n = new t();
    return t.prototype = void 0, n;
  };
})();
function ka(t, e, n) {
  switch (n.length) {
    case 0:
      return t.call(e);
    case 1:
      return t.call(e, n[0]);
    case 2:
      return t.call(e, n[0], n[1]);
    case 3:
      return t.call(e, n[0], n[1], n[2]);
  }
  return t.apply(e, n);
}
function U() {
}
function xa(t, e) {
  var n = -1, r = t.length;
  for (e || (e = Array(r)); ++n < r; )
    e[n] = t[n];
  return e;
}
var ba = 800, wa = 16, Ua = Date.now;
function Fa(t) {
  var e = 0, n = 0;
  return function() {
    var r = Ua(), s = wa - (r - n);
    if (n = r, s > 0) {
      if (++e >= ba)
        return arguments[0];
    } else
      e = 0;
    return t.apply(void 0, arguments);
  };
}
function Da(t) {
  return function() {
    return t;
  };
}
var Dt = (function() {
  try {
    var t = We(Object, "defineProperty");
    return t({}, "", {}), t;
  } catch {
  }
})(), $a = Dt ? function(t, e) {
  return Dt(t, "toString", {
    configurable: !0,
    enumerable: !1,
    value: Da(e),
    writable: !0
  });
} : gt, Ga = Fa($a);
function _s(t, e) {
  for (var n = -1, r = t == null ? 0 : t.length; ++n < r && e(t[n], n, t) !== !1; )
    ;
  return t;
}
function Ss(t, e, n, r) {
  for (var s = t.length, i = n + -1; ++i < s; )
    if (e(t[i], i, t))
      return i;
  return -1;
}
function Ba(t) {
  return t !== t;
}
function Ha(t, e, n) {
  for (var r = n - 1, s = t.length; ++r < s; )
    if (t[r] === e)
      return r;
  return -1;
}
function Kn(t, e, n) {
  return e === e ? Ha(t, e, n) : Ss(t, Ba, n);
}
function Os(t, e) {
  var n = t == null ? 0 : t.length;
  return !!n && Kn(t, e, 0) > -1;
}
var Ka = 9007199254740991, Wa = /^(?:0|[1-9]\d*)$/;
function Qt(t, e) {
  var n = typeof t;
  return e = e ?? Ka, !!e && (n == "number" || n != "symbol" && Wa.test(t)) && t > -1 && t % 1 == 0 && t < e;
}
function Wn(t, e, n) {
  e == "__proto__" && Dt ? Dt(t, e, {
    configurable: !0,
    enumerable: !0,
    value: n,
    writable: !0
  }) : t[e] = n;
}
function Et(t, e) {
  return t === e || t !== t && e !== e;
}
var ja = Object.prototype, za = ja.hasOwnProperty;
function Jt(t, e, n) {
  var r = t[e];
  (!(za.call(t, e) && Et(r, n)) || n === void 0 && !(e in t)) && Wn(t, e, n);
}
function jn(t, e, n, r) {
  var s = !n;
  n || (n = {});
  for (var i = -1, a = e.length; ++i < a; ) {
    var o = e[i], c = void 0;
    c === void 0 && (c = t[o]), s ? Wn(n, o, c) : Jt(n, o, c);
  }
  return n;
}
var Or = Math.max;
function Va(t, e, n) {
  return e = Or(e === void 0 ? t.length - 1 : e, 0), function() {
    for (var r = arguments, s = -1, i = Or(r.length - e, 0), a = Array(i); ++s < i; )
      a[s] = r[e + s];
    s = -1;
    for (var o = Array(e + 1); ++s < e; )
      o[s] = r[s];
    return o[e] = n(a), ka(t, this, o);
  };
}
function zn(t, e) {
  return Ga(Va(t, e, gt), t + "");
}
var Ya = 9007199254740991;
function Vn(t) {
  return typeof t == "number" && t > -1 && t % 1 == 0 && t <= Ya;
}
function Te(t) {
  return t != null && Vn(t.length) && !Ce(t);
}
function Ns(t, e, n) {
  if (!te(n))
    return !1;
  var r = typeof e;
  return (r == "number" ? Te(n) && Qt(e, n.length) : r == "string" && e in n) ? Et(n[e], t) : !1;
}
function Xa(t) {
  return zn(function(e, n) {
    var r = -1, s = n.length, i = s > 1 ? n[s - 1] : void 0, a = s > 2 ? n[2] : void 0;
    for (i = t.length > 3 && typeof i == "function" ? (s--, i) : void 0, a && Ns(n[0], n[1], a) && (i = s < 3 ? void 0 : i, s = 1), e = Object(e); ++r < s; ) {
      var o = n[r];
      o && t(e, o, r, i);
    }
    return e;
  });
}
var qa = Object.prototype;
function Tt(t) {
  var e = t && t.constructor, n = typeof e == "function" && e.prototype || qa;
  return t === n;
}
function Za(t, e) {
  for (var n = -1, r = Array(t); ++n < t; )
    r[n] = e(n);
  return r;
}
var Qa = "[object Arguments]";
function Nr(t) {
  return ce(t) && be(t) == Qa;
}
var Cs = Object.prototype, Ja = Cs.hasOwnProperty, eo = Cs.propertyIsEnumerable, en = Nr(/* @__PURE__ */ (function() {
  return arguments;
})()) ? Nr : function(t) {
  return ce(t) && Ja.call(t, "callee") && !eo.call(t, "callee");
};
function to() {
  return !1;
}
var Ls = typeof exports == "object" && exports && !exports.nodeType && exports, Cr = Ls && typeof module == "object" && module && !module.nodeType && module, no = Cr && Cr.exports === Ls, Lr = no ? Ee.Buffer : void 0, ro = Lr ? Lr.isBuffer : void 0, ht = ro || to, so = "[object Arguments]", io = "[object Array]", ao = "[object Boolean]", oo = "[object Date]", co = "[object Error]", uo = "[object Function]", lo = "[object Map]", ho = "[object Number]", fo = "[object Object]", po = "[object RegExp]", go = "[object Set]", Eo = "[object String]", To = "[object WeakMap]", mo = "[object ArrayBuffer]", Ao = "[object DataView]", Ro = "[object Float32Array]", Io = "[object Float64Array]", _o = "[object Int8Array]", So = "[object Int16Array]", Oo = "[object Int32Array]", No = "[object Uint8Array]", Co = "[object Uint8ClampedArray]", Lo = "[object Uint16Array]", yo = "[object Uint32Array]", P = {};
P[Ro] = P[Io] = P[_o] = P[So] = P[Oo] = P[No] = P[Co] = P[Lo] = P[yo] = !0;
P[so] = P[io] = P[mo] = P[ao] = P[Ao] = P[oo] = P[co] = P[uo] = P[lo] = P[ho] = P[fo] = P[po] = P[go] = P[Eo] = P[To] = !1;
function vo(t) {
  return ce(t) && Vn(t.length) && !!P[be(t)];
}
function tn(t) {
  return function(e) {
    return t(e);
  };
}
var ys = typeof exports == "object" && exports && !exports.nodeType && exports, ut = ys && typeof module == "object" && module && !module.nodeType && module, Po = ut && ut.exports === ys, Tn = Po && As.process, ke = (function() {
  try {
    var t = ut && ut.require && ut.require("util").types;
    return t || Tn && Tn.binding && Tn.binding("util");
  } catch {
  }
})(), yr = ke && ke.isTypedArray, Yn = yr ? tn(yr) : vo, Mo = Object.prototype, ko = Mo.hasOwnProperty;
function vs(t, e) {
  var n = N(t), r = !n && en(t), s = !n && !r && ht(t), i = !n && !r && !s && Yn(t), a = n || r || s || i, o = a ? Za(t.length, String) : [], c = o.length;
  for (var u in t)
    (e || ko.call(t, u)) && !(a && // Safari 9 has enumerable `arguments.length` in strict mode.
    (u == "length" || // Node.js 0.10 has enumerable non-index properties on buffers.
    s && (u == "offset" || u == "parent") || // PhantomJS 2 has enumerable non-index properties on typed arrays.
    i && (u == "buffer" || u == "byteLength" || u == "byteOffset") || // Skip index properties.
    Qt(u, c))) && o.push(u);
  return o;
}
function Ps(t, e) {
  return function(n) {
    return t(e(n));
  };
}
var xo = Ps(Object.keys, Object), bo = Object.prototype, wo = bo.hasOwnProperty;
function Ms(t) {
  if (!Tt(t))
    return xo(t);
  var e = [];
  for (var n in Object(t))
    wo.call(t, n) && n != "constructor" && e.push(n);
  return e;
}
function ne(t) {
  return Te(t) ? vs(t) : Ms(t);
}
var Uo = Object.prototype, Fo = Uo.hasOwnProperty, V = Xa(function(t, e) {
  if (Tt(e) || Te(e)) {
    jn(e, ne(e), t);
    return;
  }
  for (var n in e)
    Fo.call(e, n) && Jt(t, n, e[n]);
});
function Do(t) {
  var e = [];
  if (t != null)
    for (var n in Object(t))
      e.push(n);
  return e;
}
var $o = Object.prototype, Go = $o.hasOwnProperty;
function Bo(t) {
  if (!te(t))
    return Do(t);
  var e = Tt(t), n = [];
  for (var r in t)
    r == "constructor" && (e || !Go.call(t, r)) || n.push(r);
  return n;
}
function ks(t) {
  return Te(t) ? vs(t, !0) : Bo(t);
}
var Ho = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/, Ko = /^\w*$/;
function Xn(t, e) {
  if (N(t))
    return !1;
  var n = typeof t;
  return n == "number" || n == "symbol" || n == "boolean" || t == null || Xt(t) ? !0 : Ko.test(t) || !Ho.test(t) || e != null && t in Object(e);
}
var ft = We(Object, "create");
function Wo() {
  this.__data__ = ft ? ft(null) : {}, this.size = 0;
}
function jo(t) {
  var e = this.has(t) && delete this.__data__[t];
  return this.size -= e ? 1 : 0, e;
}
var zo = "__lodash_hash_undefined__", Vo = Object.prototype, Yo = Vo.hasOwnProperty;
function Xo(t) {
  var e = this.__data__;
  if (ft) {
    var n = e[t];
    return n === zo ? void 0 : n;
  }
  return Yo.call(e, t) ? e[t] : void 0;
}
var qo = Object.prototype, Zo = qo.hasOwnProperty;
function Qo(t) {
  var e = this.__data__;
  return ft ? e[t] !== void 0 : Zo.call(e, t);
}
var Jo = "__lodash_hash_undefined__";
function ec(t, e) {
  var n = this.__data__;
  return this.size += this.has(t) ? 0 : 1, n[t] = ft && e === void 0 ? Jo : e, this;
}
function $e(t) {
  var e = -1, n = t == null ? 0 : t.length;
  for (this.clear(); ++e < n; ) {
    var r = t[e];
    this.set(r[0], r[1]);
  }
}
$e.prototype.clear = Wo;
$e.prototype.delete = jo;
$e.prototype.get = Xo;
$e.prototype.has = Qo;
$e.prototype.set = ec;
function tc() {
  this.__data__ = [], this.size = 0;
}
function nn(t, e) {
  for (var n = t.length; n--; )
    if (Et(t[n][0], e))
      return n;
  return -1;
}
var nc = Array.prototype, rc = nc.splice;
function sc(t) {
  var e = this.__data__, n = nn(e, t);
  if (n < 0)
    return !1;
  var r = e.length - 1;
  return n == r ? e.pop() : rc.call(e, n, 1), --this.size, !0;
}
function ic(t) {
  var e = this.__data__, n = nn(e, t);
  return n < 0 ? void 0 : e[n][1];
}
function ac(t) {
  return nn(this.__data__, t) > -1;
}
function oc(t, e) {
  var n = this.__data__, r = nn(n, t);
  return r < 0 ? (++this.size, n.push([t, e])) : n[r][1] = e, this;
}
function Le(t) {
  var e = -1, n = t == null ? 0 : t.length;
  for (this.clear(); ++e < n; ) {
    var r = t[e];
    this.set(r[0], r[1]);
  }
}
Le.prototype.clear = tc;
Le.prototype.delete = sc;
Le.prototype.get = ic;
Le.prototype.has = ac;
Le.prototype.set = oc;
var dt = We(Ee, "Map");
function cc() {
  this.size = 0, this.__data__ = {
    hash: new $e(),
    map: new (dt || Le)(),
    string: new $e()
  };
}
function uc(t) {
  var e = typeof t;
  return e == "string" || e == "number" || e == "symbol" || e == "boolean" ? t !== "__proto__" : t === null;
}
function rn(t, e) {
  var n = t.__data__;
  return uc(e) ? n[typeof e == "string" ? "string" : "hash"] : n.map;
}
function lc(t) {
  var e = rn(this, t).delete(t);
  return this.size -= e ? 1 : 0, e;
}
function hc(t) {
  return rn(this, t).get(t);
}
function fc(t) {
  return rn(this, t).has(t);
}
function dc(t, e) {
  var n = rn(this, t), r = n.size;
  return n.set(t, e), this.size += n.size == r ? 0 : 1, this;
}
function ye(t) {
  var e = -1, n = t == null ? 0 : t.length;
  for (this.clear(); ++e < n; ) {
    var r = t[e];
    this.set(r[0], r[1]);
  }
}
ye.prototype.clear = cc;
ye.prototype.delete = lc;
ye.prototype.get = hc;
ye.prototype.has = fc;
ye.prototype.set = dc;
var pc = "Expected a function";
function qn(t, e) {
  if (typeof t != "function" || e != null && typeof e != "function")
    throw new TypeError(pc);
  var n = function() {
    var r = arguments, s = e ? e.apply(this, r) : r[0], i = n.cache;
    if (i.has(s))
      return i.get(s);
    var a = t.apply(this, r);
    return n.cache = i.set(s, a) || i, a;
  };
  return n.cache = new (qn.Cache || ye)(), n;
}
qn.Cache = ye;
var gc = 500;
function Ec(t) {
  var e = qn(t, function(r) {
    return n.size === gc && n.clear(), r;
  }), n = e.cache;
  return e;
}
var Tc = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g, mc = /\\(\\)?/g, Ac = Ec(function(t) {
  var e = [];
  return t.charCodeAt(0) === 46 && e.push(""), t.replace(Tc, function(n, r, s, i) {
    e.push(s ? i.replace(mc, "$1") : r || n);
  }), e;
});
function Rc(t) {
  return t == null ? "" : Is(t);
}
function sn(t, e) {
  return N(t) ? t : Xn(t, e) ? [t] : Ac(Rc(t));
}
function mt(t) {
  if (typeof t == "string" || Xt(t))
    return t;
  var e = t + "";
  return e == "0" && 1 / t == -1 / 0 ? "-0" : e;
}
function Zn(t, e) {
  e = sn(e, t);
  for (var n = 0, r = e.length; t != null && n < r; )
    t = t[mt(e[n++])];
  return n && n == r ? t : void 0;
}
function Ic(t, e, n) {
  var r = t == null ? void 0 : Zn(t, e);
  return r === void 0 ? n : r;
}
function Qn(t, e) {
  for (var n = -1, r = e.length, s = t.length; ++n < r; )
    t[s + n] = e[n];
  return t;
}
var vr = ee ? ee.isConcatSpreadable : void 0;
function _c(t) {
  return N(t) || en(t) || !!(vr && t && t[vr]);
}
function Jn(t, e, n, r, s) {
  var i = -1, a = t.length;
  for (n || (n = _c), s || (s = []); ++i < a; ) {
    var o = t[i];
    n(o) ? Qn(s, o) : r || (s[s.length] = o);
  }
  return s;
}
function ae(t) {
  var e = t == null ? 0 : t.length;
  return e ? Jn(t) : [];
}
var xs = Ps(Object.getPrototypeOf, Object);
function bs(t, e, n) {
  var r = -1, s = t.length;
  e < 0 && (e = -e > s ? 0 : s + e), n = n > s ? s : n, n < 0 && (n += s), s = e > n ? 0 : n - e >>> 0, e >>>= 0;
  for (var i = Array(s); ++r < s; )
    i[r] = t[r + e];
  return i;
}
function Sc(t, e, n, r) {
  var s = -1, i = t == null ? 0 : t.length;
  for (r && i && (n = t[++s]); ++s < i; )
    n = e(n, t[s], s, t);
  return n;
}
function Oc() {
  this.__data__ = new Le(), this.size = 0;
}
function Nc(t) {
  var e = this.__data__, n = e.delete(t);
  return this.size = e.size, n;
}
function Cc(t) {
  return this.__data__.get(t);
}
function Lc(t) {
  return this.__data__.has(t);
}
var yc = 200;
function vc(t, e) {
  var n = this.__data__;
  if (n instanceof Le) {
    var r = n.__data__;
    if (!dt || r.length < yc - 1)
      return r.push([t, e]), this.size = ++n.size, this;
    n = this.__data__ = new ye(r);
  }
  return n.set(t, e), this.size = n.size, this;
}
function ge(t) {
  var e = this.__data__ = new Le(t);
  this.size = e.size;
}
ge.prototype.clear = Oc;
ge.prototype.delete = Nc;
ge.prototype.get = Cc;
ge.prototype.has = Lc;
ge.prototype.set = vc;
function Pc(t, e) {
  return t && jn(e, ne(e), t);
}
var ws = typeof exports == "object" && exports && !exports.nodeType && exports, Pr = ws && typeof module == "object" && module && !module.nodeType && module, Mc = Pr && Pr.exports === ws, Mr = Mc ? Ee.Buffer : void 0, kr = Mr ? Mr.allocUnsafe : void 0;
function kc(t, e) {
  var n = t.length, r = kr ? kr(n) : new t.constructor(n);
  return t.copy(r), r;
}
function er(t, e) {
  for (var n = -1, r = t == null ? 0 : t.length, s = 0, i = []; ++n < r; ) {
    var a = t[n];
    e(a, n, t) && (i[s++] = a);
  }
  return i;
}
function Us() {
  return [];
}
var xc = Object.prototype, bc = xc.propertyIsEnumerable, xr = Object.getOwnPropertySymbols, tr = xr ? function(t) {
  return t == null ? [] : (t = Object(t), er(xr(t), function(e) {
    return bc.call(t, e);
  }));
} : Us;
function wc(t, e) {
  return jn(t, tr(t), e);
}
var Uc = Object.getOwnPropertySymbols, Fc = Uc ? function(t) {
  for (var e = []; t; )
    Qn(e, tr(t)), t = xs(t);
  return e;
} : Us;
function Fs(t, e, n) {
  var r = e(t);
  return N(t) ? r : Qn(r, n(t));
}
function Nn(t) {
  return Fs(t, ne, tr);
}
function Dc(t) {
  return Fs(t, ks, Fc);
}
var Cn = We(Ee, "DataView"), Ln = We(Ee, "Promise"), Ve = We(Ee, "Set"), br = "[object Map]", $c = "[object Object]", wr = "[object Promise]", Ur = "[object Set]", Fr = "[object WeakMap]", Dr = "[object DataView]", Gc = Ke(Cn), Bc = Ke(dt), Hc = Ke(Ln), Kc = Ke(Ve), Wc = Ke(On), Q = be;
(Cn && Q(new Cn(new ArrayBuffer(1))) != Dr || dt && Q(new dt()) != br || Ln && Q(Ln.resolve()) != wr || Ve && Q(new Ve()) != Ur || On && Q(new On()) != Fr) && (Q = function(t) {
  var e = be(t), n = e == $c ? t.constructor : void 0, r = n ? Ke(n) : "";
  if (r)
    switch (r) {
      case Gc:
        return Dr;
      case Bc:
        return br;
      case Hc:
        return wr;
      case Kc:
        return Ur;
      case Wc:
        return Fr;
    }
  return e;
});
var jc = Object.prototype, zc = jc.hasOwnProperty;
function Vc(t) {
  var e = t.length, n = new t.constructor(e);
  return e && typeof t[0] == "string" && zc.call(t, "index") && (n.index = t.index, n.input = t.input), n;
}
var $t = Ee.Uint8Array;
function Yc(t) {
  var e = new t.constructor(t.byteLength);
  return new $t(e).set(new $t(t)), e;
}
function Xc(t, e) {
  var n = t.buffer;
  return new t.constructor(n, t.byteOffset, t.byteLength);
}
var qc = /\w*$/;
function Zc(t) {
  var e = new t.constructor(t.source, qc.exec(t));
  return e.lastIndex = t.lastIndex, e;
}
var $r = ee ? ee.prototype : void 0, Gr = $r ? $r.valueOf : void 0;
function Qc(t) {
  return Gr ? Object(Gr.call(t)) : {};
}
function Jc(t, e) {
  var n = t.buffer;
  return new t.constructor(n, t.byteOffset, t.length);
}
var eu = "[object Boolean]", tu = "[object Date]", nu = "[object Map]", ru = "[object Number]", su = "[object RegExp]", iu = "[object Set]", au = "[object String]", ou = "[object Symbol]", cu = "[object ArrayBuffer]", uu = "[object DataView]", lu = "[object Float32Array]", hu = "[object Float64Array]", fu = "[object Int8Array]", du = "[object Int16Array]", pu = "[object Int32Array]", gu = "[object Uint8Array]", Eu = "[object Uint8ClampedArray]", Tu = "[object Uint16Array]", mu = "[object Uint32Array]";
function Au(t, e, n) {
  var r = t.constructor;
  switch (e) {
    case cu:
      return Yc(t);
    case eu:
    case tu:
      return new r(+t);
    case uu:
      return Xc(t);
    case lu:
    case hu:
    case fu:
    case du:
    case pu:
    case gu:
    case Eu:
    case Tu:
    case mu:
      return Jc(t);
    case nu:
      return new r();
    case ru:
    case au:
      return new r(t);
    case su:
      return Zc(t);
    case iu:
      return new r();
    case ou:
      return Qc(t);
  }
}
function Ru(t) {
  return typeof t.constructor == "function" && !Tt(t) ? Ma(xs(t)) : {};
}
var Iu = "[object Map]";
function _u(t) {
  return ce(t) && Q(t) == Iu;
}
var Br = ke && ke.isMap, Su = Br ? tn(Br) : _u, Ou = "[object Set]";
function Nu(t) {
  return ce(t) && Q(t) == Ou;
}
var Hr = ke && ke.isSet, Cu = Hr ? tn(Hr) : Nu, Ds = "[object Arguments]", Lu = "[object Array]", yu = "[object Boolean]", vu = "[object Date]", Pu = "[object Error]", $s = "[object Function]", Mu = "[object GeneratorFunction]", ku = "[object Map]", xu = "[object Number]", Gs = "[object Object]", bu = "[object RegExp]", wu = "[object Set]", Uu = "[object String]", Fu = "[object Symbol]", Du = "[object WeakMap]", $u = "[object ArrayBuffer]", Gu = "[object DataView]", Bu = "[object Float32Array]", Hu = "[object Float64Array]", Ku = "[object Int8Array]", Wu = "[object Int16Array]", ju = "[object Int32Array]", zu = "[object Uint8Array]", Vu = "[object Uint8ClampedArray]", Yu = "[object Uint16Array]", Xu = "[object Uint32Array]", y = {};
y[Ds] = y[Lu] = y[$u] = y[Gu] = y[yu] = y[vu] = y[Bu] = y[Hu] = y[Ku] = y[Wu] = y[ju] = y[ku] = y[xu] = y[Gs] = y[bu] = y[wu] = y[Uu] = y[Fu] = y[zu] = y[Vu] = y[Yu] = y[Xu] = !0;
y[Pu] = y[$s] = y[Du] = !1;
function Mt(t, e, n, r, s, i) {
  var a;
  if (a !== void 0)
    return a;
  if (!te(t))
    return t;
  var o = N(t);
  if (o)
    return a = Vc(t), xa(t, a);
  var c = Q(t), u = c == $s || c == Mu;
  if (ht(t))
    return kc(t);
  if (c == Gs || c == Ds || u && !s)
    return a = u ? {} : Ru(t), wc(t, Pc(a, t));
  if (!y[c])
    return s ? t : {};
  a = Au(t, c), i || (i = new ge());
  var l = i.get(t);
  if (l)
    return l;
  i.set(t, a), Cu(t) ? t.forEach(function(p) {
    a.add(Mt(p, e, n, p, t, i));
  }) : Su(t) && t.forEach(function(p, m) {
    a.set(m, Mt(p, e, n, m, t, i));
  });
  var h = Nn, f = o ? void 0 : h(t);
  return _s(f || t, function(p, m) {
    f && (m = p, p = t[m]), Jt(a, m, Mt(p, e, n, m, t, i));
  }), a;
}
var qu = 4;
function D(t) {
  return Mt(t, qu);
}
function At(t) {
  for (var e = -1, n = t == null ? 0 : t.length, r = 0, s = []; ++e < n; ) {
    var i = t[e];
    i && (s[r++] = i);
  }
  return s;
}
var Zu = "__lodash_hash_undefined__";
function Qu(t) {
  return this.__data__.set(t, Zu), this;
}
function Ju(t) {
  return this.__data__.has(t);
}
function Ye(t) {
  var e = -1, n = t == null ? 0 : t.length;
  for (this.__data__ = new ye(); ++e < n; )
    this.add(t[e]);
}
Ye.prototype.add = Ye.prototype.push = Qu;
Ye.prototype.has = Ju;
function Bs(t, e) {
  for (var n = -1, r = t == null ? 0 : t.length; ++n < r; )
    if (e(t[n], n, t))
      return !0;
  return !1;
}
function nr(t, e) {
  return t.has(e);
}
var el = 1, tl = 2;
function Hs(t, e, n, r, s, i) {
  var a = n & el, o = t.length, c = e.length;
  if (o != c && !(a && c > o))
    return !1;
  var u = i.get(t), l = i.get(e);
  if (u && l)
    return u == e && l == t;
  var h = -1, f = !0, p = n & tl ? new Ye() : void 0;
  for (i.set(t, e), i.set(e, t); ++h < o; ) {
    var m = t[h], I = e[h];
    if (r)
      var O = a ? r(I, m, h, e, t, i) : r(m, I, h, t, e, i);
    if (O !== void 0) {
      if (O)
        continue;
      f = !1;
      break;
    }
    if (p) {
      if (!Bs(e, function(T, g) {
        if (!nr(p, g) && (m === T || s(m, T, n, r, i)))
          return p.push(g);
      })) {
        f = !1;
        break;
      }
    } else if (!(m === I || s(m, I, n, r, i))) {
      f = !1;
      break;
    }
  }
  return i.delete(t), i.delete(e), f;
}
function nl(t) {
  var e = -1, n = Array(t.size);
  return t.forEach(function(r, s) {
    n[++e] = [s, r];
  }), n;
}
function rr(t) {
  var e = -1, n = Array(t.size);
  return t.forEach(function(r) {
    n[++e] = r;
  }), n;
}
var rl = 1, sl = 2, il = "[object Boolean]", al = "[object Date]", ol = "[object Error]", cl = "[object Map]", ul = "[object Number]", ll = "[object RegExp]", hl = "[object Set]", fl = "[object String]", dl = "[object Symbol]", pl = "[object ArrayBuffer]", gl = "[object DataView]", Kr = ee ? ee.prototype : void 0, mn = Kr ? Kr.valueOf : void 0;
function El(t, e, n, r, s, i, a) {
  switch (n) {
    case gl:
      if (t.byteLength != e.byteLength || t.byteOffset != e.byteOffset)
        return !1;
      t = t.buffer, e = e.buffer;
    case pl:
      return !(t.byteLength != e.byteLength || !i(new $t(t), new $t(e)));
    case il:
    case al:
    case ul:
      return Et(+t, +e);
    case ol:
      return t.name == e.name && t.message == e.message;
    case ll:
    case fl:
      return t == e + "";
    case cl:
      var o = nl;
    case hl:
      var c = r & rl;
      if (o || (o = rr), t.size != e.size && !c)
        return !1;
      var u = a.get(t);
      if (u)
        return u == e;
      r |= sl, a.set(t, e);
      var l = Hs(o(t), o(e), r, s, i, a);
      return a.delete(t), l;
    case dl:
      if (mn)
        return mn.call(t) == mn.call(e);
  }
  return !1;
}
var Tl = 1, ml = Object.prototype, Al = ml.hasOwnProperty;
function Rl(t, e, n, r, s, i) {
  var a = n & Tl, o = Nn(t), c = o.length, u = Nn(e), l = u.length;
  if (c != l && !a)
    return !1;
  for (var h = c; h--; ) {
    var f = o[h];
    if (!(a ? f in e : Al.call(e, f)))
      return !1;
  }
  var p = i.get(t), m = i.get(e);
  if (p && m)
    return p == e && m == t;
  var I = !0;
  i.set(t, e), i.set(e, t);
  for (var O = a; ++h < c; ) {
    f = o[h];
    var T = t[f], g = e[f];
    if (r)
      var d = a ? r(g, T, f, e, t, i) : r(T, g, f, t, e, i);
    if (!(d === void 0 ? T === g || s(T, g, n, r, i) : d)) {
      I = !1;
      break;
    }
    O || (O = f == "constructor");
  }
  if (I && !O) {
    var E = t.constructor, L = e.constructor;
    E != L && "constructor" in t && "constructor" in e && !(typeof E == "function" && E instanceof E && typeof L == "function" && L instanceof L) && (I = !1);
  }
  return i.delete(t), i.delete(e), I;
}
var Il = 1, Wr = "[object Arguments]", jr = "[object Array]", Nt = "[object Object]", _l = Object.prototype, zr = _l.hasOwnProperty;
function Sl(t, e, n, r, s, i) {
  var a = N(t), o = N(e), c = a ? jr : Q(t), u = o ? jr : Q(e);
  c = c == Wr ? Nt : c, u = u == Wr ? Nt : u;
  var l = c == Nt, h = u == Nt, f = c == u;
  if (f && ht(t)) {
    if (!ht(e))
      return !1;
    a = !0, l = !1;
  }
  if (f && !l)
    return i || (i = new ge()), a || Yn(t) ? Hs(t, e, n, r, s, i) : El(t, e, c, n, r, s, i);
  if (!(n & Il)) {
    var p = l && zr.call(t, "__wrapped__"), m = h && zr.call(e, "__wrapped__");
    if (p || m) {
      var I = p ? t.value() : t, O = m ? e.value() : e;
      return i || (i = new ge()), s(I, O, n, r, i);
    }
  }
  return f ? (i || (i = new ge()), Rl(t, e, n, r, s, i)) : !1;
}
function sr(t, e, n, r, s) {
  return t === e ? !0 : t == null || e == null || !ce(t) && !ce(e) ? t !== t && e !== e : Sl(t, e, n, r, sr, s);
}
var Ol = 1, Nl = 2;
function Cl(t, e, n, r) {
  var s = n.length, i = s;
  if (t == null)
    return !i;
  for (t = Object(t); s--; ) {
    var a = n[s];
    if (a[2] ? a[1] !== t[a[0]] : !(a[0] in t))
      return !1;
  }
  for (; ++s < i; ) {
    a = n[s];
    var o = a[0], c = t[o], u = a[1];
    if (a[2]) {
      if (c === void 0 && !(o in t))
        return !1;
    } else {
      var l = new ge(), h;
      if (!(h === void 0 ? sr(u, c, Ol | Nl, r, l) : h))
        return !1;
    }
  }
  return !0;
}
function Ks(t) {
  return t === t && !te(t);
}
function Ll(t) {
  for (var e = ne(t), n = e.length; n--; ) {
    var r = e[n], s = t[r];
    e[n] = [r, s, Ks(s)];
  }
  return e;
}
function Ws(t, e) {
  return function(n) {
    return n == null ? !1 : n[t] === e && (e !== void 0 || t in Object(n));
  };
}
function yl(t) {
  var e = Ll(t);
  return e.length == 1 && e[0][2] ? Ws(e[0][0], e[0][1]) : function(n) {
    return n === t || Cl(n, t, e);
  };
}
function vl(t, e) {
  return t != null && e in Object(t);
}
function js(t, e, n) {
  e = sn(e, t);
  for (var r = -1, s = e.length, i = !1; ++r < s; ) {
    var a = mt(e[r]);
    if (!(i = t != null && n(t, a)))
      break;
    t = t[a];
  }
  return i || ++r != s ? i : (s = t == null ? 0 : t.length, !!s && Vn(s) && Qt(a, s) && (N(t) || en(t)));
}
function Pl(t, e) {
  return t != null && js(t, e, vl);
}
var Ml = 1, kl = 2;
function xl(t, e) {
  return Xn(t) && Ks(e) ? Ws(mt(t), e) : function(n) {
    var r = Ic(n, t);
    return r === void 0 && r === e ? Pl(n, t) : sr(e, r, Ml | kl);
  };
}
function bl(t) {
  return function(e) {
    return e?.[t];
  };
}
function wl(t) {
  return function(e) {
    return Zn(e, t);
  };
}
function Ul(t) {
  return Xn(t) ? bl(mt(t)) : wl(t);
}
function me(t) {
  return typeof t == "function" ? t : t == null ? gt : typeof t == "object" ? N(t) ? xl(t[0], t[1]) : yl(t) : Ul(t);
}
function Fl(t, e, n, r) {
  for (var s = -1, i = t == null ? 0 : t.length; ++s < i; ) {
    var a = t[s];
    e(r, a, n(a), t);
  }
  return r;
}
function Dl(t) {
  return function(e, n, r) {
    for (var s = -1, i = Object(e), a = r(e), o = a.length; o--; ) {
      var c = a[++s];
      if (n(i[c], c, i) === !1)
        break;
    }
    return e;
  };
}
var $l = Dl();
function Gl(t, e) {
  return t && $l(t, e, ne);
}
function Bl(t, e) {
  return function(n, r) {
    if (n == null)
      return n;
    if (!Te(n))
      return t(n, r);
    for (var s = n.length, i = -1, a = Object(n); ++i < s && r(a[i], i, a) !== !1; )
      ;
    return n;
  };
}
var je = Bl(Gl);
function Hl(t, e, n, r) {
  return je(t, function(s, i, a) {
    e(r, s, n(s), a);
  }), r;
}
function Kl(t, e) {
  return function(n, r) {
    var s = N(n) ? Fl : Hl, i = e ? e() : {};
    return s(n, t, me(r), i);
  };
}
var zs = Object.prototype, Wl = zs.hasOwnProperty, ir = zn(function(t, e) {
  t = Object(t);
  var n = -1, r = e.length, s = r > 2 ? e[2] : void 0;
  for (s && Ns(e[0], e[1], s) && (r = 1); ++n < r; )
    for (var i = e[n], a = ks(i), o = -1, c = a.length; ++o < c; ) {
      var u = a[o], l = t[u];
      (l === void 0 || Et(l, zs[u]) && !Wl.call(t, u)) && (t[u] = i[u]);
    }
  return t;
});
function Vr(t) {
  return ce(t) && Te(t);
}
var jl = 200;
function zl(t, e, n, r) {
  var s = -1, i = Os, a = !0, o = t.length, c = [], u = e.length;
  if (!o)
    return c;
  e.length >= jl && (i = nr, a = !1, e = new Ye(e));
  e:
    for (; ++s < o; ) {
      var l = t[s], h = l;
      if (l = l !== 0 ? l : 0, a && h === h) {
        for (var f = u; f--; )
          if (e[f] === h)
            continue e;
        c.push(l);
      } else i(e, h, r) || c.push(l);
    }
  return c;
}
var an = zn(function(t, e) {
  return Vr(t) ? zl(t, Jn(e, 1, Vr, !0)) : [];
});
function Ge(t) {
  var e = t == null ? 0 : t.length;
  return e ? t[e - 1] : void 0;
}
function F(t, e, n) {
  var r = t == null ? 0 : t.length;
  return r ? (e = e === void 0 ? 1 : Zt(e), bs(t, e < 0 ? 0 : e, r)) : [];
}
function pt(t, e, n) {
  var r = t == null ? 0 : t.length;
  return r ? (e = e === void 0 ? 1 : Zt(e), e = r - e, bs(t, 0, e < 0 ? 0 : e)) : [];
}
function Vl(t) {
  return typeof t == "function" ? t : gt;
}
function _(t, e) {
  var n = N(t) ? _s : je;
  return n(t, Vl(e));
}
function Yl(t, e) {
  for (var n = -1, r = t == null ? 0 : t.length; ++n < r; )
    if (!e(t[n], n, t))
      return !1;
  return !0;
}
function Xl(t, e) {
  var n = !0;
  return je(t, function(r, s, i) {
    return n = !!e(r, s, i), n;
  }), n;
}
function oe(t, e, n) {
  var r = N(t) ? Yl : Xl;
  return r(t, me(e));
}
function Vs(t, e) {
  var n = [];
  return je(t, function(r, s, i) {
    e(r, s, i) && n.push(r);
  }), n;
}
function re(t, e) {
  var n = N(t) ? er : Vs;
  return n(t, me(e));
}
function ql(t) {
  return function(e, n, r) {
    var s = Object(e);
    if (!Te(e)) {
      var i = me(n);
      e = ne(e), n = function(o) {
        return i(s[o], o, s);
      };
    }
    var a = t(e, n, r);
    return a > -1 ? s[i ? e[a] : a] : void 0;
  };
}
var Zl = Math.max;
function Ql(t, e, n) {
  var r = t == null ? 0 : t.length;
  if (!r)
    return -1;
  var s = n == null ? 0 : Zt(n);
  return s < 0 && (s = Zl(r + s, 0)), Ss(t, me(e), s);
}
var Xe = ql(Ql);
function ue(t) {
  return t && t.length ? t[0] : void 0;
}
function Jl(t, e) {
  var n = -1, r = Te(t) ? Array(t.length) : [];
  return je(t, function(s, i, a) {
    r[++n] = e(s, i, a);
  }), r;
}
function A(t, e) {
  var n = N(t) ? qt : Jl;
  return n(t, me(e));
}
function J(t, e) {
  return Jn(A(t, e));
}
var eh = Object.prototype, th = eh.hasOwnProperty, nh = Kl(function(t, e, n) {
  th.call(t, n) ? t[n].push(e) : Wn(t, n, [e]);
}), rh = Object.prototype, sh = rh.hasOwnProperty;
function ih(t, e) {
  return t != null && sh.call(t, e);
}
function R(t, e) {
  return t != null && js(t, e, ih);
}
var ah = "[object String]";
function j(t) {
  return typeof t == "string" || !N(t) && ce(t) && be(t) == ah;
}
function oh(t, e) {
  return qt(e, function(n) {
    return t[n];
  });
}
function b(t) {
  return t == null ? [] : oh(t, ne(t));
}
var ch = Math.max;
function H(t, e, n, r) {
  t = Te(t) ? t : b(t), n = n ? Zt(n) : 0;
  var s = t.length;
  return n < 0 && (n = ch(s + n, 0)), j(t) ? n <= s && t.indexOf(e, n) > -1 : !!s && Kn(t, e, n) > -1;
}
function Yr(t, e, n) {
  var r = t == null ? 0 : t.length;
  if (!r)
    return -1;
  var s = 0;
  return Kn(t, e, s);
}
var uh = "[object Map]", lh = "[object Set]", hh = Object.prototype, fh = hh.hasOwnProperty;
function v(t) {
  if (t == null)
    return !0;
  if (Te(t) && (N(t) || typeof t == "string" || typeof t.splice == "function" || ht(t) || Yn(t) || en(t)))
    return !t.length;
  var e = Q(t);
  if (e == uh || e == lh)
    return !t.size;
  if (Tt(t))
    return !Ms(t).length;
  for (var n in t)
    if (fh.call(t, n))
      return !1;
  return !0;
}
var dh = "[object RegExp]";
function ph(t) {
  return ce(t) && be(t) == dh;
}
var Xr = ke && ke.isRegExp, Se = Xr ? tn(Xr) : ph;
function Oe(t) {
  return t === void 0;
}
var gh = "Expected a function";
function Eh(t) {
  if (typeof t != "function")
    throw new TypeError(gh);
  return function() {
    var e = arguments;
    switch (e.length) {
      case 0:
        return !t.call(this);
      case 1:
        return !t.call(this, e[0]);
      case 2:
        return !t.call(this, e[0], e[1]);
      case 3:
        return !t.call(this, e[0], e[1], e[2]);
    }
    return !t.apply(this, e);
  };
}
function Th(t, e, n, r) {
  if (!te(t))
    return t;
  e = sn(e, t);
  for (var s = -1, i = e.length, a = i - 1, o = t; o != null && ++s < i; ) {
    var c = mt(e[s]), u = n;
    if (c === "__proto__" || c === "constructor" || c === "prototype")
      return t;
    if (s != a) {
      var l = o[c];
      u = void 0, u === void 0 && (u = te(l) ? l : Qt(e[s + 1]) ? [] : {});
    }
    Jt(o, c, u), o = o[c];
  }
  return t;
}
function mh(t, e, n) {
  for (var r = -1, s = e.length, i = {}; ++r < s; ) {
    var a = e[r], o = Zn(t, a);
    n(o, a) && Th(i, sn(a, t), o);
  }
  return i;
}
function le(t, e) {
  if (t == null)
    return {};
  var n = qt(Dc(t), function(r) {
    return [r];
  });
  return e = me(e), mh(t, n, function(r, s) {
    return e(r, s[0]);
  });
}
function Ah(t, e, n, r, s) {
  return s(t, function(i, a, o) {
    n = r ? (r = !1, i) : e(n, i, a, o);
  }), n;
}
function Y(t, e, n) {
  var r = N(t) ? Sc : Ah, s = arguments.length < 3;
  return r(t, me(e), n, s, je);
}
function on(t, e) {
  var n = N(t) ? er : Vs;
  return n(t, Eh(me(e)));
}
function Rh(t, e) {
  var n;
  return je(t, function(r, s, i) {
    return n = e(r, s, i), !n;
  }), !!n;
}
function Ys(t, e, n) {
  var r = N(t) ? Bs : Rh;
  return r(t, me(e));
}
var Ih = 1 / 0, _h = Ve && 1 / rr(new Ve([, -0]))[1] == Ih ? function(t) {
  return new Ve(t);
} : U, Sh = 200;
function Oh(t, e, n) {
  var r = -1, s = Os, i = t.length, a = !0, o = [], c = o;
  if (i >= Sh) {
    var u = _h(t);
    if (u)
      return rr(u);
    a = !1, s = nr, c = new Ye();
  } else
    c = o;
  e:
    for (; ++r < i; ) {
      var l = t[r], h = l;
      if (l = l !== 0 ? l : 0, a && h === h) {
        for (var f = c.length; f--; )
          if (c[f] === h)
            continue e;
        o.push(l);
      } else s(c, h, n) || (c !== o && c.push(h), o.push(l));
    }
  return o;
}
function ar(t) {
  return t && t.length ? Oh(t) : [];
}
function yn(t) {
  console && console.error && console.error(`Error: ${t}`);
}
function Xs(t) {
  console && console.warn && console.warn(`Warning: ${t}`);
}
function qs(t) {
  const e = (/* @__PURE__ */ new Date()).getTime(), n = t();
  return { time: (/* @__PURE__ */ new Date()).getTime() - e, value: n };
}
function Zs(t) {
  function e() {
  }
  e.prototype = t;
  const n = new e();
  function r() {
    return typeof n.bar;
  }
  return r(), r(), t;
}
function Nh(t) {
  return Ch(t) ? t.LABEL : t.name;
}
function Ch(t) {
  return j(t.LABEL) && t.LABEL !== "";
}
class Ae {
  get definition() {
    return this._definition;
  }
  set definition(e) {
    this._definition = e;
  }
  constructor(e) {
    this._definition = e;
  }
  accept(e) {
    e.visit(this), _(this.definition, (n) => {
      n.accept(e);
    });
  }
}
class X extends Ae {
  constructor(e) {
    super([]), this.idx = 1, V(this, le(e, (n) => n !== void 0));
  }
  set definition(e) {
  }
  get definition() {
    return this.referencedRule !== void 0 ? this.referencedRule.definition : [];
  }
  accept(e) {
    e.visit(this);
  }
}
class qe extends Ae {
  constructor(e) {
    super(e.definition), this.orgText = "", V(this, le(e, (n) => n !== void 0));
  }
}
class z extends Ae {
  constructor(e) {
    super(e.definition), this.ignoreAmbiguities = !1, V(this, le(e, (n) => n !== void 0));
  }
}
class W extends Ae {
  constructor(e) {
    super(e.definition), this.idx = 1, V(this, le(e, (n) => n !== void 0));
  }
}
class Re extends Ae {
  constructor(e) {
    super(e.definition), this.idx = 1, V(this, le(e, (n) => n !== void 0));
  }
}
class Ie extends Ae {
  constructor(e) {
    super(e.definition), this.idx = 1, V(this, le(e, (n) => n !== void 0));
  }
}
class w extends Ae {
  constructor(e) {
    super(e.definition), this.idx = 1, V(this, le(e, (n) => n !== void 0));
  }
}
class he extends Ae {
  constructor(e) {
    super(e.definition), this.idx = 1, V(this, le(e, (n) => n !== void 0));
  }
}
class fe extends Ae {
  get definition() {
    return this._definition;
  }
  set definition(e) {
    this._definition = e;
  }
  constructor(e) {
    super(e.definition), this.idx = 1, this.ignoreAmbiguities = !1, this.hasPredicates = !1, V(this, le(e, (n) => n !== void 0));
  }
}
class M {
  constructor(e) {
    this.idx = 1, V(this, le(e, (n) => n !== void 0));
  }
  accept(e) {
    e.visit(this);
  }
}
function Lh(t) {
  return A(t, kt);
}
function kt(t) {
  function e(n) {
    return A(n, kt);
  }
  if (t instanceof X) {
    const n = {
      type: "NonTerminal",
      name: t.nonTerminalName,
      idx: t.idx
    };
    return j(t.label) && (n.label = t.label), n;
  } else {
    if (t instanceof z)
      return {
        type: "Alternative",
        definition: e(t.definition)
      };
    if (t instanceof W)
      return {
        type: "Option",
        idx: t.idx,
        definition: e(t.definition)
      };
    if (t instanceof Re)
      return {
        type: "RepetitionMandatory",
        idx: t.idx,
        definition: e(t.definition)
      };
    if (t instanceof Ie)
      return {
        type: "RepetitionMandatoryWithSeparator",
        idx: t.idx,
        separator: kt(new M({ terminalType: t.separator })),
        definition: e(t.definition)
      };
    if (t instanceof he)
      return {
        type: "RepetitionWithSeparator",
        idx: t.idx,
        separator: kt(new M({ terminalType: t.separator })),
        definition: e(t.definition)
      };
    if (t instanceof w)
      return {
        type: "Repetition",
        idx: t.idx,
        definition: e(t.definition)
      };
    if (t instanceof fe)
      return {
        type: "Alternation",
        idx: t.idx,
        definition: e(t.definition)
      };
    if (t instanceof M) {
      const n = {
        type: "Terminal",
        name: t.terminalType.name,
        label: Nh(t.terminalType),
        idx: t.idx
      };
      j(t.label) && (n.terminalLabel = t.label);
      const r = t.terminalType.PATTERN;
      return t.terminalType.PATTERN && (n.pattern = Se(r) ? r.source : r), n;
    } else {
      if (t instanceof qe)
        return {
          type: "Rule",
          name: t.name,
          orgText: t.orgText,
          definition: e(t.definition)
        };
      throw Error("non exhaustive match");
    }
  }
}
class Ze {
  visit(e) {
    const n = e;
    switch (n.constructor) {
      case X:
        return this.visitNonTerminal(n);
      case z:
        return this.visitAlternative(n);
      case W:
        return this.visitOption(n);
      case Re:
        return this.visitRepetitionMandatory(n);
      case Ie:
        return this.visitRepetitionMandatoryWithSeparator(n);
      case he:
        return this.visitRepetitionWithSeparator(n);
      case w:
        return this.visitRepetition(n);
      case fe:
        return this.visitAlternation(n);
      case M:
        return this.visitTerminal(n);
      case qe:
        return this.visitRule(n);
      /* c8 ignore next 2 */
      default:
        throw Error("non exhaustive match");
    }
  }
  /* c8 ignore next */
  visitNonTerminal(e) {
  }
  /* c8 ignore next */
  visitAlternative(e) {
  }
  /* c8 ignore next */
  visitOption(e) {
  }
  /* c8 ignore next */
  visitRepetition(e) {
  }
  /* c8 ignore next */
  visitRepetitionMandatory(e) {
  }
  /* c8 ignore next 3 */
  visitRepetitionMandatoryWithSeparator(e) {
  }
  /* c8 ignore next */
  visitRepetitionWithSeparator(e) {
  }
  /* c8 ignore next */
  visitAlternation(e) {
  }
  /* c8 ignore next */
  visitTerminal(e) {
  }
  /* c8 ignore next */
  visitRule(e) {
  }
}
function yh(t) {
  return t instanceof z || t instanceof W || t instanceof w || t instanceof Re || t instanceof Ie || t instanceof he || t instanceof M || t instanceof qe;
}
function Gt(t, e = []) {
  return t instanceof W || t instanceof w || t instanceof he ? !0 : t instanceof fe ? Ys(t.definition, (r) => Gt(r, e)) : t instanceof X && H(e, t) ? !1 : t instanceof Ae ? (t instanceof X && e.push(t), oe(t.definition, (r) => Gt(r, e))) : !1;
}
function vh(t) {
  return t instanceof fe;
}
function pe(t) {
  if (t instanceof X)
    return "SUBRULE";
  if (t instanceof W)
    return "OPTION";
  if (t instanceof fe)
    return "OR";
  if (t instanceof Re)
    return "AT_LEAST_ONE";
  if (t instanceof Ie)
    return "AT_LEAST_ONE_SEP";
  if (t instanceof he)
    return "MANY_SEP";
  if (t instanceof w)
    return "MANY";
  if (t instanceof M)
    return "CONSUME";
  throw Error("non exhaustive match");
}
class cn {
  walk(e, n = []) {
    _(e.definition, (r, s) => {
      const i = F(e.definition, s + 1);
      if (r instanceof X)
        this.walkProdRef(r, i, n);
      else if (r instanceof M)
        this.walkTerminal(r, i, n);
      else if (r instanceof z)
        this.walkFlat(r, i, n);
      else if (r instanceof W)
        this.walkOption(r, i, n);
      else if (r instanceof Re)
        this.walkAtLeastOne(r, i, n);
      else if (r instanceof Ie)
        this.walkAtLeastOneSep(r, i, n);
      else if (r instanceof he)
        this.walkManySep(r, i, n);
      else if (r instanceof w)
        this.walkMany(r, i, n);
      else if (r instanceof fe)
        this.walkOr(r, i, n);
      else
        throw Error("non exhaustive match");
    });
  }
  walkTerminal(e, n, r) {
  }
  walkProdRef(e, n, r) {
  }
  walkFlat(e, n, r) {
    const s = n.concat(r);
    this.walk(e, s);
  }
  walkOption(e, n, r) {
    const s = n.concat(r);
    this.walk(e, s);
  }
  walkAtLeastOne(e, n, r) {
    const s = [
      new W({ definition: e.definition })
    ].concat(n, r);
    this.walk(e, s);
  }
  walkAtLeastOneSep(e, n, r) {
    const s = qr(e, n, r);
    this.walk(e, s);
  }
  walkMany(e, n, r) {
    const s = [
      new W({ definition: e.definition })
    ].concat(n, r);
    this.walk(e, s);
  }
  walkManySep(e, n, r) {
    const s = qr(e, n, r);
    this.walk(e, s);
  }
  walkOr(e, n, r) {
    const s = n.concat(r);
    _(e.definition, (i) => {
      const a = new z({ definition: [i] });
      this.walk(a, s);
    });
  }
}
function qr(t, e, n) {
  return [
    new W({
      definition: [
        new M({ terminalType: t.separator })
      ].concat(t.definition)
    })
  ].concat(e, n);
}
function Rt(t) {
  if (t instanceof X)
    return Rt(t.referencedRule);
  if (t instanceof M)
    return kh(t);
  if (yh(t))
    return Ph(t);
  if (vh(t))
    return Mh(t);
  throw Error("non exhaustive match");
}
function Ph(t) {
  let e = [];
  const n = t.definition;
  let r = 0, s = n.length > r, i, a = !0;
  for (; s && a; )
    i = n[r], a = Gt(i), e = e.concat(Rt(i)), r = r + 1, s = n.length > r;
  return ar(e);
}
function Mh(t) {
  const e = A(t.definition, (n) => Rt(n));
  return ar(ae(e));
}
function kh(t) {
  return [t.terminalType];
}
const Qs = "_~IN~_";
class xh extends cn {
  constructor(e) {
    super(), this.topProd = e, this.follows = {};
  }
  startWalking() {
    return this.walk(this.topProd), this.follows;
  }
  walkTerminal(e, n, r) {
  }
  walkProdRef(e, n, r) {
    const s = wh(e.referencedRule, e.idx) + this.topProd.name, i = n.concat(r), a = new z({ definition: i }), o = Rt(a);
    this.follows[s] = o;
  }
}
function bh(t) {
  const e = {};
  return _(t, (n) => {
    const r = new xh(n).startWalking();
    V(e, r);
  }), e;
}
function wh(t, e) {
  return t.name + e + Qs;
}
function S(t) {
  return t.charCodeAt(0);
}
function An(t, e) {
  Array.isArray(t) ? t.forEach(function(n) {
    e.push(n);
  }) : e.push(t);
}
function st(t, e) {
  if (t[e] === !0)
    throw "duplicate flag " + e;
  t[e], t[e] = !0;
}
function ze(t) {
  if (t === void 0)
    throw Error("Internal Error - Should never get here!");
  return !0;
}
function Uh() {
  throw Error("Internal Error - Should never get here!");
}
function Zr(t) {
  return t.type === "Character";
}
const Bt = [];
for (let t = S("0"); t <= S("9"); t++)
  Bt.push(t);
const Ht = [S("_")].concat(Bt);
for (let t = S("a"); t <= S("z"); t++)
  Ht.push(t);
for (let t = S("A"); t <= S("Z"); t++)
  Ht.push(t);
const Qr = [
  S(" "),
  S("\f"),
  S(`
`),
  S("\r"),
  S("	"),
  S("\v"),
  S("	"),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S(" "),
  S("\u2028"),
  S("\u2029"),
  S(" "),
  S(" "),
  S("　"),
  S("\uFEFF")
], Fh = /[0-9a-fA-F]/, Ct = /[0-9]/, Dh = /[1-9]/;
class $h {
  constructor() {
    this.idx = 0, this.input = "", this.groupIdx = 0;
  }
  saveState() {
    return {
      idx: this.idx,
      input: this.input,
      groupIdx: this.groupIdx
    };
  }
  restoreState(e) {
    this.idx = e.idx, this.input = e.input, this.groupIdx = e.groupIdx;
  }
  pattern(e) {
    this.idx = 0, this.input = e, this.groupIdx = 0, this.consumeChar("/");
    const n = this.disjunction();
    this.consumeChar("/");
    const r = {
      type: "Flags",
      loc: { begin: this.idx, end: e.length },
      global: !1,
      ignoreCase: !1,
      multiLine: !1,
      unicode: !1,
      sticky: !1
    };
    for (; this.isRegExpFlag(); )
      switch (this.popChar()) {
        case "g":
          st(r, "global");
          break;
        case "i":
          st(r, "ignoreCase");
          break;
        case "m":
          st(r, "multiLine");
          break;
        case "u":
          st(r, "unicode");
          break;
        case "y":
          st(r, "sticky");
          break;
      }
    if (this.idx !== this.input.length)
      throw Error("Redundant input: " + this.input.substring(this.idx));
    return {
      type: "Pattern",
      flags: r,
      value: n,
      loc: this.loc(0)
    };
  }
  disjunction() {
    const e = [], n = this.idx;
    for (e.push(this.alternative()); this.peekChar() === "|"; )
      this.consumeChar("|"), e.push(this.alternative());
    return { type: "Disjunction", value: e, loc: this.loc(n) };
  }
  alternative() {
    const e = [], n = this.idx;
    for (; this.isTerm(); )
      e.push(this.term());
    return { type: "Alternative", value: e, loc: this.loc(n) };
  }
  term() {
    return this.isAssertion() ? this.assertion() : this.atom();
  }
  assertion() {
    const e = this.idx;
    switch (this.popChar()) {
      case "^":
        return {
          type: "StartAnchor",
          loc: this.loc(e)
        };
      case "$":
        return { type: "EndAnchor", loc: this.loc(e) };
      // '\b' or '\B'
      case "\\":
        switch (this.popChar()) {
          case "b":
            return {
              type: "WordBoundary",
              loc: this.loc(e)
            };
          case "B":
            return {
              type: "NonWordBoundary",
              loc: this.loc(e)
            };
        }
        throw Error("Invalid Assertion Escape");
      // '(?=' or '(?!'
      case "(":
        this.consumeChar("?");
        let n;
        switch (this.popChar()) {
          case "=":
            n = "Lookahead";
            break;
          case "!":
            n = "NegativeLookahead";
            break;
          case "<": {
            switch (this.popChar()) {
              case "=":
                n = "Lookbehind";
                break;
              case "!":
                n = "NegativeLookbehind";
            }
            break;
          }
        }
        ze(n);
        const r = this.disjunction();
        return this.consumeChar(")"), {
          type: n,
          value: r,
          loc: this.loc(e)
        };
    }
    return Uh();
  }
  quantifier(e = !1) {
    let n;
    const r = this.idx;
    switch (this.popChar()) {
      case "*":
        n = {
          atLeast: 0,
          atMost: 1 / 0
        };
        break;
      case "+":
        n = {
          atLeast: 1,
          atMost: 1 / 0
        };
        break;
      case "?":
        n = {
          atLeast: 0,
          atMost: 1
        };
        break;
      case "{":
        const s = this.integerIncludingZero();
        switch (this.popChar()) {
          case "}":
            n = {
              atLeast: s,
              atMost: s
            };
            break;
          case ",":
            let i;
            this.isDigit() ? (i = this.integerIncludingZero(), n = {
              atLeast: s,
              atMost: i
            }) : n = {
              atLeast: s,
              atMost: 1 / 0
            }, this.consumeChar("}");
            break;
        }
        if (e === !0 && n === void 0)
          return;
        ze(n);
        break;
    }
    if (!(e === !0 && n === void 0) && ze(n))
      return this.peekChar(0) === "?" ? (this.consumeChar("?"), n.greedy = !1) : n.greedy = !0, n.type = "Quantifier", n.loc = this.loc(r), n;
  }
  atom() {
    let e;
    const n = this.idx;
    switch (this.peekChar()) {
      case ".":
        e = this.dotAll();
        break;
      case "\\":
        e = this.atomEscape();
        break;
      case "[":
        e = this.characterClass();
        break;
      case "(":
        e = this.group();
        break;
    }
    if (e === void 0 && this.isPatternCharacter() && (e = this.patternCharacter()), ze(e))
      return e.loc = this.loc(n), this.isQuantifier() && (e.quantifier = this.quantifier()), e;
  }
  dotAll() {
    return this.consumeChar("."), {
      type: "Set",
      complement: !0,
      value: [S(`
`), S("\r"), S("\u2028"), S("\u2029")]
    };
  }
  atomEscape() {
    switch (this.consumeChar("\\"), this.peekChar()) {
      case "1":
      case "2":
      case "3":
      case "4":
      case "5":
      case "6":
      case "7":
      case "8":
      case "9":
        return this.decimalEscapeAtom();
      case "d":
      case "D":
      case "s":
      case "S":
      case "w":
      case "W":
        return this.characterClassEscape();
      case "f":
      case "n":
      case "r":
      case "t":
      case "v":
        return this.controlEscapeAtom();
      case "c":
        return this.controlLetterEscapeAtom();
      case "0":
        return this.nulCharacterAtom();
      case "x":
        return this.hexEscapeSequenceAtom();
      case "u":
        return this.regExpUnicodeEscapeSequenceAtom();
      default:
        return this.identityEscapeAtom();
    }
  }
  decimalEscapeAtom() {
    return { type: "GroupBackReference", value: this.positiveInteger() };
  }
  characterClassEscape() {
    let e, n = !1;
    switch (this.popChar()) {
      case "d":
        e = Bt;
        break;
      case "D":
        e = Bt, n = !0;
        break;
      case "s":
        e = Qr;
        break;
      case "S":
        e = Qr, n = !0;
        break;
      case "w":
        e = Ht;
        break;
      case "W":
        e = Ht, n = !0;
        break;
    }
    if (ze(e))
      return { type: "Set", value: e, complement: n };
  }
  controlEscapeAtom() {
    let e;
    switch (this.popChar()) {
      case "f":
        e = S("\f");
        break;
      case "n":
        e = S(`
`);
        break;
      case "r":
        e = S("\r");
        break;
      case "t":
        e = S("	");
        break;
      case "v":
        e = S("\v");
        break;
    }
    if (ze(e))
      return { type: "Character", value: e };
  }
  controlLetterEscapeAtom() {
    this.consumeChar("c");
    const e = this.popChar();
    if (/[a-zA-Z]/.test(e) === !1)
      throw Error("Invalid ");
    return { type: "Character", value: e.toUpperCase().charCodeAt(0) - 64 };
  }
  nulCharacterAtom() {
    return this.consumeChar("0"), { type: "Character", value: S("\0") };
  }
  hexEscapeSequenceAtom() {
    return this.consumeChar("x"), this.parseHexDigits(2);
  }
  regExpUnicodeEscapeSequenceAtom() {
    return this.consumeChar("u"), this.parseHexDigits(4);
  }
  identityEscapeAtom() {
    const e = this.popChar();
    return { type: "Character", value: S(e) };
  }
  classPatternCharacterAtom() {
    switch (this.peekChar()) {
      // istanbul ignore next
      case `
`:
      // istanbul ignore next
      case "\r":
      // istanbul ignore next
      case "\u2028":
      // istanbul ignore next
      case "\u2029":
      // istanbul ignore next
      case "\\":
      // istanbul ignore next
      case "]":
        throw Error("TBD");
      default:
        const e = this.popChar();
        return { type: "Character", value: S(e) };
    }
  }
  characterClass() {
    const e = [];
    let n = !1;
    for (this.consumeChar("["), this.peekChar(0) === "^" && (this.consumeChar("^"), n = !0); this.isClassAtom(); ) {
      const r = this.classAtom();
      if (r.type, Zr(r) && this.isRangeDash()) {
        this.consumeChar("-");
        const s = this.classAtom();
        if (s.type, Zr(s)) {
          if (s.value < r.value)
            throw Error("Range out of order in character class");
          e.push({ from: r.value, to: s.value });
        } else
          An(r.value, e), e.push(S("-")), An(s.value, e);
      } else
        An(r.value, e);
    }
    return this.consumeChar("]"), { type: "Set", complement: n, value: e };
  }
  classAtom() {
    switch (this.peekChar()) {
      // istanbul ignore next
      case "]":
      // istanbul ignore next
      case `
`:
      // istanbul ignore next
      case "\r":
      // istanbul ignore next
      case "\u2028":
      // istanbul ignore next
      case "\u2029":
        throw Error("TBD");
      case "\\":
        return this.classEscape();
      default:
        return this.classPatternCharacterAtom();
    }
  }
  classEscape() {
    switch (this.consumeChar("\\"), this.peekChar()) {
      // Matches a backspace.
      // (Not to be confused with \b word boundary outside characterClass)
      case "b":
        return this.consumeChar("b"), { type: "Character", value: S("\b") };
      case "d":
      case "D":
      case "s":
      case "S":
      case "w":
      case "W":
        return this.characterClassEscape();
      case "f":
      case "n":
      case "r":
      case "t":
      case "v":
        return this.controlEscapeAtom();
      case "c":
        return this.controlLetterEscapeAtom();
      case "0":
        return this.nulCharacterAtom();
      case "x":
        return this.hexEscapeSequenceAtom();
      case "u":
        return this.regExpUnicodeEscapeSequenceAtom();
      default:
        return this.identityEscapeAtom();
    }
  }
  group() {
    let e = !0;
    this.consumeChar("("), this.peekChar(0) === "?" ? (this.consumeChar("?"), this.consumeChar(":"), e = !1) : this.groupIdx++;
    const n = this.disjunction();
    this.consumeChar(")");
    const r = {
      type: "Group",
      capturing: e,
      value: n
    };
    return e && (r.idx = this.groupIdx), r;
  }
  positiveInteger() {
    let e = this.popChar();
    if (Dh.test(e) === !1)
      throw Error("Expecting a positive integer");
    for (; Ct.test(this.peekChar(0)); )
      e += this.popChar();
    return parseInt(e, 10);
  }
  integerIncludingZero() {
    let e = this.popChar();
    if (Ct.test(e) === !1)
      throw Error("Expecting an integer");
    for (; Ct.test(this.peekChar(0)); )
      e += this.popChar();
    return parseInt(e, 10);
  }
  patternCharacter() {
    const e = this.popChar();
    switch (e) {
      // istanbul ignore next
      case `
`:
      // istanbul ignore next
      case "\r":
      // istanbul ignore next
      case "\u2028":
      // istanbul ignore next
      case "\u2029":
      // istanbul ignore next
      case "^":
      // istanbul ignore next
      case "$":
      // istanbul ignore next
      case "\\":
      // istanbul ignore next
      case ".":
      // istanbul ignore next
      case "*":
      // istanbul ignore next
      case "+":
      // istanbul ignore next
      case "?":
      // istanbul ignore next
      case "(":
      // istanbul ignore next
      case ")":
      // istanbul ignore next
      case "[":
      // istanbul ignore next
      case "|":
        throw Error("TBD");
      default:
        return { type: "Character", value: S(e) };
    }
  }
  isRegExpFlag() {
    switch (this.peekChar(0)) {
      case "g":
      case "i":
      case "m":
      case "u":
      case "y":
        return !0;
      default:
        return !1;
    }
  }
  isRangeDash() {
    return this.peekChar() === "-" && this.isClassAtom(1);
  }
  isDigit() {
    return Ct.test(this.peekChar(0));
  }
  isClassAtom(e = 0) {
    switch (this.peekChar(e)) {
      case "]":
      case `
`:
      case "\r":
      case "\u2028":
      case "\u2029":
        return !1;
      default:
        return !0;
    }
  }
  isTerm() {
    return this.isAtom() || this.isAssertion();
  }
  isAtom() {
    if (this.isPatternCharacter())
      return !0;
    switch (this.peekChar(0)) {
      case ".":
      case "\\":
      // atomEscape
      case "[":
      // characterClass
      // TODO: isAtom must be called before isAssertion - disambiguate
      case "(":
        return !0;
      default:
        return !1;
    }
  }
  isAssertion() {
    switch (this.peekChar(0)) {
      case "^":
      case "$":
        return !0;
      // '\b' or '\B'
      case "\\":
        switch (this.peekChar(1)) {
          case "b":
          case "B":
            return !0;
          default:
            return !1;
        }
      // '(?=' or '(?!' or `(?<=` or `(?<!`
      case "(":
        return this.peekChar(1) === "?" && (this.peekChar(2) === "=" || this.peekChar(2) === "!" || this.peekChar(2) === "<" && (this.peekChar(3) === "=" || this.peekChar(3) === "!"));
      default:
        return !1;
    }
  }
  isQuantifier() {
    const e = this.saveState();
    try {
      return this.quantifier(!0) !== void 0;
    } catch {
      return !1;
    } finally {
      this.restoreState(e);
    }
  }
  isPatternCharacter() {
    switch (this.peekChar()) {
      case "^":
      case "$":
      case "\\":
      case ".":
      case "*":
      case "+":
      case "?":
      case "(":
      case ")":
      case "[":
      case "|":
      case "/":
      case `
`:
      case "\r":
      case "\u2028":
      case "\u2029":
        return !1;
      default:
        return !0;
    }
  }
  parseHexDigits(e) {
    let n = "";
    for (let s = 0; s < e; s++) {
      const i = this.popChar();
      if (Fh.test(i) === !1)
        throw Error("Expecting a HexDecimal digits");
      n += i;
    }
    return { type: "Character", value: parseInt(n, 16) };
  }
  peekChar(e = 0) {
    return this.input[this.idx + e];
  }
  popChar() {
    const e = this.peekChar(0);
    return this.consumeChar(void 0), e;
  }
  consumeChar(e) {
    if (e !== void 0 && this.input[this.idx] !== e)
      throw Error("Expected: '" + e + "' but found: '" + this.input[this.idx] + "' at offset: " + this.idx);
    if (this.idx >= this.input.length)
      throw Error("Unexpected end of input");
    this.idx++;
  }
  loc(e) {
    return { begin: e, end: this.idx };
  }
}
class or {
  visitChildren(e) {
    for (const n in e) {
      const r = e[n];
      e.hasOwnProperty(n) && (r.type !== void 0 ? this.visit(r) : Array.isArray(r) && r.forEach((s) => {
        this.visit(s);
      }, this));
    }
  }
  visit(e) {
    switch (e.type) {
      case "Pattern":
        this.visitPattern(e);
        break;
      case "Flags":
        this.visitFlags(e);
        break;
      case "Disjunction":
        this.visitDisjunction(e);
        break;
      case "Alternative":
        this.visitAlternative(e);
        break;
      case "StartAnchor":
        this.visitStartAnchor(e);
        break;
      case "EndAnchor":
        this.visitEndAnchor(e);
        break;
      case "WordBoundary":
        this.visitWordBoundary(e);
        break;
      case "NonWordBoundary":
        this.visitNonWordBoundary(e);
        break;
      case "Lookahead":
        this.visitLookahead(e);
        break;
      case "NegativeLookahead":
        this.visitNegativeLookahead(e);
        break;
      case "Lookbehind":
        this.visitLookbehind(e);
        break;
      case "NegativeLookbehind":
        this.visitNegativeLookbehind(e);
        break;
      case "Character":
        this.visitCharacter(e);
        break;
      case "Set":
        this.visitSet(e);
        break;
      case "Group":
        this.visitGroup(e);
        break;
      case "GroupBackReference":
        this.visitGroupBackReference(e);
        break;
      case "Quantifier":
        this.visitQuantifier(e);
        break;
    }
    this.visitChildren(e);
  }
  visitPattern(e) {
  }
  visitFlags(e) {
  }
  visitDisjunction(e) {
  }
  visitAlternative(e) {
  }
  // Assertion
  visitStartAnchor(e) {
  }
  visitEndAnchor(e) {
  }
  visitWordBoundary(e) {
  }
  visitNonWordBoundary(e) {
  }
  visitLookahead(e) {
  }
  visitNegativeLookahead(e) {
  }
  visitLookbehind(e) {
  }
  visitNegativeLookbehind(e) {
  }
  // atoms
  visitCharacter(e) {
  }
  visitSet(e) {
  }
  visitGroup(e) {
  }
  visitGroupBackReference(e) {
  }
  visitQuantifier(e) {
  }
}
let xt = {};
const Gh = new $h();
function un(t) {
  const e = t.toString();
  if (xt.hasOwnProperty(e))
    return xt[e];
  {
    const n = Gh.pattern(e);
    return xt[e] = n, n;
  }
}
function Bh() {
  xt = {};
}
const Js = "Complement Sets are not supported for first char optimization", Kt = `Unable to use "first char" lexer optimizations:
`;
function Hh(t, e = !1) {
  try {
    const n = un(t);
    return vn(n.value, {}, n.flags.ignoreCase);
  } catch (n) {
    if (n.message === Js)
      e && Xs(`${Kt}	Unable to optimize: < ${t.toString()} >
	Complement Sets cannot be automatically optimized.
	This will disable the lexer's first char optimizations.
	See: https://chevrotain.io/docs/guide/resolving_lexer_errors.html#COMPLEMENT for details.`);
    else {
      let r = "";
      e && (r = `
	This will disable the lexer's first char optimizations.
	See: https://chevrotain.io/docs/guide/resolving_lexer_errors.html#REGEXP_PARSING for details.`), yn(`${Kt}
	Failed parsing: < ${t.toString()} >
	Using the @chevrotain/regexp-to-ast library
	Please open an issue at: https://github.com/chevrotain/chevrotain/issues` + r);
    }
  }
  return [];
}
function vn(t, e, n) {
  switch (t.type) {
    case "Disjunction":
      for (let s = 0; s < t.value.length; s++)
        vn(t.value[s], e, n);
      break;
    case "Alternative":
      const r = t.value;
      for (let s = 0; s < r.length; s++) {
        const i = r[s];
        switch (i.type) {
          case "EndAnchor":
          // A group back reference cannot affect potential starting char.
          // because if a back reference is the first production than automatically
          // the group being referenced has had to come BEFORE so its codes have already been added
          case "GroupBackReference":
          // assertions do not affect potential starting codes
          case "Lookahead":
          case "NegativeLookahead":
          case "Lookbehind":
          case "NegativeLookbehind":
          case "StartAnchor":
          case "WordBoundary":
          case "NonWordBoundary":
            continue;
        }
        const a = i;
        switch (a.type) {
          case "Character":
            Lt(a.value, e, n);
            break;
          case "Set":
            if (a.complement === !0)
              throw Error(Js);
            _(a.value, (c) => {
              if (typeof c == "number")
                Lt(c, e, n);
              else {
                const u = c;
                if (n === !0)
                  for (let l = u.from; l <= u.to; l++)
                    Lt(l, e, n);
                else {
                  for (let l = u.from; l <= u.to && l < ot; l++)
                    Lt(l, e, n);
                  if (u.to >= ot) {
                    const l = u.from >= ot ? u.from : ot, h = u.to, f = xe(l), p = xe(h);
                    for (let m = f; m <= p; m++)
                      e[m] = m;
                  }
                }
              }
            });
            break;
          case "Group":
            vn(a.value, e, n);
            break;
          /* istanbul ignore next */
          default:
            throw Error("Non Exhaustive Match");
        }
        const o = a.quantifier !== void 0 && a.quantifier.atLeast === 0;
        if (
          // A group may be optional due to empty contents /(?:)/
          // or if everything inside it is optional /((a)?)/
          a.type === "Group" && Pn(a) === !1 || // If this term is not a group it may only be optional if it has an optional quantifier
          a.type !== "Group" && o === !1
        )
          break;
      }
      break;
    /* istanbul ignore next */
    default:
      throw Error("non exhaustive match!");
  }
  return b(e);
}
function Lt(t, e, n) {
  const r = xe(t);
  e[r] = r, n === !0 && Kh(t, e);
}
function Kh(t, e) {
  const n = String.fromCharCode(t), r = n.toUpperCase();
  if (r !== n) {
    const s = xe(r.charCodeAt(0));
    e[s] = s;
  } else {
    const s = n.toLowerCase();
    if (s !== n) {
      const i = xe(s.charCodeAt(0));
      e[i] = i;
    }
  }
}
function Jr(t, e) {
  return Xe(t.value, (n) => {
    if (typeof n == "number")
      return H(e, n);
    {
      const r = n;
      return Xe(e, (s) => r.from <= s && s <= r.to) !== void 0;
    }
  });
}
function Pn(t) {
  const e = t.quantifier;
  return e && e.atLeast === 0 ? !0 : t.value ? N(t.value) ? oe(t.value, Pn) : Pn(t.value) : !1;
}
class Wh extends or {
  constructor(e) {
    super(), this.targetCharCodes = e, this.found = !1;
  }
  visitChildren(e) {
    if (this.found !== !0) {
      switch (e.type) {
        case "Lookahead":
          this.visitLookahead(e);
          return;
        case "NegativeLookahead":
          this.visitNegativeLookahead(e);
          return;
        case "Lookbehind":
          this.visitLookbehind(e);
          return;
        case "NegativeLookbehind":
          this.visitNegativeLookbehind(e);
          return;
      }
      super.visitChildren(e);
    }
  }
  visitCharacter(e) {
    H(this.targetCharCodes, e.value) && (this.found = !0);
  }
  visitSet(e) {
    e.complement ? Jr(e, this.targetCharCodes) === void 0 && (this.found = !0) : Jr(e, this.targetCharCodes) !== void 0 && (this.found = !0);
  }
}
function cr(t, e) {
  if (e instanceof RegExp) {
    const n = un(e), r = new Wh(t);
    return r.visit(n), r.found;
  } else
    return Xe(e, (n) => H(t, n.charCodeAt(0))) !== void 0;
}
const Be = "PATTERN", at = "defaultMode", yt = "modes";
function jh(t, e) {
  e = ir(e, {
    debug: !1,
    safeMode: !1,
    positionTracking: "full",
    lineTerminatorCharacters: ["\r", `
`],
    tracer: (g, d) => d()
  });
  const n = e.tracer;
  n("initCharCodeToOptimizedIndexMap", () => {
    Ef();
  });
  let r;
  n("Reject Lexer.NA", () => {
    r = on(t, (g) => g[Be] === G.NA);
  });
  let s = !1, i;
  n("Transform Patterns", () => {
    s = !1, i = A(r, (g) => {
      const d = g[Be];
      if (Se(d)) {
        const E = d.source;
        return E.length === 1 && // only these regExp meta characters which can appear in a length one regExp
        E !== "^" && E !== "$" && E !== "." && !d.ignoreCase ? E : E.length === 2 && E[0] === "\\" && // not a meta character
        !H([
          "d",
          "D",
          "s",
          "S",
          "t",
          "r",
          "n",
          "t",
          "0",
          "c",
          "b",
          "B",
          "f",
          "v",
          "w",
          "W"
        ], E[1]) ? E[1] : es(d);
      } else {
        if (Ce(d))
          return s = !0, { exec: d };
        if (typeof d == "object")
          return s = !0, d;
        if (typeof d == "string") {
          if (d.length === 1)
            return d;
          {
            const E = d.replace(/[\\^$.*+?()[\]{}|]/g, "\\$&"), L = new RegExp(E);
            return es(L);
          }
        } else
          throw Error("non exhaustive match");
      }
    });
  });
  let a, o, c, u, l;
  n("misc mapping", () => {
    a = A(r, (g) => g.tokenTypeIdx), o = A(r, (g) => {
      const d = g.GROUP;
      if (d !== G.SKIPPED) {
        if (j(d))
          return d;
        if (Oe(d))
          return !1;
        throw Error("non exhaustive match");
      }
    }), c = A(r, (g) => {
      const d = g.LONGER_ALT;
      if (d)
        return N(d) ? A(d, (L) => Yr(r, L)) : [Yr(r, d)];
    }), u = A(r, (g) => g.PUSH_MODE), l = A(r, (g) => R(g, "POP_MODE"));
  });
  let h;
  n("Line Terminator Handling", () => {
    const g = ni(e.lineTerminatorCharacters);
    h = A(r, (d) => !1), e.positionTracking !== "onlyOffset" && (h = A(r, (d) => R(d, "LINE_BREAKS") ? !!d.LINE_BREAKS : ti(d, g) === !1 && cr(g, d.PATTERN)));
  });
  let f, p, m, I;
  n("Misc Mapping #2", () => {
    f = A(r, ei), p = A(i, df), m = Y(r, (g, d) => {
      const E = d.GROUP;
      return j(E) && E !== G.SKIPPED && (g[E] = []), g;
    }, {}), I = A(i, (g, d) => ({
      pattern: i[d],
      longerAlt: c[d],
      canLineTerminator: h[d],
      isCustom: f[d],
      short: p[d],
      group: o[d],
      push: u[d],
      pop: l[d],
      tokenTypeIdx: a[d],
      tokenType: r[d]
    }));
  });
  let O = !0, T = [];
  return e.safeMode || n("First Char Optimization", () => {
    T = Y(r, (g, d, E) => {
      if (typeof d.PATTERN == "string") {
        const L = d.PATTERN.charCodeAt(0), se = xe(L);
        Rn(g, se, I[E]);
      } else if (N(d.START_CHARS_HINT)) {
        let L;
        _(d.START_CHARS_HINT, (se) => {
          const ve = typeof se == "string" ? se.charCodeAt(0) : se, Ue = xe(ve);
          L !== Ue && (L = Ue, Rn(g, Ue, I[E]));
        });
      } else if (Se(d.PATTERN))
        if (d.PATTERN.unicode)
          O = !1, e.ensureOptimizations && yn(`${Kt}	Unable to analyze < ${d.PATTERN.toString()} > pattern.
	The regexp unicode flag is not currently supported by the regexp-to-ast library.
	This will disable the lexer's first char optimizations.
	For details See: https://chevrotain.io/docs/guide/resolving_lexer_errors.html#UNICODE_OPTIMIZE`);
        else {
          const L = Hh(d.PATTERN, e.ensureOptimizations);
          v(L) && (O = !1), _(L, (se) => {
            Rn(g, se, I[E]);
          });
        }
      else
        e.ensureOptimizations && yn(`${Kt}	TokenType: <${d.name}> is using a custom token pattern without providing <start_chars_hint> parameter.
	This will disable the lexer's first char optimizations.
	For details See: https://chevrotain.io/docs/guide/resolving_lexer_errors.html#CUSTOM_OPTIMIZE`), O = !1;
      return g;
    }, []);
  }), {
    emptyGroups: m,
    patternIdxToConfig: I,
    charCodeToPatternIdxToConfig: T,
    hasCustom: s,
    canBeOptimized: O
  };
}
function zh(t, e) {
  let n = [];
  const r = Yh(t);
  n = n.concat(r.errors);
  const s = Xh(r.valid), i = s.valid;
  return n = n.concat(s.errors), n = n.concat(Vh(i)), n = n.concat(rf(i)), n = n.concat(sf(i, e)), n = n.concat(af(i)), n;
}
function Vh(t) {
  let e = [];
  const n = re(t, (r) => Se(r[Be]));
  return e = e.concat(Zh(n)), e = e.concat(ef(n)), e = e.concat(tf(n)), e = e.concat(nf(n)), e = e.concat(Qh(n)), e;
}
function Yh(t) {
  const e = re(t, (s) => !R(s, Be)), n = A(e, (s) => ({
    message: "Token Type: ->" + s.name + "<- missing static 'PATTERN' property",
    type: x.MISSING_PATTERN,
    tokenTypes: [s]
  })), r = an(t, e);
  return { errors: n, valid: r };
}
function Xh(t) {
  const e = re(t, (s) => {
    const i = s[Be];
    return !Se(i) && !Ce(i) && !R(i, "exec") && !j(i);
  }), n = A(e, (s) => ({
    message: "Token Type: ->" + s.name + "<- static 'PATTERN' can only be a RegExp, a Function matching the {CustomPatternMatcherFunc} type or an Object matching the {ICustomPattern} interface.",
    type: x.INVALID_PATTERN,
    tokenTypes: [s]
  })), r = an(t, e);
  return { errors: n, valid: r };
}
const qh = /[^\\][$]/;
function Zh(t) {
  class e extends or {
    constructor() {
      super(...arguments), this.found = !1;
    }
    visitEndAnchor(i) {
      this.found = !0;
    }
  }
  const n = re(t, (s) => {
    const i = s.PATTERN;
    try {
      const a = un(i), o = new e();
      return o.visit(a), o.found;
    } catch {
      return qh.test(i.source);
    }
  });
  return A(n, (s) => ({
    message: `Unexpected RegExp Anchor Error:
	Token Type: ->` + s.name + `<- static 'PATTERN' cannot contain end of input anchor '$'
	See chevrotain.io/docs/guide/resolving_lexer_errors.html#ANCHORS	for details.`,
    type: x.EOI_ANCHOR_FOUND,
    tokenTypes: [s]
  }));
}
function Qh(t) {
  const e = re(t, (r) => r.PATTERN.test(""));
  return A(e, (r) => ({
    message: "Token Type: ->" + r.name + "<- static 'PATTERN' must not match an empty string",
    type: x.EMPTY_MATCH_PATTERN,
    tokenTypes: [r]
  }));
}
const Jh = /[^\\[][\^]|^\^/;
function ef(t) {
  class e extends or {
    constructor() {
      super(...arguments), this.found = !1;
    }
    visitStartAnchor(i) {
      this.found = !0;
    }
  }
  const n = re(t, (s) => {
    const i = s.PATTERN;
    try {
      const a = un(i), o = new e();
      return o.visit(a), o.found;
    } catch {
      return Jh.test(i.source);
    }
  });
  return A(n, (s) => ({
    message: `Unexpected RegExp Anchor Error:
	Token Type: ->` + s.name + `<- static 'PATTERN' cannot contain start of input anchor '^'
	See https://chevrotain.io/docs/guide/resolving_lexer_errors.html#ANCHORS	for details.`,
    type: x.SOI_ANCHOR_FOUND,
    tokenTypes: [s]
  }));
}
function tf(t) {
  const e = re(t, (r) => {
    const s = r[Be];
    return s instanceof RegExp && (s.multiline || s.global);
  });
  return A(e, (r) => ({
    message: "Token Type: ->" + r.name + "<- static 'PATTERN' may NOT contain global('g') or multiline('m')",
    type: x.UNSUPPORTED_FLAGS_FOUND,
    tokenTypes: [r]
  }));
}
function nf(t) {
  const e = [];
  let n = A(t, (i) => Y(t, (a, o) => (i.PATTERN.source === o.PATTERN.source && !H(e, o) && o.PATTERN !== G.NA && (e.push(o), a.push(o)), a), []));
  n = At(n);
  const r = re(n, (i) => i.length > 1);
  return A(r, (i) => {
    const a = A(i, (c) => c.name);
    return {
      message: `The same RegExp pattern ->${ue(i).PATTERN}<-has been used in all of the following Token Types: ${a.join(", ")} <-`,
      type: x.DUPLICATE_PATTERNS_FOUND,
      tokenTypes: i
    };
  });
}
function rf(t) {
  const e = re(t, (r) => {
    if (!R(r, "GROUP"))
      return !1;
    const s = r.GROUP;
    return s !== G.SKIPPED && s !== G.NA && !j(s);
  });
  return A(e, (r) => ({
    message: "Token Type: ->" + r.name + "<- static 'GROUP' can only be Lexer.SKIPPED/Lexer.NA/A String",
    type: x.INVALID_GROUP_TYPE_FOUND,
    tokenTypes: [r]
  }));
}
function sf(t, e) {
  const n = re(t, (s) => s.PUSH_MODE !== void 0 && !H(e, s.PUSH_MODE));
  return A(n, (s) => ({
    message: `Token Type: ->${s.name}<- static 'PUSH_MODE' value cannot refer to a Lexer Mode ->${s.PUSH_MODE}<-which does not exist`,
    type: x.PUSH_MODE_DOES_NOT_EXIST,
    tokenTypes: [s]
  }));
}
function af(t) {
  const e = [], n = Y(t, (r, s, i) => {
    const a = s.PATTERN;
    return a === G.NA || (j(a) ? r.push({ str: a, idx: i, tokenType: s }) : Se(a) && cf(a) && r.push({ str: a.source, idx: i, tokenType: s })), r;
  }, []);
  return _(t, (r, s) => {
    _(n, ({ str: i, idx: a, tokenType: o }) => {
      if (s < a && of(i, r.PATTERN)) {
        const c = `Token: ->${o.name}<- can never be matched.
Because it appears AFTER the Token Type ->${r.name}<-in the lexer's definition.
See https://chevrotain.io/docs/guide/resolving_lexer_errors.html#UNREACHABLE`;
        e.push({
          message: c,
          type: x.UNREACHABLE_PATTERN,
          tokenTypes: [r, o]
        });
      }
    });
  }), e;
}
function of(t, e) {
  if (Se(e)) {
    if (uf(e))
      return !1;
    const n = e.exec(t);
    return n !== null && n.index === 0;
  } else {
    if (Ce(e))
      return e(t, 0, [], {});
    if (R(e, "exec"))
      return e.exec(t, 0, [], {});
    if (typeof e == "string")
      return e === t;
    throw Error("non exhaustive match");
  }
}
function cf(t) {
  return Xe([
    ".",
    "\\",
    "[",
    "]",
    "|",
    "^",
    "$",
    "(",
    ")",
    "?",
    "*",
    "+",
    "{"
  ], (n) => t.source.indexOf(n) !== -1) === void 0;
}
function uf(t) {
  return /(\(\?=)|(\(\?!)|(\(\?<=)|(\(\?<!)/.test(t.source);
}
function es(t) {
  const e = t.ignoreCase ? "iy" : "y";
  return new RegExp(`${t.source}`, e);
}
function lf(t, e, n) {
  const r = [];
  return R(t, at) || r.push({
    message: "A MultiMode Lexer cannot be initialized without a <" + at + `> property in its definition
`,
    type: x.MULTI_MODE_LEXER_WITHOUT_DEFAULT_MODE
  }), R(t, yt) || r.push({
    message: "A MultiMode Lexer cannot be initialized without a <" + yt + `> property in its definition
`,
    type: x.MULTI_MODE_LEXER_WITHOUT_MODES_PROPERTY
  }), R(t, yt) && R(t, at) && !R(t.modes, t.defaultMode) && r.push({
    message: `A MultiMode Lexer cannot be initialized with a ${at}: <${t.defaultMode}>which does not exist
`,
    type: x.MULTI_MODE_LEXER_DEFAULT_MODE_VALUE_DOES_NOT_EXIST
  }), R(t, yt) && _(t.modes, (s, i) => {
    _(s, (a, o) => {
      if (Oe(a))
        r.push({
          message: `A Lexer cannot be initialized using an undefined Token Type. Mode:<${i}> at index: <${o}>
`,
          type: x.LEXER_DEFINITION_CANNOT_CONTAIN_UNDEFINED
        });
      else if (R(a, "LONGER_ALT")) {
        const c = N(a.LONGER_ALT) ? a.LONGER_ALT : [a.LONGER_ALT];
        _(c, (u) => {
          !Oe(u) && !H(s, u) && r.push({
            message: `A MultiMode Lexer cannot be initialized with a longer_alt <${u.name}> on token <${a.name}> outside of mode <${i}>
`,
            type: x.MULTI_MODE_LEXER_LONGER_ALT_NOT_IN_CURRENT_MODE
          });
        });
      }
    });
  }), r;
}
function hf(t, e, n) {
  const r = [];
  let s = !1;
  const i = At(ae(b(t.modes))), a = on(i, (c) => c[Be] === G.NA), o = ni(n);
  return e && _(a, (c) => {
    const u = ti(c, o);
    if (u !== !1) {
      const h = {
        message: gf(c, u),
        type: u.issue,
        tokenType: c
      };
      r.push(h);
    } else
      R(c, "LINE_BREAKS") ? c.LINE_BREAKS === !0 && (s = !0) : cr(o, c.PATTERN) && (s = !0);
  }), e && !s && r.push({
    message: `Warning: No LINE_BREAKS Found.
	This Lexer has been defined to track line and column information,
	But none of the Token Types can be identified as matching a line terminator.
	See https://chevrotain.io/docs/guide/resolving_lexer_errors.html#LINE_BREAKS 
	for details.`,
    type: x.NO_LINE_BREAKS_FLAGS
  }), r;
}
function ff(t) {
  const e = {}, n = ne(t);
  return _(n, (r) => {
    const s = t[r];
    if (N(s))
      e[r] = [];
    else
      throw Error("non exhaustive match");
  }), e;
}
function ei(t) {
  const e = t.PATTERN;
  if (Se(e))
    return !1;
  if (Ce(e))
    return !0;
  if (R(e, "exec"))
    return !0;
  if (j(e))
    return !1;
  throw Error("non exhaustive match");
}
function df(t) {
  return j(t) && t.length === 1 ? t.charCodeAt(0) : !1;
}
const pf = {
  // implements /\n|\r\n?/g.test
  test: function(t) {
    const e = t.length;
    for (let n = this.lastIndex; n < e; n++) {
      const r = t.charCodeAt(n);
      if (r === 10)
        return this.lastIndex = n + 1, !0;
      if (r === 13)
        return t.charCodeAt(n + 1) === 10 ? this.lastIndex = n + 2 : this.lastIndex = n + 1, !0;
    }
    return !1;
  },
  lastIndex: 0
};
function ti(t, e) {
  if (R(t, "LINE_BREAKS"))
    return !1;
  if (Se(t.PATTERN)) {
    try {
      cr(e, t.PATTERN);
    } catch (n) {
      return {
        issue: x.IDENTIFY_TERMINATOR,
        errMsg: n.message
      };
    }
    return !1;
  } else {
    if (j(t.PATTERN))
      return !1;
    if (ei(t))
      return { issue: x.CUSTOM_LINE_BREAK };
    throw Error("non exhaustive match");
  }
}
function gf(t, e) {
  if (e.issue === x.IDENTIFY_TERMINATOR)
    return `Warning: unable to identify line terminator usage in pattern.
	The problem is in the <${t.name}> Token Type
	 Root cause: ${e.errMsg}.
	For details See: https://chevrotain.io/docs/guide/resolving_lexer_errors.html#IDENTIFY_TERMINATOR`;
  if (e.issue === x.CUSTOM_LINE_BREAK)
    return `Warning: A Custom Token Pattern should specify the <line_breaks> option.
	The problem is in the <${t.name}> Token Type
	For details See: https://chevrotain.io/docs/guide/resolving_lexer_errors.html#CUSTOM_LINE_BREAK`;
  throw Error("non exhaustive match");
}
function ni(t) {
  return A(t, (n) => j(n) ? n.charCodeAt(0) : n);
}
function Rn(t, e, n) {
  t[e] === void 0 ? t[e] = [n] : t[e].push(n);
}
const ot = 256;
let bt = [];
function xe(t) {
  return t < ot ? t : bt[t];
}
function Ef() {
  if (v(bt)) {
    bt = new Array(65536);
    for (let t = 0; t < 65536; t++)
      bt[t] = t > 255 ? 255 + ~~(t / 255) : t;
  }
}
function It(t, e) {
  const n = t.tokenTypeIdx;
  return n === e.tokenTypeIdx ? !0 : e.isParent === !0 && e.categoryMatchesMap[n] === !0;
}
function Wt(t, e) {
  return t.tokenTypeIdx === e.tokenTypeIdx;
}
let ts = 1;
const ri = {};
function _t(t) {
  const e = Tf(t);
  mf(e), Rf(e), Af(e), _(e, (n) => {
    n.isParent = n.categoryMatches.length > 0;
  });
}
function Tf(t) {
  let e = D(t), n = t, r = !0;
  for (; r; ) {
    n = At(ae(A(n, (i) => i.CATEGORIES)));
    const s = an(n, e);
    e = e.concat(s), v(s) ? r = !1 : n = s;
  }
  return e;
}
function mf(t) {
  _(t, (e) => {
    ii(e) || (ri[ts] = e, e.tokenTypeIdx = ts++), ns(e) && !N(e.CATEGORIES) && (e.CATEGORIES = [e.CATEGORIES]), ns(e) || (e.CATEGORIES = []), If(e) || (e.categoryMatches = []), _f(e) || (e.categoryMatchesMap = {});
  });
}
function Af(t) {
  _(t, (e) => {
    e.categoryMatches = [], _(e.categoryMatchesMap, (n, r) => {
      e.categoryMatches.push(ri[r].tokenTypeIdx);
    });
  });
}
function Rf(t) {
  _(t, (e) => {
    si([], e);
  });
}
function si(t, e) {
  _(t, (n) => {
    e.categoryMatchesMap[n.tokenTypeIdx] = !0;
  }), _(e.CATEGORIES, (n) => {
    const r = t.concat(e);
    H(r, n) || si(r, n);
  });
}
function ii(t) {
  return R(t, "tokenTypeIdx");
}
function ns(t) {
  return R(t, "CATEGORIES");
}
function If(t) {
  return R(t, "categoryMatches");
}
function _f(t) {
  return R(t, "categoryMatchesMap");
}
function Sf(t) {
  return R(t, "tokenTypeIdx");
}
const Of = {
  buildUnableToPopLexerModeMessage(t) {
    return `Unable to pop Lexer Mode after encountering Token ->${t.image}<- The Mode Stack is empty`;
  },
  buildUnexpectedCharactersMessage(t, e, n, r, s, i) {
    return `unexpected character: ->${t.charAt(e)}<- at offset: ${e}, skipped ${n} characters.`;
  }
};
var x;
(function(t) {
  t[t.MISSING_PATTERN = 0] = "MISSING_PATTERN", t[t.INVALID_PATTERN = 1] = "INVALID_PATTERN", t[t.EOI_ANCHOR_FOUND = 2] = "EOI_ANCHOR_FOUND", t[t.UNSUPPORTED_FLAGS_FOUND = 3] = "UNSUPPORTED_FLAGS_FOUND", t[t.DUPLICATE_PATTERNS_FOUND = 4] = "DUPLICATE_PATTERNS_FOUND", t[t.INVALID_GROUP_TYPE_FOUND = 5] = "INVALID_GROUP_TYPE_FOUND", t[t.PUSH_MODE_DOES_NOT_EXIST = 6] = "PUSH_MODE_DOES_NOT_EXIST", t[t.MULTI_MODE_LEXER_WITHOUT_DEFAULT_MODE = 7] = "MULTI_MODE_LEXER_WITHOUT_DEFAULT_MODE", t[t.MULTI_MODE_LEXER_WITHOUT_MODES_PROPERTY = 8] = "MULTI_MODE_LEXER_WITHOUT_MODES_PROPERTY", t[t.MULTI_MODE_LEXER_DEFAULT_MODE_VALUE_DOES_NOT_EXIST = 9] = "MULTI_MODE_LEXER_DEFAULT_MODE_VALUE_DOES_NOT_EXIST", t[t.LEXER_DEFINITION_CANNOT_CONTAIN_UNDEFINED = 10] = "LEXER_DEFINITION_CANNOT_CONTAIN_UNDEFINED", t[t.SOI_ANCHOR_FOUND = 11] = "SOI_ANCHOR_FOUND", t[t.EMPTY_MATCH_PATTERN = 12] = "EMPTY_MATCH_PATTERN", t[t.NO_LINE_BREAKS_FLAGS = 13] = "NO_LINE_BREAKS_FLAGS", t[t.UNREACHABLE_PATTERN = 14] = "UNREACHABLE_PATTERN", t[t.IDENTIFY_TERMINATOR = 15] = "IDENTIFY_TERMINATOR", t[t.CUSTOM_LINE_BREAK = 16] = "CUSTOM_LINE_BREAK", t[t.MULTI_MODE_LEXER_LONGER_ALT_NOT_IN_CURRENT_MODE = 17] = "MULTI_MODE_LEXER_LONGER_ALT_NOT_IN_CURRENT_MODE";
})(x || (x = {}));
const ct = {
  deferDefinitionErrorsHandling: !1,
  positionTracking: "full",
  lineTerminatorsPattern: /\n|\r\n?/g,
  lineTerminatorCharacters: [`
`, "\r"],
  ensureOptimizations: !1,
  safeMode: !1,
  errorMessageProvider: Of,
  traceInitPerf: !1,
  skipValidations: !1,
  recoveryEnabled: !0
};
Object.freeze(ct);
class G {
  constructor(e, n = ct) {
    if (this.lexerDefinition = e, this.lexerDefinitionErrors = [], this.lexerDefinitionWarning = [], this.patternIdxToConfig = {}, this.charCodeToPatternIdxToConfig = {}, this.modes = [], this.emptyGroups = {}, this.trackStartLines = !0, this.trackEndLines = !0, this.hasCustom = !1, this.canModeBeOptimized = {}, this.TRACE_INIT = (s, i) => {
      if (this.traceInitPerf === !0) {
        this.traceInitIndent++;
        const a = new Array(this.traceInitIndent + 1).join("	");
        this.traceInitIndent < this.traceInitMaxIdent && console.log(`${a}--> <${s}>`);
        const { time: o, value: c } = qs(i), u = o > 10 ? console.warn : console.log;
        return this.traceInitIndent < this.traceInitMaxIdent && u(`${a}<-- <${s}> time: ${o}ms`), this.traceInitIndent--, c;
      } else
        return i();
    }, typeof n == "boolean")
      throw Error(`The second argument to the Lexer constructor is now an ILexerConfig Object.
a boolean 2nd argument is no longer supported`);
    this.config = V({}, ct, n);
    const r = this.config.traceInitPerf;
    r === !0 ? (this.traceInitMaxIdent = 1 / 0, this.traceInitPerf = !0) : typeof r == "number" && (this.traceInitMaxIdent = r, this.traceInitPerf = !0), this.traceInitIndent = -1, this.TRACE_INIT("Lexer Constructor", () => {
      let s, i = !0;
      this.TRACE_INIT("Lexer Config handling", () => {
        if (this.config.lineTerminatorsPattern === ct.lineTerminatorsPattern)
          this.config.lineTerminatorsPattern = pf;
        else if (this.config.lineTerminatorCharacters === ct.lineTerminatorCharacters)
          throw Error(`Error: Missing <lineTerminatorCharacters> property on the Lexer config.
	For details See: https://chevrotain.io/docs/guide/resolving_lexer_errors.html#MISSING_LINE_TERM_CHARS`);
        if (n.safeMode && n.ensureOptimizations)
          throw Error('"safeMode" and "ensureOptimizations" flags are mutually exclusive.');
        this.trackStartLines = /full|onlyStart/i.test(this.config.positionTracking), this.trackEndLines = /full/i.test(this.config.positionTracking), N(e) ? s = {
          modes: { defaultMode: D(e) },
          defaultMode: at
        } : (i = !1, s = D(e));
      }), this.config.skipValidations === !1 && (this.TRACE_INIT("performRuntimeChecks", () => {
        this.lexerDefinitionErrors = this.lexerDefinitionErrors.concat(lf(s, this.trackStartLines, this.config.lineTerminatorCharacters));
      }), this.TRACE_INIT("performWarningRuntimeChecks", () => {
        this.lexerDefinitionWarning = this.lexerDefinitionWarning.concat(hf(s, this.trackStartLines, this.config.lineTerminatorCharacters));
      })), s.modes = s.modes ? s.modes : {}, _(s.modes, (o, c) => {
        s.modes[c] = on(o, (u) => Oe(u));
      });
      const a = ne(s.modes);
      if (_(s.modes, (o, c) => {
        this.TRACE_INIT(`Mode: <${c}> processing`, () => {
          if (this.modes.push(c), this.config.skipValidations === !1 && this.TRACE_INIT("validatePatterns", () => {
            this.lexerDefinitionErrors = this.lexerDefinitionErrors.concat(zh(o, a));
          }), v(this.lexerDefinitionErrors)) {
            _t(o);
            let u;
            this.TRACE_INIT("analyzeTokenTypes", () => {
              u = jh(o, {
                lineTerminatorCharacters: this.config.lineTerminatorCharacters,
                positionTracking: n.positionTracking,
                ensureOptimizations: n.ensureOptimizations,
                safeMode: n.safeMode,
                tracer: this.TRACE_INIT
              });
            }), this.patternIdxToConfig[c] = u.patternIdxToConfig, this.charCodeToPatternIdxToConfig[c] = u.charCodeToPatternIdxToConfig, this.emptyGroups = V({}, this.emptyGroups, u.emptyGroups), this.hasCustom = u.hasCustom || this.hasCustom, this.canModeBeOptimized[c] = u.canBeOptimized;
          }
        });
      }), this.defaultMode = s.defaultMode, !v(this.lexerDefinitionErrors) && !this.config.deferDefinitionErrorsHandling) {
        const c = A(this.lexerDefinitionErrors, (u) => u.message).join(`-----------------------
`);
        throw new Error(`Errors detected in definition of Lexer:
` + c);
      }
      _(this.lexerDefinitionWarning, (o) => {
        Xs(o.message);
      }), this.TRACE_INIT("Choosing sub-methods implementations", () => {
        if (i && (this.handleModes = U), this.trackStartLines === !1 && (this.computeNewColumn = gt), this.trackEndLines === !1 && (this.updateTokenEndLineColumnLocation = U), /full/i.test(this.config.positionTracking))
          this.createTokenInstance = this.createFullToken;
        else if (/onlyStart/i.test(this.config.positionTracking))
          this.createTokenInstance = this.createStartOnlyToken;
        else if (/onlyOffset/i.test(this.config.positionTracking))
          this.createTokenInstance = this.createOffsetOnlyToken;
        else
          throw Error(`Invalid <positionTracking> config option: "${this.config.positionTracking}"`);
        this.hasCustom ? (this.addToken = this.addTokenUsingPush, this.handlePayload = this.handlePayloadWithCustom) : (this.addToken = this.addTokenUsingMemberAccess, this.handlePayload = this.handlePayloadNoCustom);
      }), this.TRACE_INIT("Failed Optimization Warnings", () => {
        const o = Y(this.canModeBeOptimized, (c, u, l) => (u === !1 && c.push(l), c), []);
        if (n.ensureOptimizations && !v(o))
          throw Error(`Lexer Modes: < ${o.join(", ")} > cannot be optimized.
	 Disable the "ensureOptimizations" lexer config flag to silently ignore this and run the lexer in an un-optimized mode.
	 Or inspect the console log for details on how to resolve these issues.`);
      }), this.TRACE_INIT("clearRegExpParserCache", () => {
        Bh();
      }), this.TRACE_INIT("toFastProperties", () => {
        Zs(this);
      });
    });
  }
  tokenize(e, n = this.defaultMode) {
    if (!v(this.lexerDefinitionErrors)) {
      const s = A(this.lexerDefinitionErrors, (i) => i.message).join(`-----------------------
`);
      throw new Error(`Unable to Tokenize because Errors detected in definition of Lexer:
` + s);
    }
    return this.tokenizeInternal(e, n);
  }
  // There is quite a bit of duplication between this and "tokenizeInternalLazy"
  // This is intentional due to performance considerations.
  // this method also used quite a bit of `!` none null assertions because it is too optimized
  // for `tsc` to always understand it is "safe"
  tokenizeInternal(e, n) {
    let r, s, i, a, o, c, u, l, h, f, p, m, I, O, T;
    const g = e, d = g.length;
    let E = 0, L = 0;
    const se = this.hasCustom ? 0 : Math.floor(e.length / 10), ve = new Array(se), Ue = [];
    let Qe = this.trackStartLines ? 1 : void 0, de = this.trackStartLines ? 1 : void 0;
    const Je = ff(this.emptyGroups), zi = this.trackStartLines, dn = this.config.lineTerminatorsPattern;
    let Ot = 0, Pe = [], et = [];
    const tt = [], dr = [];
    Object.freeze(dr);
    let nt = !1;
    const Vi = (K) => {
      if (tt.length === 1 && // if we have both a POP_MODE and a PUSH_MODE this is in-fact a "transition"
      // So no error should occur.
      K.tokenType.PUSH_MODE === void 0) {
        const q = this.config.errorMessageProvider.buildUnableToPopLexerModeMessage(K);
        Ue.push({
          offset: K.startOffset,
          line: K.startLine,
          column: K.startColumn,
          length: K.image.length,
          message: q
        });
      } else {
        tt.pop();
        const q = Ge(tt);
        Pe = this.patternIdxToConfig[q], et = this.charCodeToPatternIdxToConfig[q], Ot = Pe.length;
        const pn = this.canModeBeOptimized[q] && this.config.safeMode === !1;
        et && pn ? nt = !0 : nt = !1;
      }
    };
    function pr(K) {
      tt.push(K), et = this.charCodeToPatternIdxToConfig[K], Pe = this.patternIdxToConfig[K], Ot = Pe.length, Ot = Pe.length;
      const q = this.canModeBeOptimized[K] && this.config.safeMode === !1;
      et && q ? nt = !0 : nt = !1;
    }
    pr.call(this, n);
    let ie;
    const gr = this.config.recoveryEnabled;
    for (; E < d; ) {
      c = null, h = -1;
      const K = g.charCodeAt(E);
      let q;
      if (nt) {
        const $ = xe(K), Z = et[$];
        q = Z !== void 0 ? Z : dr;
      } else
        q = Pe;
      const pn = q.length;
      for (r = 0; r < pn; r++) {
        ie = q[r];
        const $ = ie.pattern;
        u = null;
        const Z = ie.short;
        if (Z !== !1 ? K === Z && (h = 1, c = $) : ie.isCustom === !0 ? (T = $.exec(g, E, ve, Je), T !== null ? (c = T[0], h = c.length, T.payload !== void 0 && (u = T.payload)) : c = null) : ($.lastIndex = E, h = this.matchLength($, e, E)), h !== -1) {
          if (o = ie.longerAlt, o !== void 0) {
            c = e.substring(E, E + h);
            const Me = o.length;
            for (i = 0; i < Me; i++) {
              const _e = Pe[o[i]], Fe = _e.pattern;
              if (l = null, _e.isCustom === !0 ? (T = Fe.exec(g, E, ve, Je), T !== null ? (a = T[0], T.payload !== void 0 && (l = T.payload)) : a = null) : (Fe.lastIndex = E, a = this.match(Fe, e, E)), a && a.length > c.length) {
                c = a, h = a.length, u = l, ie = _e;
                break;
              }
            }
          }
          break;
        }
      }
      if (h !== -1) {
        if (f = ie.group, f !== void 0 && (c = c !== null ? c : e.substring(E, E + h), p = ie.tokenTypeIdx, m = this.createTokenInstance(c, E, p, ie.tokenType, Qe, de, h), this.handlePayload(m, u), f === !1 ? L = this.addToken(ve, L, m) : Je[f].push(m)), zi === !0 && ie.canLineTerminator === !0) {
          let $ = 0, Z, Me;
          dn.lastIndex = 0;
          do
            c = c !== null ? c : e.substring(E, E + h), Z = dn.test(c), Z === !0 && (Me = dn.lastIndex - 1, $++);
          while (Z === !0);
          $ !== 0 ? (Qe = Qe + $, de = h - Me, this.updateTokenEndLineColumnLocation(m, f, Me, $, Qe, de, h)) : de = this.computeNewColumn(de, h);
        } else
          de = this.computeNewColumn(de, h);
        E = E + h, this.handleModes(ie, Vi, pr, m);
      } else {
        const $ = E, Z = Qe, Me = de;
        let _e = gr === !1;
        for (; _e === !1 && E < d; )
          for (E++, s = 0; s < Ot; s++) {
            const Fe = Pe[s], gn = Fe.pattern, Er = Fe.short;
            if (Er !== !1 ? g.charCodeAt(E) === Er && (_e = !0) : Fe.isCustom === !0 ? _e = gn.exec(g, E, ve, Je) !== null : (gn.lastIndex = E, _e = gn.exec(e) !== null), _e === !0)
              break;
          }
        if (I = E - $, de = this.computeNewColumn(de, I), O = this.config.errorMessageProvider.buildUnexpectedCharactersMessage(g, $, I, Z, Me, Ge(tt)), Ue.push({
          offset: $,
          line: Z,
          column: Me,
          length: I,
          message: O
        }), gr === !1)
          break;
      }
    }
    return this.hasCustom || (ve.length = L), {
      tokens: ve,
      groups: Je,
      errors: Ue
    };
  }
  handleModes(e, n, r, s) {
    if (e.pop === !0) {
      const i = e.push;
      n(s), i !== void 0 && r.call(this, i);
    } else e.push !== void 0 && r.call(this, e.push);
  }
  // TODO: decrease this under 600 characters? inspect stripping comments option in TSC compiler
  updateTokenEndLineColumnLocation(e, n, r, s, i, a, o) {
    let c, u;
    n !== void 0 && (c = r === o - 1, u = c ? -1 : 0, s === 1 && c === !0 || (e.endLine = i + u, e.endColumn = a - 1 + -u));
  }
  computeNewColumn(e, n) {
    return e + n;
  }
  createOffsetOnlyToken(e, n, r, s) {
    return {
      image: e,
      startOffset: n,
      tokenTypeIdx: r,
      tokenType: s
    };
  }
  createStartOnlyToken(e, n, r, s, i, a) {
    return {
      image: e,
      startOffset: n,
      startLine: i,
      startColumn: a,
      tokenTypeIdx: r,
      tokenType: s
    };
  }
  createFullToken(e, n, r, s, i, a, o) {
    return {
      image: e,
      startOffset: n,
      endOffset: n + o - 1,
      startLine: i,
      endLine: i,
      startColumn: a,
      endColumn: a + o - 1,
      tokenTypeIdx: r,
      tokenType: s
    };
  }
  addTokenUsingPush(e, n, r) {
    return e.push(r), n;
  }
  addTokenUsingMemberAccess(e, n, r) {
    return e[n] = r, n++, n;
  }
  handlePayloadNoCustom(e, n) {
  }
  handlePayloadWithCustom(e, n) {
    n !== null && (e.payload = n);
  }
  match(e, n, r) {
    return e.test(n) === !0 ? n.substring(r, e.lastIndex) : null;
  }
  matchLength(e, n, r) {
    return e.test(n) === !0 ? e.lastIndex - r : -1;
  }
}
G.SKIPPED = "This marks a skipped Token pattern, this means each token identified by it will be consumed and then thrown into oblivion, this can be used to for example to completely ignore whitespace.";
G.NA = /NOT_APPLICABLE/;
function lt(t) {
  return ai(t) ? t.LABEL : t.name;
}
function ai(t) {
  return j(t.LABEL) && t.LABEL !== "";
}
const Nf = "parent", rs = "categories", ss = "label", is = "group", as = "push_mode", os = "pop_mode", cs = "longer_alt", us = "line_breaks", ls = "start_chars_hint";
function C(t) {
  return Cf(t);
}
function Cf(t) {
  const e = t.pattern, n = {};
  if (n.name = t.name, Oe(e) || (n.PATTERN = e), R(t, Nf))
    throw `The parent property is no longer supported.
See: https://github.com/chevrotain/chevrotain/issues/564#issuecomment-349062346 for details.`;
  return R(t, rs) && (n.CATEGORIES = t[rs]), _t([n]), R(t, ss) && (n.LABEL = t[ss]), R(t, is) && (n.GROUP = t[is]), R(t, os) && (n.POP_MODE = t[os]), R(t, as) && (n.PUSH_MODE = t[as]), R(t, cs) && (n.LONGER_ALT = t[cs]), R(t, us) && (n.LINE_BREAKS = t[us]), R(t, ls) && (n.START_CHARS_HINT = t[ls]), n;
}
const He = C({ name: "EOF", pattern: G.NA });
_t([He]);
function ur(t, e, n, r, s, i, a, o) {
  return {
    image: e,
    startOffset: n,
    endOffset: r,
    startLine: s,
    endLine: i,
    startColumn: a,
    endColumn: o,
    tokenTypeIdx: t.tokenTypeIdx,
    tokenType: t
  };
}
function Lf(t, e) {
  return It(t, e);
}
const oi = {
  buildMismatchTokenMessage({ expected: t, actual: e, previous: n, ruleName: r }) {
    return `Expecting ${ai(t) ? `--> ${lt(t)} <--` : `token of type --> ${t.name} <--`} but found --> '${e.image}' <--`;
  },
  buildNotAllInputParsedMessage({ firstRedundant: t, ruleName: e }) {
    return "Redundant input, expecting EOF but found: " + t.image;
  },
  buildNoViableAltMessage({ expectedPathsPerAlt: t, actual: e, previous: n, customUserDescription: r, ruleName: s }) {
    const i = "Expecting: ", o = `
but found: '` + ue(e).image + "'";
    if (r)
      return i + r + o;
    {
      const c = Y(t, (f, p) => f.concat(p), []), u = A(c, (f) => `[${A(f, (p) => lt(p)).join(", ")}]`), h = `one of these possible Token sequences:
${A(u, (f, p) => `  ${p + 1}. ${f}`).join(`
`)}`;
      return i + h + o;
    }
  },
  buildEarlyExitMessage({ expectedIterationPaths: t, actual: e, customUserDescription: n, ruleName: r }) {
    const s = "Expecting: ", a = `
but found: '` + ue(e).image + "'";
    if (n)
      return s + n + a;
    {
      const c = `expecting at least one iteration which starts with one of these possible Token sequences::
  <${A(t, (u) => `[${A(u, (l) => lt(l)).join(",")}]`).join(" ,")}>`;
      return s + c + a;
    }
  }
};
Object.freeze(oi);
const yf = {
  buildRuleNotFoundError(t, e) {
    return "Invalid grammar, reference to a rule which is not defined: ->" + e.nonTerminalName + `<-
inside top level rule: ->` + t.name + "<-";
  }
}, De = {
  buildDuplicateFoundError(t, e) {
    function n(l) {
      return l instanceof M ? l.terminalType.name : l instanceof X ? l.nonTerminalName : "";
    }
    const r = t.name, s = ue(e), i = s.idx, a = pe(s), o = n(s), c = i > 0;
    let u = `->${a}${c ? i : ""}<- ${o ? `with argument: ->${o}<-` : ""}
                  appears more than once (${e.length} times) in the top level rule: ->${r}<-.                  
                  For further details see: https://chevrotain.io/docs/FAQ.html#NUMERICAL_SUFFIXES 
                  `;
    return u = u.replace(/[ \t]+/g, " "), u = u.replace(/\s\s+/g, `
`), u;
  },
  buildNamespaceConflictError(t) {
    return `Namespace conflict found in grammar.
The grammar has both a Terminal(Token) and a Non-Terminal(Rule) named: <${t.name}>.
To resolve this make sure each Terminal and Non-Terminal names are unique
This is easy to accomplish by using the convention that Terminal names start with an uppercase letter
and Non-Terminal names start with a lower case letter.`;
  },
  buildAlternationPrefixAmbiguityError(t) {
    const e = A(t.prefixPath, (s) => lt(s)).join(", "), n = t.alternation.idx === 0 ? "" : t.alternation.idx;
    return `Ambiguous alternatives: <${t.ambiguityIndices.join(" ,")}> due to common lookahead prefix
in <OR${n}> inside <${t.topLevelRule.name}> Rule,
<${e}> may appears as a prefix path in all these alternatives.
See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#COMMON_PREFIX
For Further details.`;
  },
  buildAlternationAmbiguityError(t) {
    const e = t.alternation.idx === 0 ? "" : t.alternation.idx, n = t.prefixPath.length === 0;
    let r = `Ambiguous Alternatives Detected: <${t.ambiguityIndices.join(" ,")}> in <OR${e}> inside <${t.topLevelRule.name}> Rule,
`;
    if (n)
      r += `These alternatives are all empty (match no tokens), making them indistinguishable.
Only the last alternative may be empty.
`;
    else {
      const s = A(t.prefixPath, (i) => lt(i)).join(", ");
      r += `<${s}> may appears as a prefix path in all these alternatives.
`;
    }
    return r += `See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#AMBIGUOUS_ALTERNATIVES
For Further details.`, r;
  },
  buildEmptyRepetitionError(t) {
    let e = pe(t.repetition);
    return t.repetition.idx !== 0 && (e += t.repetition.idx), `The repetition <${e}> within Rule <${t.topLevelRule.name}> can never consume any tokens.
This could lead to an infinite loop.`;
  },
  // TODO: remove - `errors_public` from nyc.config.js exclude
  //       once this method is fully removed from this file
  buildTokenNameError(t) {
    return "deprecated";
  },
  buildEmptyAlternationError(t) {
    return `Ambiguous empty alternative: <${t.emptyChoiceIdx + 1}> in <OR${t.alternation.idx}> inside <${t.topLevelRule.name}> Rule.
Only the last alternative may be an empty alternative.`;
  },
  buildTooManyAlternativesError(t) {
    return `An Alternation cannot have more than 256 alternatives:
<OR${t.alternation.idx}> inside <${t.topLevelRule.name}> Rule.
 has ${t.alternation.definition.length + 1} alternatives.`;
  },
  buildLeftRecursionError(t) {
    const e = t.topLevelRule.name, n = A(t.leftRecursionPath, (i) => i.name), r = `${e} --> ${n.concat([e]).join(" --> ")}`;
    return `Left Recursion found in grammar.
rule: <${e}> can be invoked from itself (directly or indirectly)
without consuming any Tokens. The grammar path that causes this is: 
 ${r}
 To fix this refactor your grammar to remove the left recursion.
see: https://en.wikipedia.org/wiki/LL_parser#Left_factoring.`;
  },
  // TODO: remove - `errors_public` from nyc.config.js exclude
  //       once this method is fully removed from this file
  buildInvalidRuleNameError(t) {
    return "deprecated";
  },
  buildDuplicateRuleNameError(t) {
    let e;
    return t.topLevelRule instanceof qe ? e = t.topLevelRule.name : e = t.topLevelRule, `Duplicate definition, rule: ->${e}<- is already defined in the grammar: ->${t.grammarName}<-`;
  }
};
function vf(t, e) {
  const n = new Pf(t, e);
  return n.resolveRefs(), n.errors;
}
class Pf extends Ze {
  constructor(e, n) {
    super(), this.nameToTopRule = e, this.errMsgProvider = n, this.errors = [];
  }
  resolveRefs() {
    _(b(this.nameToTopRule), (e) => {
      this.currTopLevel = e, e.accept(this);
    });
  }
  visitNonTerminal(e) {
    const n = this.nameToTopRule[e.nonTerminalName];
    if (n)
      e.referencedRule = n;
    else {
      const r = this.errMsgProvider.buildRuleNotFoundError(this.currTopLevel, e);
      this.errors.push({
        message: r,
        type: B.UNRESOLVED_SUBRULE_REF,
        ruleName: this.currTopLevel.name,
        unresolvedRefName: e.nonTerminalName
      });
    }
  }
}
class Mf extends cn {
  constructor(e, n) {
    super(), this.topProd = e, this.path = n, this.possibleTokTypes = [], this.nextProductionName = "", this.nextProductionOccurrence = 0, this.found = !1, this.isAtEndOfPath = !1;
  }
  startWalking() {
    if (this.found = !1, this.path.ruleStack[0] !== this.topProd.name)
      throw Error("The path does not start with the walker's top Rule!");
    return this.ruleStack = D(this.path.ruleStack).reverse(), this.occurrenceStack = D(this.path.occurrenceStack).reverse(), this.ruleStack.pop(), this.occurrenceStack.pop(), this.updateExpectedNext(), this.walk(this.topProd), this.possibleTokTypes;
  }
  walk(e, n = []) {
    this.found || super.walk(e, n);
  }
  walkProdRef(e, n, r) {
    if (e.referencedRule.name === this.nextProductionName && e.idx === this.nextProductionOccurrence) {
      const s = n.concat(r);
      this.updateExpectedNext(), this.walk(e.referencedRule, s);
    }
  }
  updateExpectedNext() {
    v(this.ruleStack) ? (this.nextProductionName = "", this.nextProductionOccurrence = 0, this.isAtEndOfPath = !0) : (this.nextProductionName = this.ruleStack.pop(), this.nextProductionOccurrence = this.occurrenceStack.pop());
  }
}
class kf extends Mf {
  constructor(e, n) {
    super(e, n), this.path = n, this.nextTerminalName = "", this.nextTerminalOccurrence = 0, this.nextTerminalName = this.path.lastTok.name, this.nextTerminalOccurrence = this.path.lastTokOccurrence;
  }
  walkTerminal(e, n, r) {
    if (this.isAtEndOfPath && e.terminalType.name === this.nextTerminalName && e.idx === this.nextTerminalOccurrence && !this.found) {
      const s = n.concat(r), i = new z({ definition: s });
      this.possibleTokTypes = Rt(i), this.found = !0;
    }
  }
}
class ln extends cn {
  constructor(e, n) {
    super(), this.topRule = e, this.occurrence = n, this.result = {
      token: void 0,
      occurrence: void 0,
      isEndOfRule: void 0
    };
  }
  startWalking() {
    return this.walk(this.topRule), this.result;
  }
}
class xf extends ln {
  walkMany(e, n, r) {
    if (e.idx === this.occurrence) {
      const s = ue(n.concat(r));
      this.result.isEndOfRule = s === void 0, s instanceof M && (this.result.token = s.terminalType, this.result.occurrence = s.idx);
    } else
      super.walkMany(e, n, r);
  }
}
class hs extends ln {
  walkManySep(e, n, r) {
    if (e.idx === this.occurrence) {
      const s = ue(n.concat(r));
      this.result.isEndOfRule = s === void 0, s instanceof M && (this.result.token = s.terminalType, this.result.occurrence = s.idx);
    } else
      super.walkManySep(e, n, r);
  }
}
class bf extends ln {
  walkAtLeastOne(e, n, r) {
    if (e.idx === this.occurrence) {
      const s = ue(n.concat(r));
      this.result.isEndOfRule = s === void 0, s instanceof M && (this.result.token = s.terminalType, this.result.occurrence = s.idx);
    } else
      super.walkAtLeastOne(e, n, r);
  }
}
class fs extends ln {
  walkAtLeastOneSep(e, n, r) {
    if (e.idx === this.occurrence) {
      const s = ue(n.concat(r));
      this.result.isEndOfRule = s === void 0, s instanceof M && (this.result.token = s.terminalType, this.result.occurrence = s.idx);
    } else
      super.walkAtLeastOneSep(e, n, r);
  }
}
function Mn(t, e, n = []) {
  n = D(n);
  let r = [], s = 0;
  function i(o) {
    return o.concat(F(t, s + 1));
  }
  function a(o) {
    const c = Mn(i(o), e, n);
    return r.concat(c);
  }
  for (; n.length < e && s < t.length; ) {
    const o = t[s];
    if (o instanceof z)
      return a(o.definition);
    if (o instanceof X)
      return a(o.definition);
    if (o instanceof W)
      r = a(o.definition);
    else if (o instanceof Re) {
      const c = o.definition.concat([
        new w({
          definition: o.definition
        })
      ]);
      return a(c);
    } else if (o instanceof Ie) {
      const c = [
        new z({ definition: o.definition }),
        new w({
          definition: [new M({ terminalType: o.separator })].concat(o.definition)
        })
      ];
      return a(c);
    } else if (o instanceof he) {
      const c = o.definition.concat([
        new w({
          definition: [new M({ terminalType: o.separator })].concat(o.definition)
        })
      ]);
      r = a(c);
    } else if (o instanceof w) {
      const c = o.definition.concat([
        new w({
          definition: o.definition
        })
      ]);
      r = a(c);
    } else {
      if (o instanceof fe)
        return _(o.definition, (c) => {
          v(c.definition) === !1 && (r = a(c.definition));
        }), r;
      if (o instanceof M)
        n.push(o.terminalType);
      else
        throw Error("non exhaustive match");
    }
    s++;
  }
  return r.push({
    partialPath: n,
    suffixDef: F(t, s)
  }), r;
}
function ci(t, e, n, r) {
  const s = "EXIT_NONE_TERMINAL", i = [s], a = "EXIT_ALTERNATIVE";
  let o = !1;
  const c = e.length, u = c - r - 1, l = [], h = [];
  for (h.push({
    idx: -1,
    def: t,
    ruleStack: [],
    occurrenceStack: []
  }); !v(h); ) {
    const f = h.pop();
    if (f === a) {
      o && Ge(h).idx <= u && h.pop();
      continue;
    }
    const p = f.def, m = f.idx, I = f.ruleStack, O = f.occurrenceStack;
    if (v(p))
      continue;
    const T = p[0];
    if (T === s) {
      const g = {
        idx: m,
        def: F(p),
        ruleStack: pt(I),
        occurrenceStack: pt(O)
      };
      h.push(g);
    } else if (T instanceof M)
      if (m < c - 1) {
        const g = m + 1, d = e[g];
        if (n(d, T.terminalType)) {
          const E = {
            idx: g,
            def: F(p),
            ruleStack: I,
            occurrenceStack: O
          };
          h.push(E);
        }
      } else if (m === c - 1)
        l.push({
          nextTokenType: T.terminalType,
          nextTokenOccurrence: T.idx,
          ruleStack: I,
          occurrenceStack: O
        }), o = !0;
      else
        throw Error("non exhaustive match");
    else if (T instanceof X) {
      const g = D(I);
      g.push(T.nonTerminalName);
      const d = D(O);
      d.push(T.idx);
      const E = {
        idx: m,
        def: T.definition.concat(i, F(p)),
        ruleStack: g,
        occurrenceStack: d
      };
      h.push(E);
    } else if (T instanceof W) {
      const g = {
        idx: m,
        def: F(p),
        ruleStack: I,
        occurrenceStack: O
      };
      h.push(g), h.push(a);
      const d = {
        idx: m,
        def: T.definition.concat(F(p)),
        ruleStack: I,
        occurrenceStack: O
      };
      h.push(d);
    } else if (T instanceof Re) {
      const g = new w({
        definition: T.definition,
        idx: T.idx
      }), d = T.definition.concat([g], F(p)), E = {
        idx: m,
        def: d,
        ruleStack: I,
        occurrenceStack: O
      };
      h.push(E);
    } else if (T instanceof Ie) {
      const g = new M({
        terminalType: T.separator
      }), d = new w({
        definition: [g].concat(T.definition),
        idx: T.idx
      }), E = T.definition.concat([d], F(p)), L = {
        idx: m,
        def: E,
        ruleStack: I,
        occurrenceStack: O
      };
      h.push(L);
    } else if (T instanceof he) {
      const g = {
        idx: m,
        def: F(p),
        ruleStack: I,
        occurrenceStack: O
      };
      h.push(g), h.push(a);
      const d = new M({
        terminalType: T.separator
      }), E = new w({
        definition: [d].concat(T.definition),
        idx: T.idx
      }), L = T.definition.concat([E], F(p)), se = {
        idx: m,
        def: L,
        ruleStack: I,
        occurrenceStack: O
      };
      h.push(se);
    } else if (T instanceof w) {
      const g = {
        idx: m,
        def: F(p),
        ruleStack: I,
        occurrenceStack: O
      };
      h.push(g), h.push(a);
      const d = new w({
        definition: T.definition,
        idx: T.idx
      }), E = T.definition.concat([d], F(p)), L = {
        idx: m,
        def: E,
        ruleStack: I,
        occurrenceStack: O
      };
      h.push(L);
    } else if (T instanceof fe)
      for (let g = T.definition.length - 1; g >= 0; g--) {
        const d = T.definition[g], E = {
          idx: m,
          def: d.definition.concat(F(p)),
          ruleStack: I,
          occurrenceStack: O
        };
        h.push(E), h.push(a);
      }
    else if (T instanceof z)
      h.push({
        idx: m,
        def: T.definition.concat(F(p)),
        ruleStack: I,
        occurrenceStack: O
      });
    else if (T instanceof qe)
      h.push(wf(T, m, I, O));
    else
      throw Error("non exhaustive match");
  }
  return l;
}
function wf(t, e, n, r) {
  const s = D(n);
  s.push(t.name);
  const i = D(r);
  return i.push(1), {
    idx: e,
    def: t.definition,
    ruleStack: s,
    occurrenceStack: i
  };
}
var k;
(function(t) {
  t[t.OPTION = 0] = "OPTION", t[t.REPETITION = 1] = "REPETITION", t[t.REPETITION_MANDATORY = 2] = "REPETITION_MANDATORY", t[t.REPETITION_MANDATORY_WITH_SEPARATOR = 3] = "REPETITION_MANDATORY_WITH_SEPARATOR", t[t.REPETITION_WITH_SEPARATOR = 4] = "REPETITION_WITH_SEPARATOR", t[t.ALTERNATION = 5] = "ALTERNATION";
})(k || (k = {}));
function ui(t) {
  if (t instanceof W || t === "Option")
    return k.OPTION;
  if (t instanceof w || t === "Repetition")
    return k.REPETITION;
  if (t instanceof Re || t === "RepetitionMandatory")
    return k.REPETITION_MANDATORY;
  if (t instanceof Ie || t === "RepetitionMandatoryWithSeparator")
    return k.REPETITION_MANDATORY_WITH_SEPARATOR;
  if (t instanceof he || t === "RepetitionWithSeparator")
    return k.REPETITION_WITH_SEPARATOR;
  if (t instanceof fe || t === "Alternation")
    return k.ALTERNATION;
  throw Error("non exhaustive match");
}
function Uf(t, e, n, r, s, i) {
  const a = lr(t, e, n), o = fi(a) ? Wt : It;
  return i(a, r, o, s);
}
function Ff(t, e, n, r, s, i) {
  const a = hr(t, e, s, n), o = fi(a) ? Wt : It;
  return i(a[0], o, r);
}
function Df(t, e, n, r) {
  const s = t.length, i = oe(t, (a) => oe(a, (o) => o.length === 1));
  if (e)
    return function(a) {
      const o = A(a, (c) => c.GATE);
      for (let c = 0; c < s; c++) {
        const u = t[c], l = u.length, h = o[c];
        if (!(h !== void 0 && h.call(this) === !1))
          e: for (let f = 0; f < l; f++) {
            const p = u[f], m = p.length;
            for (let I = 0; I < m; I++) {
              const O = this.LA(I + 1);
              if (n(O, p[I]) === !1)
                continue e;
            }
            return c;
          }
      }
    };
  if (i && !r) {
    const a = A(t, (c) => ae(c)), o = Y(a, (c, u, l) => (_(u, (h) => {
      R(c, h.tokenTypeIdx) || (c[h.tokenTypeIdx] = l), _(h.categoryMatches, (f) => {
        R(c, f) || (c[f] = l);
      });
    }), c), {});
    return function() {
      const c = this.LA(1);
      return o[c.tokenTypeIdx];
    };
  } else
    return function() {
      for (let a = 0; a < s; a++) {
        const o = t[a], c = o.length;
        e: for (let u = 0; u < c; u++) {
          const l = o[u], h = l.length;
          for (let f = 0; f < h; f++) {
            const p = this.LA(f + 1);
            if (n(p, l[f]) === !1)
              continue e;
          }
          return a;
        }
      }
    };
}
function $f(t, e, n) {
  const r = oe(t, (i) => i.length === 1), s = t.length;
  if (r && !n) {
    const i = ae(t);
    if (i.length === 1 && v(i[0].categoryMatches)) {
      const o = i[0].tokenTypeIdx;
      return function() {
        return this.LA(1).tokenTypeIdx === o;
      };
    } else {
      const a = Y(i, (o, c, u) => (o[c.tokenTypeIdx] = !0, _(c.categoryMatches, (l) => {
        o[l] = !0;
      }), o), []);
      return function() {
        const o = this.LA(1);
        return a[o.tokenTypeIdx] === !0;
      };
    }
  } else
    return function() {
      e: for (let i = 0; i < s; i++) {
        const a = t[i], o = a.length;
        for (let c = 0; c < o; c++) {
          const u = this.LA(c + 1);
          if (e(u, a[c]) === !1)
            continue e;
        }
        return !0;
      }
      return !1;
    };
}
class Gf extends cn {
  constructor(e, n, r) {
    super(), this.topProd = e, this.targetOccurrence = n, this.targetProdType = r;
  }
  startWalking() {
    return this.walk(this.topProd), this.restDef;
  }
  checkIsTarget(e, n, r, s) {
    return e.idx === this.targetOccurrence && this.targetProdType === n ? (this.restDef = r.concat(s), !0) : !1;
  }
  walkOption(e, n, r) {
    this.checkIsTarget(e, k.OPTION, n, r) || super.walkOption(e, n, r);
  }
  walkAtLeastOne(e, n, r) {
    this.checkIsTarget(e, k.REPETITION_MANDATORY, n, r) || super.walkOption(e, n, r);
  }
  walkAtLeastOneSep(e, n, r) {
    this.checkIsTarget(e, k.REPETITION_MANDATORY_WITH_SEPARATOR, n, r) || super.walkOption(e, n, r);
  }
  walkMany(e, n, r) {
    this.checkIsTarget(e, k.REPETITION, n, r) || super.walkOption(e, n, r);
  }
  walkManySep(e, n, r) {
    this.checkIsTarget(e, k.REPETITION_WITH_SEPARATOR, n, r) || super.walkOption(e, n, r);
  }
}
class li extends Ze {
  constructor(e, n, r) {
    super(), this.targetOccurrence = e, this.targetProdType = n, this.targetRef = r, this.result = [];
  }
  checkIsTarget(e, n) {
    e.idx === this.targetOccurrence && this.targetProdType === n && (this.targetRef === void 0 || e === this.targetRef) && (this.result = e.definition);
  }
  visitOption(e) {
    this.checkIsTarget(e, k.OPTION);
  }
  visitRepetition(e) {
    this.checkIsTarget(e, k.REPETITION);
  }
  visitRepetitionMandatory(e) {
    this.checkIsTarget(e, k.REPETITION_MANDATORY);
  }
  visitRepetitionMandatoryWithSeparator(e) {
    this.checkIsTarget(e, k.REPETITION_MANDATORY_WITH_SEPARATOR);
  }
  visitRepetitionWithSeparator(e) {
    this.checkIsTarget(e, k.REPETITION_WITH_SEPARATOR);
  }
  visitAlternation(e) {
    this.checkIsTarget(e, k.ALTERNATION);
  }
}
function ds(t) {
  const e = new Array(t);
  for (let n = 0; n < t; n++)
    e[n] = [];
  return e;
}
function In(t) {
  let e = [""];
  for (let n = 0; n < t.length; n++) {
    const r = t[n], s = [];
    for (let i = 0; i < e.length; i++) {
      const a = e[i];
      s.push(a + "_" + r.tokenTypeIdx);
      for (let o = 0; o < r.categoryMatches.length; o++) {
        const c = "_" + r.categoryMatches[o];
        s.push(a + c);
      }
    }
    e = s;
  }
  return e;
}
function Bf(t, e, n) {
  for (let r = 0; r < t.length; r++) {
    if (r === n)
      continue;
    const s = t[r];
    for (let i = 0; i < e.length; i++) {
      const a = e[i];
      if (s[a] === !0)
        return !1;
    }
  }
  return !0;
}
function hi(t, e) {
  const n = A(t, (a) => Mn([a], 1)), r = ds(n.length), s = A(n, (a) => {
    const o = {};
    return _(a, (c) => {
      const u = In(c.partialPath);
      _(u, (l) => {
        o[l] = !0;
      });
    }), o;
  });
  let i = n;
  for (let a = 1; a <= e; a++) {
    const o = i;
    i = ds(o.length);
    for (let c = 0; c < o.length; c++) {
      const u = o[c];
      for (let l = 0; l < u.length; l++) {
        const h = u[l].partialPath, f = u[l].suffixDef, p = In(h);
        if (Bf(s, p, c) || v(f) || h.length === e) {
          const I = r[c];
          if (kn(I, h) === !1) {
            I.push(h);
            for (let O = 0; O < p.length; O++) {
              const T = p[O];
              s[c][T] = !0;
            }
          }
        } else {
          const I = Mn(f, a + 1, h);
          i[c] = i[c].concat(I), _(I, (O) => {
            const T = In(O.partialPath);
            _(T, (g) => {
              s[c][g] = !0;
            });
          });
        }
      }
    }
  }
  return r;
}
function lr(t, e, n, r) {
  const s = new li(t, k.ALTERNATION, r);
  return e.accept(s), hi(s.result, n);
}
function hr(t, e, n, r) {
  const s = new li(t, n);
  e.accept(s);
  const i = s.result, o = new Gf(e, t, n).startWalking(), c = new z({ definition: i }), u = new z({ definition: o });
  return hi([c, u], r);
}
function kn(t, e) {
  e: for (let n = 0; n < t.length; n++) {
    const r = t[n];
    if (r.length === e.length) {
      for (let s = 0; s < r.length; s++) {
        const i = e[s], a = r[s];
        if ((i === a || a.categoryMatchesMap[i.tokenTypeIdx] !== void 0) === !1)
          continue e;
      }
      return !0;
    }
  }
  return !1;
}
function Hf(t, e) {
  return t.length < e.length && oe(t, (n, r) => {
    const s = e[r];
    return n === s || s.categoryMatchesMap[n.tokenTypeIdx];
  });
}
function fi(t) {
  return oe(t, (e) => oe(e, (n) => oe(n, (r) => v(r.categoryMatches))));
}
function Kf(t) {
  const e = t.lookaheadStrategy.validate({
    rules: t.rules,
    tokenTypes: t.tokenTypes,
    grammarName: t.grammarName
  });
  return A(e, (n) => Object.assign({ type: B.CUSTOM_LOOKAHEAD_VALIDATION }, n));
}
function Wf(t, e, n, r) {
  const s = J(t, (c) => jf(c, n)), i = rd(t, e, n), a = J(t, (c) => Jf(c, n)), o = J(t, (c) => Yf(c, t, r, n));
  return s.concat(i, a, o);
}
function jf(t, e) {
  const n = new Vf();
  t.accept(n);
  const r = n.allProductions, s = nh(r, zf), i = le(s, (o) => o.length > 1);
  return A(b(i), (o) => {
    const c = ue(o), u = e.buildDuplicateFoundError(t, o), l = pe(c), h = {
      message: u,
      type: B.DUPLICATE_PRODUCTIONS,
      ruleName: t.name,
      dslName: l,
      occurrence: c.idx
    }, f = di(c);
    return f && (h.parameter = f), h;
  });
}
function zf(t) {
  return `${pe(t)}_#_${t.idx}_#_${di(t)}`;
}
function di(t) {
  return t instanceof M ? t.terminalType.name : t instanceof X ? t.nonTerminalName : "";
}
class Vf extends Ze {
  constructor() {
    super(...arguments), this.allProductions = [];
  }
  visitNonTerminal(e) {
    this.allProductions.push(e);
  }
  visitOption(e) {
    this.allProductions.push(e);
  }
  visitRepetitionWithSeparator(e) {
    this.allProductions.push(e);
  }
  visitRepetitionMandatory(e) {
    this.allProductions.push(e);
  }
  visitRepetitionMandatoryWithSeparator(e) {
    this.allProductions.push(e);
  }
  visitRepetition(e) {
    this.allProductions.push(e);
  }
  visitAlternation(e) {
    this.allProductions.push(e);
  }
  visitTerminal(e) {
    this.allProductions.push(e);
  }
}
function Yf(t, e, n, r) {
  const s = [];
  if (Y(e, (a, o) => o.name === t.name ? a + 1 : a, 0) > 1) {
    const a = r.buildDuplicateRuleNameError({
      topLevelRule: t,
      grammarName: n
    });
    s.push({
      message: a,
      type: B.DUPLICATE_RULE_NAME,
      ruleName: t.name
    });
  }
  return s;
}
function Xf(t, e, n) {
  const r = [];
  let s;
  return H(e, t) || (s = `Invalid rule override, rule: ->${t}<- cannot be overridden in the grammar: ->${n}<-as it is not defined in any of the super grammars `, r.push({
    message: s,
    type: B.INVALID_RULE_OVERRIDE,
    ruleName: t
  })), r;
}
function pi(t, e, n, r = []) {
  const s = [], i = wt(e.definition);
  if (v(i))
    return [];
  {
    const a = t.name;
    H(i, t) && s.push({
      message: n.buildLeftRecursionError({
        topLevelRule: t,
        leftRecursionPath: r
      }),
      type: B.LEFT_RECURSION,
      ruleName: a
    });
    const c = an(i, r.concat([t])), u = J(c, (l) => {
      const h = D(r);
      return h.push(l), pi(t, l, n, h);
    });
    return s.concat(u);
  }
}
function wt(t) {
  let e = [];
  if (v(t))
    return e;
  const n = ue(t);
  if (n instanceof X)
    e.push(n.referencedRule);
  else if (n instanceof z || n instanceof W || n instanceof Re || n instanceof Ie || n instanceof he || n instanceof w)
    e = e.concat(wt(n.definition));
  else if (n instanceof fe)
    e = ae(A(n.definition, (i) => wt(i.definition)));
  else if (!(n instanceof M)) throw Error("non exhaustive match");
  const r = Gt(n), s = t.length > 1;
  if (r && s) {
    const i = F(t);
    return e.concat(wt(i));
  } else
    return e;
}
class fr extends Ze {
  constructor() {
    super(...arguments), this.alternations = [];
  }
  visitAlternation(e) {
    this.alternations.push(e);
  }
}
function qf(t, e) {
  const n = new fr();
  t.accept(n);
  const r = n.alternations;
  return J(r, (i) => {
    const a = pt(i.definition);
    return J(a, (o, c) => {
      const u = ci([o], [], It, 1);
      return v(u) ? [
        {
          message: e.buildEmptyAlternationError({
            topLevelRule: t,
            alternation: i,
            emptyChoiceIdx: c
          }),
          type: B.NONE_LAST_EMPTY_ALT,
          ruleName: t.name,
          occurrence: i.idx,
          alternative: c + 1
        }
      ] : [];
    });
  });
}
function Zf(t, e, n) {
  const r = new fr();
  t.accept(r);
  let s = r.alternations;
  return s = on(s, (a) => a.ignoreAmbiguities === !0), J(s, (a) => {
    const o = a.idx, c = a.maxLookahead || e, u = lr(o, t, c, a), l = td(u, a, t, n), h = nd(u, a, t, n);
    return l.concat(h);
  });
}
class Qf extends Ze {
  constructor() {
    super(...arguments), this.allProductions = [];
  }
  visitRepetitionWithSeparator(e) {
    this.allProductions.push(e);
  }
  visitRepetitionMandatory(e) {
    this.allProductions.push(e);
  }
  visitRepetitionMandatoryWithSeparator(e) {
    this.allProductions.push(e);
  }
  visitRepetition(e) {
    this.allProductions.push(e);
  }
}
function Jf(t, e) {
  const n = new fr();
  t.accept(n);
  const r = n.alternations;
  return J(r, (i) => i.definition.length > 255 ? [
    {
      message: e.buildTooManyAlternativesError({
        topLevelRule: t,
        alternation: i
      }),
      type: B.TOO_MANY_ALTS,
      ruleName: t.name,
      occurrence: i.idx
    }
  ] : []);
}
function ed(t, e, n) {
  const r = [];
  return _(t, (s) => {
    const i = new Qf();
    s.accept(i);
    const a = i.allProductions;
    _(a, (o) => {
      const c = ui(o), u = o.maxLookahead || e, l = o.idx, f = hr(l, s, c, u)[0];
      if (v(ae(f))) {
        const p = n.buildEmptyRepetitionError({
          topLevelRule: s,
          repetition: o
        });
        r.push({
          message: p,
          type: B.NO_NON_EMPTY_LOOKAHEAD,
          ruleName: s.name
        });
      }
    });
  }), r;
}
function td(t, e, n, r) {
  const s = [], i = Y(t, (o, c, u) => (e.definition[u].ignoreAmbiguities === !0 || _(c, (l) => {
    const h = [u];
    _(t, (f, p) => {
      u !== p && kn(f, l) && // ignore (skip) ambiguities with this "other" alternative
      e.definition[p].ignoreAmbiguities !== !0 && h.push(p);
    }), h.length > 1 && !kn(s, l) && (s.push(l), o.push({
      alts: h,
      path: l
    }));
  }), o), []);
  return A(i, (o) => {
    const c = A(o.alts, (l) => l + 1);
    return {
      message: r.buildAlternationAmbiguityError({
        topLevelRule: n,
        alternation: e,
        ambiguityIndices: c,
        prefixPath: o.path
      }),
      type: B.AMBIGUOUS_ALTS,
      ruleName: n.name,
      occurrence: e.idx,
      alternatives: o.alts
    };
  });
}
function nd(t, e, n, r) {
  const s = Y(t, (a, o, c) => {
    const u = A(o, (l) => ({ idx: c, path: l }));
    return a.concat(u);
  }, []);
  return At(J(s, (a) => {
    if (e.definition[a.idx].ignoreAmbiguities === !0)
      return [];
    const c = a.idx, u = a.path, l = re(s, (f) => (
      // ignore (skip) ambiguities with this "other" alternative
      e.definition[f.idx].ignoreAmbiguities !== !0 && f.idx < c && // checking for strict prefix because identical lookaheads
      // will be be detected using a different validation.
      Hf(f.path, u)
    ));
    return A(l, (f) => {
      const p = [f.idx + 1, c + 1], m = e.idx === 0 ? "" : e.idx;
      return {
        message: r.buildAlternationPrefixAmbiguityError({
          topLevelRule: n,
          alternation: e,
          ambiguityIndices: p,
          prefixPath: f.path
        }),
        type: B.AMBIGUOUS_PREFIX_ALTS,
        ruleName: n.name,
        occurrence: m,
        alternatives: p
      };
    });
  }));
}
function rd(t, e, n) {
  const r = [], s = A(e, (i) => i.name);
  return _(t, (i) => {
    const a = i.name;
    if (H(s, a)) {
      const o = n.buildNamespaceConflictError(i);
      r.push({
        message: o,
        type: B.CONFLICT_TOKENS_RULES_NAMESPACE,
        ruleName: a
      });
    }
  }), r;
}
function sd(t) {
  const e = ir(t, {
    errMsgProvider: yf
  }), n = {};
  return _(t.rules, (r) => {
    n[r.name] = r;
  }), vf(n, e.errMsgProvider);
}
function id(t) {
  return t = ir(t, {
    errMsgProvider: De
  }), Wf(t.rules, t.tokenTypes, t.errMsgProvider, t.grammarName);
}
const gi = "MismatchedTokenException", Ei = "NoViableAltException", Ti = "EarlyExitException", mi = "NotAllInputParsedException", Ai = [
  gi,
  Ei,
  Ti,
  mi
];
Object.freeze(Ai);
function jt(t) {
  return H(Ai, t.name);
}
class hn extends Error {
  constructor(e, n) {
    super(e), this.token = n, this.resyncedTokens = [], Object.setPrototypeOf(this, new.target.prototype), Error.captureStackTrace && Error.captureStackTrace(this, this.constructor);
  }
}
class Ri extends hn {
  constructor(e, n, r) {
    super(e, n), this.previousToken = r, this.name = gi;
  }
}
class ad extends hn {
  constructor(e, n, r) {
    super(e, n), this.previousToken = r, this.name = Ei;
  }
}
class od extends hn {
  constructor(e, n) {
    super(e, n), this.name = mi;
  }
}
class cd extends hn {
  constructor(e, n, r) {
    super(e, n), this.previousToken = r, this.name = Ti;
  }
}
const _n = {}, Ii = "InRuleRecoveryException";
class ud extends Error {
  constructor(e) {
    super(e), this.name = Ii;
  }
}
class ld {
  initRecoverable(e) {
    this.firstAfterRepMap = {}, this.resyncFollows = {}, this.recoveryEnabled = R(e, "recoveryEnabled") ? e.recoveryEnabled : Ne.recoveryEnabled, this.recoveryEnabled && (this.attemptInRepetitionRecovery = hd);
  }
  getTokenToInsert(e) {
    const n = ur(e, "", NaN, NaN, NaN, NaN, NaN, NaN);
    return n.isInsertedInRecovery = !0, n;
  }
  canTokenTypeBeInsertedInRecovery(e) {
    return !0;
  }
  canTokenTypeBeDeletedInRecovery(e) {
    return !0;
  }
  tryInRepetitionRecovery(e, n, r, s) {
    const i = this.findReSyncTokenType(), a = this.exportLexerState(), o = [];
    let c = !1;
    const u = this.LA(1);
    let l = this.LA(1);
    const h = () => {
      const f = this.LA(0), p = this.errorMessageProvider.buildMismatchTokenMessage({
        expected: s,
        actual: u,
        previous: f,
        ruleName: this.getCurrRuleFullName()
      }), m = new Ri(p, u, this.LA(0));
      m.resyncedTokens = pt(o), this.SAVE_ERROR(m);
    };
    for (; !c; )
      if (this.tokenMatcher(l, s)) {
        h();
        return;
      } else if (r.call(this)) {
        h(), e.apply(this, n);
        return;
      } else this.tokenMatcher(l, i) ? c = !0 : (l = this.SKIP_TOKEN(), this.addToResyncTokens(l, o));
    this.importLexerState(a);
  }
  shouldInRepetitionRecoveryBeTried(e, n, r) {
    return !(r === !1 || this.tokenMatcher(this.LA(1), e) || this.isBackTracking() || this.canPerformInRuleRecovery(e, this.getFollowsForInRuleRecovery(e, n)));
  }
  // Error Recovery functionality
  getFollowsForInRuleRecovery(e, n) {
    const r = this.getCurrentGrammarPath(e, n);
    return this.getNextPossibleTokenTypes(r);
  }
  tryInRuleRecovery(e, n) {
    if (this.canRecoverWithSingleTokenInsertion(e, n))
      return this.getTokenToInsert(e);
    if (this.canRecoverWithSingleTokenDeletion(e)) {
      const r = this.SKIP_TOKEN();
      return this.consumeToken(), r;
    }
    throw new ud("sad sad panda");
  }
  canPerformInRuleRecovery(e, n) {
    return this.canRecoverWithSingleTokenInsertion(e, n) || this.canRecoverWithSingleTokenDeletion(e);
  }
  canRecoverWithSingleTokenInsertion(e, n) {
    if (!this.canTokenTypeBeInsertedInRecovery(e) || v(n))
      return !1;
    const r = this.LA(1);
    return Xe(n, (i) => this.tokenMatcher(r, i)) !== void 0;
  }
  canRecoverWithSingleTokenDeletion(e) {
    return this.canTokenTypeBeDeletedInRecovery(e) ? this.tokenMatcher(this.LA(2), e) : !1;
  }
  isInCurrentRuleReSyncSet(e) {
    const n = this.getCurrFollowKey(), r = this.getFollowSetFromFollowKey(n);
    return H(r, e);
  }
  findReSyncTokenType() {
    const e = this.flattenFollowSet();
    let n = this.LA(1), r = 2;
    for (; ; ) {
      const s = Xe(e, (i) => Lf(n, i));
      if (s !== void 0)
        return s;
      n = this.LA(r), r++;
    }
  }
  getCurrFollowKey() {
    if (this.RULE_STACK.length === 1)
      return _n;
    const e = this.getLastExplicitRuleShortName(), n = this.getLastExplicitRuleOccurrenceIndex(), r = this.getPreviousExplicitRuleShortName();
    return {
      ruleName: this.shortRuleNameToFullName(e),
      idxInCallingRule: n,
      inRule: this.shortRuleNameToFullName(r)
    };
  }
  buildFullFollowKeyStack() {
    const e = this.RULE_STACK, n = this.RULE_OCCURRENCE_STACK;
    return A(e, (r, s) => s === 0 ? _n : {
      ruleName: this.shortRuleNameToFullName(r),
      idxInCallingRule: n[s],
      inRule: this.shortRuleNameToFullName(e[s - 1])
    });
  }
  flattenFollowSet() {
    const e = A(this.buildFullFollowKeyStack(), (n) => this.getFollowSetFromFollowKey(n));
    return ae(e);
  }
  getFollowSetFromFollowKey(e) {
    if (e === _n)
      return [He];
    const n = e.ruleName + e.idxInCallingRule + Qs + e.inRule;
    return this.resyncFollows[n];
  }
  // It does not make any sense to include a virtual EOF token in the list of resynced tokens
  // as EOF does not really exist and thus does not contain any useful information (line/column numbers)
  addToResyncTokens(e, n) {
    return this.tokenMatcher(e, He) || n.push(e), n;
  }
  reSyncTo(e) {
    const n = [];
    let r = this.LA(1);
    for (; this.tokenMatcher(r, e) === !1; )
      r = this.SKIP_TOKEN(), this.addToResyncTokens(r, n);
    return pt(n);
  }
  attemptInRepetitionRecovery(e, n, r, s, i, a, o) {
  }
  getCurrentGrammarPath(e, n) {
    const r = this.getHumanReadableRuleStack(), s = D(this.RULE_OCCURRENCE_STACK);
    return {
      ruleStack: r,
      occurrenceStack: s,
      lastTok: e,
      lastTokOccurrence: n
    };
  }
  getHumanReadableRuleStack() {
    return A(this.RULE_STACK, (e) => this.shortRuleNameToFullName(e));
  }
}
function hd(t, e, n, r, s, i, a) {
  const o = this.getKeyForAutomaticLookahead(r, s);
  let c = this.firstAfterRepMap[o];
  if (c === void 0) {
    const f = this.getCurrRuleFullName(), p = this.getGAstProductions()[f];
    c = new i(p, s).startWalking(), this.firstAfterRepMap[o] = c;
  }
  let u = c.token, l = c.occurrence;
  const h = c.isEndOfRule;
  this.RULE_STACK.length === 1 && h && u === void 0 && (u = He, l = 1), !(u === void 0 || l === void 0) && this.shouldInRepetitionRecoveryBeTried(u, l, a) && this.tryInRepetitionRecovery(t, e, n, u);
}
const fd = 4, we = 8, _i = 1 << we, Si = 2 << we, xn = 3 << we, bn = 4 << we, wn = 5 << we, Ut = 6 << we;
function Sn(t, e, n) {
  return n | e | t;
}
class dd {
  constructor(e) {
    var n;
    this.maxLookahead = (n = e?.maxLookahead) !== null && n !== void 0 ? n : Ne.maxLookahead;
  }
  validate(e) {
    const n = this.validateNoLeftRecursion(e.rules);
    if (v(n)) {
      const r = this.validateEmptyOrAlternatives(e.rules), s = this.validateAmbiguousAlternationAlternatives(e.rules, this.maxLookahead), i = this.validateSomeNonEmptyLookaheadPath(e.rules, this.maxLookahead);
      return [
        ...n,
        ...r,
        ...s,
        ...i
      ];
    }
    return n;
  }
  validateNoLeftRecursion(e) {
    return J(e, (n) => pi(n, n, De));
  }
  validateEmptyOrAlternatives(e) {
    return J(e, (n) => qf(n, De));
  }
  validateAmbiguousAlternationAlternatives(e, n) {
    return J(e, (r) => Zf(r, n, De));
  }
  validateSomeNonEmptyLookaheadPath(e, n) {
    return ed(e, n, De);
  }
  buildLookaheadForAlternation(e) {
    return Uf(e.prodOccurrence, e.rule, e.maxLookahead, e.hasPredicates, e.dynamicTokensEnabled, Df);
  }
  buildLookaheadForOptional(e) {
    return Ff(e.prodOccurrence, e.rule, e.maxLookahead, e.dynamicTokensEnabled, ui(e.prodType), $f);
  }
}
class pd {
  initLooksAhead(e) {
    this.dynamicTokensEnabled = R(e, "dynamicTokensEnabled") ? e.dynamicTokensEnabled : Ne.dynamicTokensEnabled, this.maxLookahead = R(e, "maxLookahead") ? e.maxLookahead : Ne.maxLookahead, this.lookaheadStrategy = R(e, "lookaheadStrategy") ? e.lookaheadStrategy : new dd({ maxLookahead: this.maxLookahead }), this.lookAheadFuncsCache = /* @__PURE__ */ new Map();
  }
  preComputeLookaheadFunctions(e) {
    _(e, (n) => {
      this.TRACE_INIT(`${n.name} Rule Lookahead`, () => {
        const { alternation: r, repetition: s, option: i, repetitionMandatory: a, repetitionMandatoryWithSeparator: o, repetitionWithSeparator: c } = Ed(n);
        _(r, (u) => {
          const l = u.idx === 0 ? "" : u.idx;
          this.TRACE_INIT(`${pe(u)}${l}`, () => {
            const h = this.lookaheadStrategy.buildLookaheadForAlternation({
              prodOccurrence: u.idx,
              rule: n,
              maxLookahead: u.maxLookahead || this.maxLookahead,
              hasPredicates: u.hasPredicates,
              dynamicTokensEnabled: this.dynamicTokensEnabled
            }), f = Sn(this.fullRuleNameToShort[n.name], _i, u.idx);
            this.setLaFuncCache(f, h);
          });
        }), _(s, (u) => {
          this.computeLookaheadFunc(n, u.idx, xn, "Repetition", u.maxLookahead, pe(u));
        }), _(i, (u) => {
          this.computeLookaheadFunc(n, u.idx, Si, "Option", u.maxLookahead, pe(u));
        }), _(a, (u) => {
          this.computeLookaheadFunc(n, u.idx, bn, "RepetitionMandatory", u.maxLookahead, pe(u));
        }), _(o, (u) => {
          this.computeLookaheadFunc(n, u.idx, Ut, "RepetitionMandatoryWithSeparator", u.maxLookahead, pe(u));
        }), _(c, (u) => {
          this.computeLookaheadFunc(n, u.idx, wn, "RepetitionWithSeparator", u.maxLookahead, pe(u));
        });
      });
    });
  }
  computeLookaheadFunc(e, n, r, s, i, a) {
    this.TRACE_INIT(`${a}${n === 0 ? "" : n}`, () => {
      const o = this.lookaheadStrategy.buildLookaheadForOptional({
        prodOccurrence: n,
        rule: e,
        maxLookahead: i || this.maxLookahead,
        dynamicTokensEnabled: this.dynamicTokensEnabled,
        prodType: s
      }), c = Sn(this.fullRuleNameToShort[e.name], r, n);
      this.setLaFuncCache(c, o);
    });
  }
  // this actually returns a number, but it is always used as a string (object prop key)
  getKeyForAutomaticLookahead(e, n) {
    const r = this.getLastExplicitRuleShortName();
    return Sn(r, e, n);
  }
  getLaFuncFromCache(e) {
    return this.lookAheadFuncsCache.get(e);
  }
  /* istanbul ignore next */
  setLaFuncCache(e, n) {
    this.lookAheadFuncsCache.set(e, n);
  }
}
class gd extends Ze {
  constructor() {
    super(...arguments), this.dslMethods = {
      option: [],
      alternation: [],
      repetition: [],
      repetitionWithSeparator: [],
      repetitionMandatory: [],
      repetitionMandatoryWithSeparator: []
    };
  }
  reset() {
    this.dslMethods = {
      option: [],
      alternation: [],
      repetition: [],
      repetitionWithSeparator: [],
      repetitionMandatory: [],
      repetitionMandatoryWithSeparator: []
    };
  }
  visitOption(e) {
    this.dslMethods.option.push(e);
  }
  visitRepetitionWithSeparator(e) {
    this.dslMethods.repetitionWithSeparator.push(e);
  }
  visitRepetitionMandatory(e) {
    this.dslMethods.repetitionMandatory.push(e);
  }
  visitRepetitionMandatoryWithSeparator(e) {
    this.dslMethods.repetitionMandatoryWithSeparator.push(e);
  }
  visitRepetition(e) {
    this.dslMethods.repetition.push(e);
  }
  visitAlternation(e) {
    this.dslMethods.alternation.push(e);
  }
}
const vt = new gd();
function Ed(t) {
  vt.reset(), t.accept(vt);
  const e = vt.dslMethods;
  return vt.reset(), e;
}
function ps(t, e) {
  isNaN(t.startOffset) === !0 ? (t.startOffset = e.startOffset, t.endOffset = e.endOffset) : t.endOffset < e.endOffset && (t.endOffset = e.endOffset);
}
function gs(t, e) {
  isNaN(t.startOffset) === !0 ? (t.startOffset = e.startOffset, t.startColumn = e.startColumn, t.startLine = e.startLine, t.endOffset = e.endOffset, t.endColumn = e.endColumn, t.endLine = e.endLine) : t.endOffset < e.endOffset && (t.endOffset = e.endOffset, t.endColumn = e.endColumn, t.endLine = e.endLine);
}
function Td(t, e, n) {
  t.children[n] === void 0 ? t.children[n] = [e] : t.children[n].push(e);
}
function md(t, e, n) {
  t.children[e] === void 0 ? t.children[e] = [n] : t.children[e].push(n);
}
const Ad = "name";
function Oi(t, e) {
  Object.defineProperty(t, Ad, {
    enumerable: !1,
    configurable: !0,
    writable: !1,
    value: e
  });
}
function Rd(t, e) {
  const n = ne(t), r = n.length;
  for (let s = 0; s < r; s++) {
    const i = n[s], a = t[i], o = a.length;
    for (let c = 0; c < o; c++) {
      const u = a[c];
      u.tokenTypeIdx === void 0 && this[u.name](u.children, e);
    }
  }
}
function Id(t, e) {
  const n = function() {
  };
  Oi(n, t + "BaseSemantics");
  const r = {
    visit: function(s, i) {
      if (N(s) && (s = s[0]), !Oe(s))
        return this[s.name](s.children, i);
    },
    validateVisitor: function() {
      const s = Sd(this, e);
      if (!v(s)) {
        const i = A(s, (a) => a.msg);
        throw Error(`Errors Detected in CST Visitor <${this.constructor.name}>:
	${i.join(`

`).replace(/\n/g, `
	`)}`);
      }
    }
  };
  return n.prototype = r, n.prototype.constructor = n, n._RULE_NAMES = e, n;
}
function _d(t, e, n) {
  const r = function() {
  };
  Oi(r, t + "BaseSemanticsWithDefaults");
  const s = Object.create(n.prototype);
  return _(e, (i) => {
    s[i] = Rd;
  }), r.prototype = s, r.prototype.constructor = r, r;
}
var Un;
(function(t) {
  t[t.REDUNDANT_METHOD = 0] = "REDUNDANT_METHOD", t[t.MISSING_METHOD = 1] = "MISSING_METHOD";
})(Un || (Un = {}));
function Sd(t, e) {
  return Od(t, e);
}
function Od(t, e) {
  const n = re(e, (s) => Ce(t[s]) === !1), r = A(n, (s) => ({
    msg: `Missing visitor method: <${s}> on ${t.constructor.name} CST Visitor.`,
    type: Un.MISSING_METHOD,
    methodName: s
  }));
  return At(r);
}
class Nd {
  initTreeBuilder(e) {
    if (this.CST_STACK = [], this.outputCst = e.outputCst, this.nodeLocationTracking = R(e, "nodeLocationTracking") ? e.nodeLocationTracking : Ne.nodeLocationTracking, !this.outputCst)
      this.cstInvocationStateUpdate = U, this.cstFinallyStateUpdate = U, this.cstPostTerminal = U, this.cstPostNonTerminal = U, this.cstPostRule = U;
    else if (/full/i.test(this.nodeLocationTracking))
      this.recoveryEnabled ? (this.setNodeLocationFromToken = gs, this.setNodeLocationFromNode = gs, this.cstPostRule = U, this.setInitialNodeLocation = this.setInitialNodeLocationFullRecovery) : (this.setNodeLocationFromToken = U, this.setNodeLocationFromNode = U, this.cstPostRule = this.cstPostRuleFull, this.setInitialNodeLocation = this.setInitialNodeLocationFullRegular);
    else if (/onlyOffset/i.test(this.nodeLocationTracking))
      this.recoveryEnabled ? (this.setNodeLocationFromToken = ps, this.setNodeLocationFromNode = ps, this.cstPostRule = U, this.setInitialNodeLocation = this.setInitialNodeLocationOnlyOffsetRecovery) : (this.setNodeLocationFromToken = U, this.setNodeLocationFromNode = U, this.cstPostRule = this.cstPostRuleOnlyOffset, this.setInitialNodeLocation = this.setInitialNodeLocationOnlyOffsetRegular);
    else if (/none/i.test(this.nodeLocationTracking))
      this.setNodeLocationFromToken = U, this.setNodeLocationFromNode = U, this.cstPostRule = U, this.setInitialNodeLocation = U;
    else
      throw Error(`Invalid <nodeLocationTracking> config option: "${e.nodeLocationTracking}"`);
  }
  setInitialNodeLocationOnlyOffsetRecovery(e) {
    e.location = {
      startOffset: NaN,
      endOffset: NaN
    };
  }
  setInitialNodeLocationOnlyOffsetRegular(e) {
    e.location = {
      // without error recovery the starting Location of a new CstNode is guaranteed
      // To be the next Token's startOffset (for valid inputs).
      // For invalid inputs there won't be any CSTOutput so this potential
      // inaccuracy does not matter
      startOffset: this.LA(1).startOffset,
      endOffset: NaN
    };
  }
  setInitialNodeLocationFullRecovery(e) {
    e.location = {
      startOffset: NaN,
      startLine: NaN,
      startColumn: NaN,
      endOffset: NaN,
      endLine: NaN,
      endColumn: NaN
    };
  }
  /**
       *  @see setInitialNodeLocationOnlyOffsetRegular for explanation why this work
  
       * @param cstNode
       */
  setInitialNodeLocationFullRegular(e) {
    const n = this.LA(1);
    e.location = {
      startOffset: n.startOffset,
      startLine: n.startLine,
      startColumn: n.startColumn,
      endOffset: NaN,
      endLine: NaN,
      endColumn: NaN
    };
  }
  cstInvocationStateUpdate(e) {
    const n = {
      name: e,
      children: /* @__PURE__ */ Object.create(null)
    };
    this.setInitialNodeLocation(n), this.CST_STACK.push(n);
  }
  cstFinallyStateUpdate() {
    this.CST_STACK.pop();
  }
  cstPostRuleFull(e) {
    const n = this.LA(0), r = e.location;
    r.startOffset <= n.startOffset ? (r.endOffset = n.endOffset, r.endLine = n.endLine, r.endColumn = n.endColumn) : (r.startOffset = NaN, r.startLine = NaN, r.startColumn = NaN);
  }
  cstPostRuleOnlyOffset(e) {
    const n = this.LA(0), r = e.location;
    r.startOffset <= n.startOffset ? r.endOffset = n.endOffset : r.startOffset = NaN;
  }
  cstPostTerminal(e, n) {
    const r = this.CST_STACK[this.CST_STACK.length - 1];
    Td(r, n, e), this.setNodeLocationFromToken(r.location, n);
  }
  cstPostNonTerminal(e, n) {
    const r = this.CST_STACK[this.CST_STACK.length - 1];
    md(r, n, e), this.setNodeLocationFromNode(r.location, e.location);
  }
  getBaseCstVisitorConstructor() {
    if (Oe(this.baseCstVisitorConstructor)) {
      const e = Id(this.className, ne(this.gastProductionsCache));
      return this.baseCstVisitorConstructor = e, e;
    }
    return this.baseCstVisitorConstructor;
  }
  getBaseCstVisitorConstructorWithDefaults() {
    if (Oe(this.baseCstVisitorWithDefaultsConstructor)) {
      const e = _d(this.className, ne(this.gastProductionsCache), this.getBaseCstVisitorConstructor());
      return this.baseCstVisitorWithDefaultsConstructor = e, e;
    }
    return this.baseCstVisitorWithDefaultsConstructor;
  }
  getLastExplicitRuleShortName() {
    const e = this.RULE_STACK;
    return e[e.length - 1];
  }
  getPreviousExplicitRuleShortName() {
    const e = this.RULE_STACK;
    return e[e.length - 2];
  }
  getLastExplicitRuleOccurrenceIndex() {
    const e = this.RULE_OCCURRENCE_STACK;
    return e[e.length - 1];
  }
}
class Cd {
  initLexerAdapter() {
    this.tokVector = [], this.tokVectorLength = 0, this.currIdx = -1;
  }
  set input(e) {
    if (this.selfAnalysisDone !== !0)
      throw Error("Missing <performSelfAnalysis> invocation at the end of the Parser's constructor.");
    this.reset(), this.tokVector = e, this.tokVectorLength = e.length;
  }
  get input() {
    return this.tokVector;
  }
  // skips a token and returns the next token
  SKIP_TOKEN() {
    return this.currIdx <= this.tokVector.length - 2 ? (this.consumeToken(), this.LA(1)) : Vt;
  }
  // Lexer (accessing Token vector) related methods which can be overridden to implement lazy lexers
  // or lexers dependent on parser context.
  LA(e) {
    const n = this.currIdx + e;
    return n < 0 || this.tokVectorLength <= n ? Vt : this.tokVector[n];
  }
  consumeToken() {
    this.currIdx++;
  }
  exportLexerState() {
    return this.currIdx;
  }
  importLexerState(e) {
    this.currIdx = e;
  }
  resetLexerState() {
    this.currIdx = -1;
  }
  moveToTerminatedState() {
    this.currIdx = this.tokVector.length - 1;
  }
  getLexerPosition() {
    return this.exportLexerState();
  }
}
class Ld {
  ACTION(e) {
    return e.call(this);
  }
  consume(e, n, r) {
    return this.consumeInternal(n, e, r);
  }
  subrule(e, n, r) {
    return this.subruleInternal(n, e, r);
  }
  option(e, n) {
    return this.optionInternal(n, e);
  }
  or(e, n) {
    return this.orInternal(n, e);
  }
  many(e, n) {
    return this.manyInternal(e, n);
  }
  atLeastOne(e, n) {
    return this.atLeastOneInternal(e, n);
  }
  CONSUME(e, n) {
    return this.consumeInternal(e, 0, n);
  }
  CONSUME1(e, n) {
    return this.consumeInternal(e, 1, n);
  }
  CONSUME2(e, n) {
    return this.consumeInternal(e, 2, n);
  }
  CONSUME3(e, n) {
    return this.consumeInternal(e, 3, n);
  }
  CONSUME4(e, n) {
    return this.consumeInternal(e, 4, n);
  }
  CONSUME5(e, n) {
    return this.consumeInternal(e, 5, n);
  }
  CONSUME6(e, n) {
    return this.consumeInternal(e, 6, n);
  }
  CONSUME7(e, n) {
    return this.consumeInternal(e, 7, n);
  }
  CONSUME8(e, n) {
    return this.consumeInternal(e, 8, n);
  }
  CONSUME9(e, n) {
    return this.consumeInternal(e, 9, n);
  }
  SUBRULE(e, n) {
    return this.subruleInternal(e, 0, n);
  }
  SUBRULE1(e, n) {
    return this.subruleInternal(e, 1, n);
  }
  SUBRULE2(e, n) {
    return this.subruleInternal(e, 2, n);
  }
  SUBRULE3(e, n) {
    return this.subruleInternal(e, 3, n);
  }
  SUBRULE4(e, n) {
    return this.subruleInternal(e, 4, n);
  }
  SUBRULE5(e, n) {
    return this.subruleInternal(e, 5, n);
  }
  SUBRULE6(e, n) {
    return this.subruleInternal(e, 6, n);
  }
  SUBRULE7(e, n) {
    return this.subruleInternal(e, 7, n);
  }
  SUBRULE8(e, n) {
    return this.subruleInternal(e, 8, n);
  }
  SUBRULE9(e, n) {
    return this.subruleInternal(e, 9, n);
  }
  OPTION(e) {
    return this.optionInternal(e, 0);
  }
  OPTION1(e) {
    return this.optionInternal(e, 1);
  }
  OPTION2(e) {
    return this.optionInternal(e, 2);
  }
  OPTION3(e) {
    return this.optionInternal(e, 3);
  }
  OPTION4(e) {
    return this.optionInternal(e, 4);
  }
  OPTION5(e) {
    return this.optionInternal(e, 5);
  }
  OPTION6(e) {
    return this.optionInternal(e, 6);
  }
  OPTION7(e) {
    return this.optionInternal(e, 7);
  }
  OPTION8(e) {
    return this.optionInternal(e, 8);
  }
  OPTION9(e) {
    return this.optionInternal(e, 9);
  }
  OR(e) {
    return this.orInternal(e, 0);
  }
  OR1(e) {
    return this.orInternal(e, 1);
  }
  OR2(e) {
    return this.orInternal(e, 2);
  }
  OR3(e) {
    return this.orInternal(e, 3);
  }
  OR4(e) {
    return this.orInternal(e, 4);
  }
  OR5(e) {
    return this.orInternal(e, 5);
  }
  OR6(e) {
    return this.orInternal(e, 6);
  }
  OR7(e) {
    return this.orInternal(e, 7);
  }
  OR8(e) {
    return this.orInternal(e, 8);
  }
  OR9(e) {
    return this.orInternal(e, 9);
  }
  MANY(e) {
    this.manyInternal(0, e);
  }
  MANY1(e) {
    this.manyInternal(1, e);
  }
  MANY2(e) {
    this.manyInternal(2, e);
  }
  MANY3(e) {
    this.manyInternal(3, e);
  }
  MANY4(e) {
    this.manyInternal(4, e);
  }
  MANY5(e) {
    this.manyInternal(5, e);
  }
  MANY6(e) {
    this.manyInternal(6, e);
  }
  MANY7(e) {
    this.manyInternal(7, e);
  }
  MANY8(e) {
    this.manyInternal(8, e);
  }
  MANY9(e) {
    this.manyInternal(9, e);
  }
  MANY_SEP(e) {
    this.manySepFirstInternal(0, e);
  }
  MANY_SEP1(e) {
    this.manySepFirstInternal(1, e);
  }
  MANY_SEP2(e) {
    this.manySepFirstInternal(2, e);
  }
  MANY_SEP3(e) {
    this.manySepFirstInternal(3, e);
  }
  MANY_SEP4(e) {
    this.manySepFirstInternal(4, e);
  }
  MANY_SEP5(e) {
    this.manySepFirstInternal(5, e);
  }
  MANY_SEP6(e) {
    this.manySepFirstInternal(6, e);
  }
  MANY_SEP7(e) {
    this.manySepFirstInternal(7, e);
  }
  MANY_SEP8(e) {
    this.manySepFirstInternal(8, e);
  }
  MANY_SEP9(e) {
    this.manySepFirstInternal(9, e);
  }
  AT_LEAST_ONE(e) {
    this.atLeastOneInternal(0, e);
  }
  AT_LEAST_ONE1(e) {
    return this.atLeastOneInternal(1, e);
  }
  AT_LEAST_ONE2(e) {
    this.atLeastOneInternal(2, e);
  }
  AT_LEAST_ONE3(e) {
    this.atLeastOneInternal(3, e);
  }
  AT_LEAST_ONE4(e) {
    this.atLeastOneInternal(4, e);
  }
  AT_LEAST_ONE5(e) {
    this.atLeastOneInternal(5, e);
  }
  AT_LEAST_ONE6(e) {
    this.atLeastOneInternal(6, e);
  }
  AT_LEAST_ONE7(e) {
    this.atLeastOneInternal(7, e);
  }
  AT_LEAST_ONE8(e) {
    this.atLeastOneInternal(8, e);
  }
  AT_LEAST_ONE9(e) {
    this.atLeastOneInternal(9, e);
  }
  AT_LEAST_ONE_SEP(e) {
    this.atLeastOneSepFirstInternal(0, e);
  }
  AT_LEAST_ONE_SEP1(e) {
    this.atLeastOneSepFirstInternal(1, e);
  }
  AT_LEAST_ONE_SEP2(e) {
    this.atLeastOneSepFirstInternal(2, e);
  }
  AT_LEAST_ONE_SEP3(e) {
    this.atLeastOneSepFirstInternal(3, e);
  }
  AT_LEAST_ONE_SEP4(e) {
    this.atLeastOneSepFirstInternal(4, e);
  }
  AT_LEAST_ONE_SEP5(e) {
    this.atLeastOneSepFirstInternal(5, e);
  }
  AT_LEAST_ONE_SEP6(e) {
    this.atLeastOneSepFirstInternal(6, e);
  }
  AT_LEAST_ONE_SEP7(e) {
    this.atLeastOneSepFirstInternal(7, e);
  }
  AT_LEAST_ONE_SEP8(e) {
    this.atLeastOneSepFirstInternal(8, e);
  }
  AT_LEAST_ONE_SEP9(e) {
    this.atLeastOneSepFirstInternal(9, e);
  }
  RULE(e, n, r = Yt) {
    if (H(this.definedRulesNames, e)) {
      const a = {
        message: De.buildDuplicateRuleNameError({
          topLevelRule: e,
          grammarName: this.className
        }),
        type: B.DUPLICATE_RULE_NAME,
        ruleName: e
      };
      this.definitionErrors.push(a);
    }
    this.definedRulesNames.push(e);
    const s = this.defineRule(e, n, r);
    return this[e] = s, s;
  }
  OVERRIDE_RULE(e, n, r = Yt) {
    const s = Xf(e, this.definedRulesNames, this.className);
    this.definitionErrors = this.definitionErrors.concat(s);
    const i = this.defineRule(e, n, r);
    return this[e] = i, i;
  }
  BACKTRACK(e, n) {
    return function() {
      this.isBackTrackingStack.push(1);
      const r = this.saveRecogState();
      try {
        return e.apply(this, n), !0;
      } catch (s) {
        if (jt(s))
          return !1;
        throw s;
      } finally {
        this.reloadRecogState(r), this.isBackTrackingStack.pop();
      }
    };
  }
  // GAST export APIs
  getGAstProductions() {
    return this.gastProductionsCache;
  }
  getSerializedGastProductions() {
    return Lh(b(this.gastProductionsCache));
  }
}
class yd {
  initRecognizerEngine(e, n) {
    if (this.className = this.constructor.name, this.shortRuleNameToFull = {}, this.fullRuleNameToShort = {}, this.ruleShortNameIdx = 256, this.tokenMatcher = Wt, this.subruleIdx = 0, this.definedRulesNames = [], this.tokensMap = {}, this.isBackTrackingStack = [], this.RULE_STACK = [], this.RULE_OCCURRENCE_STACK = [], this.gastProductionsCache = {}, R(n, "serializedGrammar"))
      throw Error(`The Parser's configuration can no longer contain a <serializedGrammar> property.
	See: https://chevrotain.io/docs/changes/BREAKING_CHANGES.html#_6-0-0
	For Further details.`);
    if (N(e)) {
      if (v(e))
        throw Error(`A Token Vocabulary cannot be empty.
	Note that the first argument for the parser constructor
	is no longer a Token vector (since v4.0).`);
      if (typeof e[0].startOffset == "number")
        throw Error(`The Parser constructor no longer accepts a token vector as the first argument.
	See: https://chevrotain.io/docs/changes/BREAKING_CHANGES.html#_4-0-0
	For Further details.`);
    }
    if (N(e))
      this.tokensMap = Y(e, (i, a) => (i[a.name] = a, i), {});
    else if (R(e, "modes") && oe(ae(b(e.modes)), Sf)) {
      const i = ae(b(e.modes)), a = ar(i);
      this.tokensMap = Y(a, (o, c) => (o[c.name] = c, o), {});
    } else if (te(e))
      this.tokensMap = D(e);
    else
      throw new Error("<tokensDictionary> argument must be An Array of Token constructors, A dictionary of Token constructors or an IMultiModeLexerDefinition");
    this.tokensMap.EOF = He;
    const r = R(e, "modes") ? ae(b(e.modes)) : b(e), s = oe(r, (i) => v(i.categoryMatches));
    this.tokenMatcher = s ? Wt : It, _t(b(this.tokensMap));
  }
  defineRule(e, n, r) {
    if (this.selfAnalysisDone)
      throw Error(`Grammar rule <${e}> may not be defined after the 'performSelfAnalysis' method has been called'
Make sure that all grammar rule definitions are done before 'performSelfAnalysis' is called.`);
    const s = R(r, "resyncEnabled") ? r.resyncEnabled : Yt.resyncEnabled, i = R(r, "recoveryValueFunc") ? r.recoveryValueFunc : Yt.recoveryValueFunc, a = this.ruleShortNameIdx << fd + we;
    this.ruleShortNameIdx++, this.shortRuleNameToFull[a] = e, this.fullRuleNameToShort[e] = a;
    let o;
    return this.outputCst === !0 ? o = function(...l) {
      try {
        this.ruleInvocationStateUpdate(a, e, this.subruleIdx), n.apply(this, l);
        const h = this.CST_STACK[this.CST_STACK.length - 1];
        return this.cstPostRule(h), h;
      } catch (h) {
        return this.invokeRuleCatch(h, s, i);
      } finally {
        this.ruleFinallyStateUpdate();
      }
    } : o = function(...l) {
      try {
        return this.ruleInvocationStateUpdate(a, e, this.subruleIdx), n.apply(this, l);
      } catch (h) {
        return this.invokeRuleCatch(h, s, i);
      } finally {
        this.ruleFinallyStateUpdate();
      }
    }, Object.assign(o, { ruleName: e, originalGrammarAction: n });
  }
  invokeRuleCatch(e, n, r) {
    const s = this.RULE_STACK.length === 1, i = n && !this.isBackTracking() && this.recoveryEnabled;
    if (jt(e)) {
      const a = e;
      if (i) {
        const o = this.findReSyncTokenType();
        if (this.isInCurrentRuleReSyncSet(o))
          if (a.resyncedTokens = this.reSyncTo(o), this.outputCst) {
            const c = this.CST_STACK[this.CST_STACK.length - 1];
            return c.recoveredNode = !0, c;
          } else
            return r(e);
        else {
          if (this.outputCst) {
            const c = this.CST_STACK[this.CST_STACK.length - 1];
            c.recoveredNode = !0, a.partialCstResult = c;
          }
          throw a;
        }
      } else {
        if (s)
          return this.moveToTerminatedState(), r(e);
        throw a;
      }
    } else
      throw e;
  }
  // Implementation of parsing DSL
  optionInternal(e, n) {
    const r = this.getKeyForAutomaticLookahead(Si, n);
    return this.optionInternalLogic(e, n, r);
  }
  optionInternalLogic(e, n, r) {
    let s = this.getLaFuncFromCache(r), i;
    if (typeof e != "function") {
      i = e.DEF;
      const a = e.GATE;
      if (a !== void 0) {
        const o = s;
        s = () => a.call(this) && o.call(this);
      }
    } else
      i = e;
    if (s.call(this) === !0)
      return i.call(this);
  }
  atLeastOneInternal(e, n) {
    const r = this.getKeyForAutomaticLookahead(bn, e);
    return this.atLeastOneInternalLogic(e, n, r);
  }
  atLeastOneInternalLogic(e, n, r) {
    let s = this.getLaFuncFromCache(r), i;
    if (typeof n != "function") {
      i = n.DEF;
      const a = n.GATE;
      if (a !== void 0) {
        const o = s;
        s = () => a.call(this) && o.call(this);
      }
    } else
      i = n;
    if (s.call(this) === !0) {
      let a = this.doSingleRepetition(i);
      for (; s.call(this) === !0 && a === !0; )
        a = this.doSingleRepetition(i);
    } else
      throw this.raiseEarlyExitException(e, k.REPETITION_MANDATORY, n.ERR_MSG);
    this.attemptInRepetitionRecovery(this.atLeastOneInternal, [e, n], s, bn, e, bf);
  }
  atLeastOneSepFirstInternal(e, n) {
    const r = this.getKeyForAutomaticLookahead(Ut, e);
    this.atLeastOneSepFirstInternalLogic(e, n, r);
  }
  atLeastOneSepFirstInternalLogic(e, n, r) {
    const s = n.DEF, i = n.SEP;
    if (this.getLaFuncFromCache(r).call(this) === !0) {
      s.call(this);
      const o = () => this.tokenMatcher(this.LA(1), i);
      for (; this.tokenMatcher(this.LA(1), i) === !0; )
        this.CONSUME(i), s.call(this);
      this.attemptInRepetitionRecovery(this.repetitionSepSecondInternal, [
        e,
        i,
        o,
        s,
        fs
      ], o, Ut, e, fs);
    } else
      throw this.raiseEarlyExitException(e, k.REPETITION_MANDATORY_WITH_SEPARATOR, n.ERR_MSG);
  }
  manyInternal(e, n) {
    const r = this.getKeyForAutomaticLookahead(xn, e);
    return this.manyInternalLogic(e, n, r);
  }
  manyInternalLogic(e, n, r) {
    let s = this.getLaFuncFromCache(r), i;
    if (typeof n != "function") {
      i = n.DEF;
      const o = n.GATE;
      if (o !== void 0) {
        const c = s;
        s = () => o.call(this) && c.call(this);
      }
    } else
      i = n;
    let a = !0;
    for (; s.call(this) === !0 && a === !0; )
      a = this.doSingleRepetition(i);
    this.attemptInRepetitionRecovery(
      this.manyInternal,
      [e, n],
      s,
      xn,
      e,
      xf,
      // The notStuck parameter is only relevant when "attemptInRepetitionRecovery"
      // is invoked from manyInternal, in the MANY_SEP case and AT_LEAST_ONE[_SEP]
      // An infinite loop cannot occur as:
      // - Either the lookahead is guaranteed to consume something (Single Token Separator)
      // - AT_LEAST_ONE by definition is guaranteed to consume something (or error out).
      a
    );
  }
  manySepFirstInternal(e, n) {
    const r = this.getKeyForAutomaticLookahead(wn, e);
    this.manySepFirstInternalLogic(e, n, r);
  }
  manySepFirstInternalLogic(e, n, r) {
    const s = n.DEF, i = n.SEP;
    if (this.getLaFuncFromCache(r).call(this) === !0) {
      s.call(this);
      const o = () => this.tokenMatcher(this.LA(1), i);
      for (; this.tokenMatcher(this.LA(1), i) === !0; )
        this.CONSUME(i), s.call(this);
      this.attemptInRepetitionRecovery(this.repetitionSepSecondInternal, [
        e,
        i,
        o,
        s,
        hs
      ], o, wn, e, hs);
    }
  }
  repetitionSepSecondInternal(e, n, r, s, i) {
    for (; r(); )
      this.CONSUME(n), s.call(this);
    this.attemptInRepetitionRecovery(this.repetitionSepSecondInternal, [
      e,
      n,
      r,
      s,
      i
    ], r, Ut, e, i);
  }
  doSingleRepetition(e) {
    const n = this.getLexerPosition();
    return e.call(this), this.getLexerPosition() > n;
  }
  orInternal(e, n) {
    const r = this.getKeyForAutomaticLookahead(_i, n), s = N(e) ? e : e.DEF, a = this.getLaFuncFromCache(r).call(this, s);
    if (a !== void 0)
      return s[a].ALT.call(this);
    this.raiseNoAltException(n, e.ERR_MSG);
  }
  ruleFinallyStateUpdate() {
    if (this.RULE_STACK.pop(), this.RULE_OCCURRENCE_STACK.pop(), this.cstFinallyStateUpdate(), this.RULE_STACK.length === 0 && this.isAtEndOfInput() === !1) {
      const e = this.LA(1), n = this.errorMessageProvider.buildNotAllInputParsedMessage({
        firstRedundant: e,
        ruleName: this.getCurrRuleFullName()
      });
      this.SAVE_ERROR(new od(n, e));
    }
  }
  subruleInternal(e, n, r) {
    let s;
    try {
      const i = r !== void 0 ? r.ARGS : void 0;
      return this.subruleIdx = n, s = e.apply(this, i), this.cstPostNonTerminal(s, r !== void 0 && r.LABEL !== void 0 ? r.LABEL : e.ruleName), s;
    } catch (i) {
      throw this.subruleInternalError(i, r, e.ruleName);
    }
  }
  subruleInternalError(e, n, r) {
    throw jt(e) && e.partialCstResult !== void 0 && (this.cstPostNonTerminal(e.partialCstResult, n !== void 0 && n.LABEL !== void 0 ? n.LABEL : r), delete e.partialCstResult), e;
  }
  consumeInternal(e, n, r) {
    let s;
    try {
      const i = this.LA(1);
      this.tokenMatcher(i, e) === !0 ? (this.consumeToken(), s = i) : this.consumeInternalError(e, i, r);
    } catch (i) {
      s = this.consumeInternalRecovery(e, n, i);
    }
    return this.cstPostTerminal(r !== void 0 && r.LABEL !== void 0 ? r.LABEL : e.name, s), s;
  }
  consumeInternalError(e, n, r) {
    let s;
    const i = this.LA(0);
    throw r !== void 0 && r.ERR_MSG ? s = r.ERR_MSG : s = this.errorMessageProvider.buildMismatchTokenMessage({
      expected: e,
      actual: n,
      previous: i,
      ruleName: this.getCurrRuleFullName()
    }), this.SAVE_ERROR(new Ri(s, n, i));
  }
  consumeInternalRecovery(e, n, r) {
    if (this.recoveryEnabled && // TODO: more robust checking of the exception type. Perhaps Typescript extending expressions?
    r.name === "MismatchedTokenException" && !this.isBackTracking()) {
      const s = this.getFollowsForInRuleRecovery(e, n);
      try {
        return this.tryInRuleRecovery(e, s);
      } catch (i) {
        throw i.name === Ii ? r : i;
      }
    } else
      throw r;
  }
  saveRecogState() {
    const e = this.errors, n = D(this.RULE_STACK);
    return {
      errors: e,
      lexerState: this.exportLexerState(),
      RULE_STACK: n,
      CST_STACK: this.CST_STACK
    };
  }
  reloadRecogState(e) {
    this.errors = e.errors, this.importLexerState(e.lexerState), this.RULE_STACK = e.RULE_STACK;
  }
  ruleInvocationStateUpdate(e, n, r) {
    this.RULE_OCCURRENCE_STACK.push(r), this.RULE_STACK.push(e), this.cstInvocationStateUpdate(n);
  }
  isBackTracking() {
    return this.isBackTrackingStack.length !== 0;
  }
  getCurrRuleFullName() {
    const e = this.getLastExplicitRuleShortName();
    return this.shortRuleNameToFull[e];
  }
  shortRuleNameToFullName(e) {
    return this.shortRuleNameToFull[e];
  }
  isAtEndOfInput() {
    return this.tokenMatcher(this.LA(1), He);
  }
  reset() {
    this.resetLexerState(), this.subruleIdx = 0, this.isBackTrackingStack = [], this.errors = [], this.RULE_STACK = [], this.CST_STACK = [], this.RULE_OCCURRENCE_STACK = [];
  }
}
class vd {
  initErrorHandler(e) {
    this._errors = [], this.errorMessageProvider = R(e, "errorMessageProvider") ? e.errorMessageProvider : Ne.errorMessageProvider;
  }
  SAVE_ERROR(e) {
    if (jt(e))
      return e.context = {
        ruleStack: this.getHumanReadableRuleStack(),
        ruleOccurrenceStack: D(this.RULE_OCCURRENCE_STACK)
      }, this._errors.push(e), e;
    throw Error("Trying to save an Error which is not a RecognitionException");
  }
  get errors() {
    return D(this._errors);
  }
  set errors(e) {
    this._errors = e;
  }
  // TODO: consider caching the error message computed information
  raiseEarlyExitException(e, n, r) {
    const s = this.getCurrRuleFullName(), i = this.getGAstProductions()[s], o = hr(e, i, n, this.maxLookahead)[0], c = [];
    for (let l = 1; l <= this.maxLookahead; l++)
      c.push(this.LA(l));
    const u = this.errorMessageProvider.buildEarlyExitMessage({
      expectedIterationPaths: o,
      actual: c,
      previous: this.LA(0),
      customUserDescription: r,
      ruleName: s
    });
    throw this.SAVE_ERROR(new cd(u, this.LA(1), this.LA(0)));
  }
  // TODO: consider caching the error message computed information
  raiseNoAltException(e, n) {
    const r = this.getCurrRuleFullName(), s = this.getGAstProductions()[r], i = lr(e, s, this.maxLookahead), a = [];
    for (let u = 1; u <= this.maxLookahead; u++)
      a.push(this.LA(u));
    const o = this.LA(0), c = this.errorMessageProvider.buildNoViableAltMessage({
      expectedPathsPerAlt: i,
      actual: a,
      previous: o,
      customUserDescription: n,
      ruleName: this.getCurrRuleFullName()
    });
    throw this.SAVE_ERROR(new ad(c, this.LA(1), o));
  }
}
class Pd {
  initContentAssist() {
  }
  computeContentAssist(e, n) {
    const r = this.gastProductionsCache[e];
    if (Oe(r))
      throw Error(`Rule ->${e}<- does not exist in this grammar.`);
    return ci([r], n, this.tokenMatcher, this.maxLookahead);
  }
  // TODO: should this be a member method or a utility? it does not have any state or usage of 'this'...
  // TODO: should this be more explicitly part of the public API?
  getNextPossibleTokenTypes(e) {
    const n = ue(e.ruleStack), s = this.getGAstProductions()[n];
    return new kf(s, e).startWalking();
  }
}
const fn = {
  description: "This Object indicates the Parser is during Recording Phase"
};
Object.freeze(fn);
const Es = !0, Ts = Math.pow(2, we) - 1, Ni = C({ name: "RECORDING_PHASE_TOKEN", pattern: G.NA });
_t([Ni]);
const Ci = ur(
  Ni,
  `This IToken indicates the Parser is in Recording Phase
	See: https://chevrotain.io/docs/guide/internals.html#grammar-recording for details`,
  // Using "-1" instead of NaN (as in EOF) because an actual number is less likely to
  // cause errors if the output of LA or CONSUME would be (incorrectly) used during the recording phase.
  -1,
  -1,
  -1,
  -1,
  -1,
  -1
);
Object.freeze(Ci);
const Md = {
  name: `This CSTNode indicates the Parser is in Recording Phase
	See: https://chevrotain.io/docs/guide/internals.html#grammar-recording for details`,
  children: {}
};
class kd {
  initGastRecorder(e) {
    this.recordingProdStack = [], this.RECORDING_PHASE = !1;
  }
  enableRecording() {
    this.RECORDING_PHASE = !0, this.TRACE_INIT("Enable Recording", () => {
      for (let e = 0; e < 10; e++) {
        const n = e > 0 ? e : "";
        this[`CONSUME${n}`] = function(r, s) {
          return this.consumeInternalRecord(r, e, s);
        }, this[`SUBRULE${n}`] = function(r, s) {
          return this.subruleInternalRecord(r, e, s);
        }, this[`OPTION${n}`] = function(r) {
          return this.optionInternalRecord(r, e);
        }, this[`OR${n}`] = function(r) {
          return this.orInternalRecord(r, e);
        }, this[`MANY${n}`] = function(r) {
          this.manyInternalRecord(e, r);
        }, this[`MANY_SEP${n}`] = function(r) {
          this.manySepFirstInternalRecord(e, r);
        }, this[`AT_LEAST_ONE${n}`] = function(r) {
          this.atLeastOneInternalRecord(e, r);
        }, this[`AT_LEAST_ONE_SEP${n}`] = function(r) {
          this.atLeastOneSepFirstInternalRecord(e, r);
        };
      }
      this.consume = function(e, n, r) {
        return this.consumeInternalRecord(n, e, r);
      }, this.subrule = function(e, n, r) {
        return this.subruleInternalRecord(n, e, r);
      }, this.option = function(e, n) {
        return this.optionInternalRecord(n, e);
      }, this.or = function(e, n) {
        return this.orInternalRecord(n, e);
      }, this.many = function(e, n) {
        this.manyInternalRecord(e, n);
      }, this.atLeastOne = function(e, n) {
        this.atLeastOneInternalRecord(e, n);
      }, this.ACTION = this.ACTION_RECORD, this.BACKTRACK = this.BACKTRACK_RECORD, this.LA = this.LA_RECORD;
    });
  }
  disableRecording() {
    this.RECORDING_PHASE = !1, this.TRACE_INIT("Deleting Recording methods", () => {
      const e = this;
      for (let n = 0; n < 10; n++) {
        const r = n > 0 ? n : "";
        delete e[`CONSUME${r}`], delete e[`SUBRULE${r}`], delete e[`OPTION${r}`], delete e[`OR${r}`], delete e[`MANY${r}`], delete e[`MANY_SEP${r}`], delete e[`AT_LEAST_ONE${r}`], delete e[`AT_LEAST_ONE_SEP${r}`];
      }
      delete e.consume, delete e.subrule, delete e.option, delete e.or, delete e.many, delete e.atLeastOne, delete e.ACTION, delete e.BACKTRACK, delete e.LA;
    });
  }
  //   Parser methods are called inside an ACTION?
  //   Maybe try/catch/finally on ACTIONS while disabling the recorders state changes?
  // @ts-expect-error -- noop place holder
  ACTION_RECORD(e) {
  }
  // Executing backtracking logic will break our recording logic assumptions
  BACKTRACK_RECORD(e, n) {
    return () => !0;
  }
  // LA is part of the official API and may be used for custom lookahead logic
  // by end users who may forget to wrap it in ACTION or inside a GATE
  LA_RECORD(e) {
    return Vt;
  }
  topLevelRuleRecord(e, n) {
    try {
      const r = new qe({ definition: [], name: e });
      return r.name = e, this.recordingProdStack.push(r), n.call(this), this.recordingProdStack.pop(), r;
    } catch (r) {
      if (r.KNOWN_RECORDER_ERROR !== !0)
        try {
          r.message = r.message + `
	 This error was thrown during the "grammar recording phase" For more info see:
	https://chevrotain.io/docs/guide/internals.html#grammar-recording`;
        } catch {
          throw r;
        }
      throw r;
    }
  }
  // Implementation of parsing DSL
  optionInternalRecord(e, n) {
    return it.call(this, W, e, n);
  }
  atLeastOneInternalRecord(e, n) {
    it.call(this, Re, n, e);
  }
  atLeastOneSepFirstInternalRecord(e, n) {
    it.call(this, Ie, n, e, Es);
  }
  manyInternalRecord(e, n) {
    it.call(this, w, n, e);
  }
  manySepFirstInternalRecord(e, n) {
    it.call(this, he, n, e, Es);
  }
  orInternalRecord(e, n) {
    return xd.call(this, e, n);
  }
  subruleInternalRecord(e, n, r) {
    if (zt(n), !e || R(e, "ruleName") === !1) {
      const o = new Error(`<SUBRULE${ms(n)}> argument is invalid expecting a Parser method reference but got: <${JSON.stringify(e)}>
 inside top level rule: <${this.recordingProdStack[0].name}>`);
      throw o.KNOWN_RECORDER_ERROR = !0, o;
    }
    const s = Ge(this.recordingProdStack), i = e.ruleName, a = new X({
      idx: n,
      nonTerminalName: i,
      label: r?.LABEL,
      // The resolving of the `referencedRule` property will be done once all the Rule's GASTs have been created
      referencedRule: void 0
    });
    return s.definition.push(a), this.outputCst ? Md : fn;
  }
  consumeInternalRecord(e, n, r) {
    if (zt(n), !ii(e)) {
      const a = new Error(`<CONSUME${ms(n)}> argument is invalid expecting a TokenType reference but got: <${JSON.stringify(e)}>
 inside top level rule: <${this.recordingProdStack[0].name}>`);
      throw a.KNOWN_RECORDER_ERROR = !0, a;
    }
    const s = Ge(this.recordingProdStack), i = new M({
      idx: n,
      terminalType: e,
      label: r?.LABEL
    });
    return s.definition.push(i), Ci;
  }
}
function it(t, e, n, r = !1) {
  zt(n);
  const s = Ge(this.recordingProdStack), i = Ce(e) ? e : e.DEF, a = new t({ definition: [], idx: n });
  return r && (a.separator = e.SEP), R(e, "MAX_LOOKAHEAD") && (a.maxLookahead = e.MAX_LOOKAHEAD), this.recordingProdStack.push(a), i.call(this), s.definition.push(a), this.recordingProdStack.pop(), fn;
}
function xd(t, e) {
  zt(e);
  const n = Ge(this.recordingProdStack), r = N(t) === !1, s = r === !1 ? t : t.DEF, i = new fe({
    definition: [],
    idx: e,
    ignoreAmbiguities: r && t.IGNORE_AMBIGUITIES === !0
  });
  R(t, "MAX_LOOKAHEAD") && (i.maxLookahead = t.MAX_LOOKAHEAD);
  const a = Ys(s, (o) => Ce(o.GATE));
  return i.hasPredicates = a, n.definition.push(i), _(s, (o) => {
    const c = new z({ definition: [] });
    i.definition.push(c), R(o, "IGNORE_AMBIGUITIES") ? c.ignoreAmbiguities = o.IGNORE_AMBIGUITIES : R(o, "GATE") && (c.ignoreAmbiguities = !0), this.recordingProdStack.push(c), o.ALT.call(this), this.recordingProdStack.pop();
  }), fn;
}
function ms(t) {
  return t === 0 ? "" : `${t}`;
}
function zt(t) {
  if (t < 0 || t > Ts) {
    const e = new Error(
      // The stack trace will contain all the needed details
      `Invalid DSL Method idx value: <${t}>
	Idx value must be a none negative value smaller than ${Ts + 1}`
    );
    throw e.KNOWN_RECORDER_ERROR = !0, e;
  }
}
class bd {
  initPerformanceTracer(e) {
    if (R(e, "traceInitPerf")) {
      const n = e.traceInitPerf, r = typeof n == "number";
      this.traceInitMaxIdent = r ? n : 1 / 0, this.traceInitPerf = r ? n > 0 : n;
    } else
      this.traceInitMaxIdent = 0, this.traceInitPerf = Ne.traceInitPerf;
    this.traceInitIndent = -1;
  }
  TRACE_INIT(e, n) {
    if (this.traceInitPerf === !0) {
      this.traceInitIndent++;
      const r = new Array(this.traceInitIndent + 1).join("	");
      this.traceInitIndent < this.traceInitMaxIdent && console.log(`${r}--> <${e}>`);
      const { time: s, value: i } = qs(n), a = s > 10 ? console.warn : console.log;
      return this.traceInitIndent < this.traceInitMaxIdent && a(`${r}<-- <${e}> time: ${s}ms`), this.traceInitIndent--, i;
    } else
      return n();
  }
}
function wd(t, e) {
  e.forEach((n) => {
    const r = n.prototype;
    Object.getOwnPropertyNames(r).forEach((s) => {
      if (s === "constructor")
        return;
      const i = Object.getOwnPropertyDescriptor(r, s);
      i && (i.get || i.set) ? Object.defineProperty(t.prototype, s, i) : t.prototype[s] = n.prototype[s];
    });
  });
}
const Vt = ur(He, "", NaN, NaN, NaN, NaN, NaN, NaN);
Object.freeze(Vt);
const Ne = Object.freeze({
  recoveryEnabled: !1,
  maxLookahead: 3,
  dynamicTokensEnabled: !1,
  outputCst: !0,
  errorMessageProvider: oi,
  nodeLocationTracking: "none",
  traceInitPerf: !1,
  skipValidations: !1
}), Yt = Object.freeze({
  recoveryValueFunc: () => {
  },
  resyncEnabled: !0
});
var B;
(function(t) {
  t[t.INVALID_RULE_NAME = 0] = "INVALID_RULE_NAME", t[t.DUPLICATE_RULE_NAME = 1] = "DUPLICATE_RULE_NAME", t[t.INVALID_RULE_OVERRIDE = 2] = "INVALID_RULE_OVERRIDE", t[t.DUPLICATE_PRODUCTIONS = 3] = "DUPLICATE_PRODUCTIONS", t[t.UNRESOLVED_SUBRULE_REF = 4] = "UNRESOLVED_SUBRULE_REF", t[t.LEFT_RECURSION = 5] = "LEFT_RECURSION", t[t.NONE_LAST_EMPTY_ALT = 6] = "NONE_LAST_EMPTY_ALT", t[t.AMBIGUOUS_ALTS = 7] = "AMBIGUOUS_ALTS", t[t.CONFLICT_TOKENS_RULES_NAMESPACE = 8] = "CONFLICT_TOKENS_RULES_NAMESPACE", t[t.INVALID_TOKEN_NAME = 9] = "INVALID_TOKEN_NAME", t[t.NO_NON_EMPTY_LOOKAHEAD = 10] = "NO_NON_EMPTY_LOOKAHEAD", t[t.AMBIGUOUS_PREFIX_ALTS = 11] = "AMBIGUOUS_PREFIX_ALTS", t[t.TOO_MANY_ALTS = 12] = "TOO_MANY_ALTS", t[t.CUSTOM_LOOKAHEAD_VALIDATION = 13] = "CUSTOM_LOOKAHEAD_VALIDATION";
})(B || (B = {}));
class St {
  /**
   *  @deprecated use the **instance** method with the same name instead
   */
  static performSelfAnalysis(e) {
    throw Error("The **static** `performSelfAnalysis` method has been deprecated.	\nUse the **instance** method with the same name instead.");
  }
  performSelfAnalysis() {
    this.TRACE_INIT("performSelfAnalysis", () => {
      let e;
      this.selfAnalysisDone = !0;
      const n = this.className;
      this.TRACE_INIT("toFastProps", () => {
        Zs(this);
      }), this.TRACE_INIT("Grammar Recording", () => {
        try {
          this.enableRecording(), _(this.definedRulesNames, (s) => {
            const a = this[s].originalGrammarAction;
            let o;
            this.TRACE_INIT(`${s} Rule`, () => {
              o = this.topLevelRuleRecord(s, a);
            }), this.gastProductionsCache[s] = o;
          });
        } finally {
          this.disableRecording();
        }
      });
      let r = [];
      if (this.TRACE_INIT("Grammar Resolving", () => {
        r = sd({
          rules: b(this.gastProductionsCache)
        }), this.definitionErrors = this.definitionErrors.concat(r);
      }), this.TRACE_INIT("Grammar Validations", () => {
        if (v(r) && this.skipValidations === !1) {
          const s = id({
            rules: b(this.gastProductionsCache),
            tokenTypes: b(this.tokensMap),
            errMsgProvider: De,
            grammarName: n
          }), i = Kf({
            lookaheadStrategy: this.lookaheadStrategy,
            rules: b(this.gastProductionsCache),
            tokenTypes: b(this.tokensMap),
            grammarName: n
          });
          this.definitionErrors = this.definitionErrors.concat(s, i);
        }
      }), v(this.definitionErrors) && (this.recoveryEnabled && this.TRACE_INIT("computeAllProdsFollows", () => {
        const s = bh(b(this.gastProductionsCache));
        this.resyncFollows = s;
      }), this.TRACE_INIT("ComputeLookaheadFunctions", () => {
        var s, i;
        (i = (s = this.lookaheadStrategy).initialize) === null || i === void 0 || i.call(s, {
          rules: b(this.gastProductionsCache)
        }), this.preComputeLookaheadFunctions(b(this.gastProductionsCache));
      })), !St.DEFER_DEFINITION_ERRORS_HANDLING && !v(this.definitionErrors))
        throw e = A(this.definitionErrors, (s) => s.message), new Error(`Parser Definition Errors detected:
 ${e.join(`
-------------------------------
`)}`);
    });
  }
  constructor(e, n) {
    this.definitionErrors = [], this.selfAnalysisDone = !1;
    const r = this;
    if (r.initErrorHandler(n), r.initLexerAdapter(), r.initLooksAhead(n), r.initRecognizerEngine(e, n), r.initRecoverable(n), r.initTreeBuilder(n), r.initContentAssist(), r.initGastRecorder(n), r.initPerformanceTracer(n), R(n, "ignoredIssues"))
      throw new Error(`The <ignoredIssues> IParserConfig property has been deprecated.
	Please use the <IGNORE_AMBIGUITIES> flag on the relevant DSL method instead.
	See: https://chevrotain.io/docs/guide/resolving_grammar_errors.html#IGNORING_AMBIGUITIES
	For further details.`);
    this.skipValidations = R(n, "skipValidations") ? n.skipValidations : Ne.skipValidations;
  }
}
St.DEFER_DEFINITION_ERRORS_HANDLING = !1;
wd(St, [
  ld,
  pd,
  Nd,
  Cd,
  yd,
  Ld,
  vd,
  Pd,
  kd,
  bd
]);
class Ud extends St {
  constructor(e, n = Ne) {
    const r = D(n);
    r.outputCst = !0, super(e, r);
  }
}
const Li = C({
  name: "HTML_COMMENT",
  pattern: /<!--[\s\S]*?-->/,
  line_breaks: !0
}), yi = C({
  name: "HTML_CONDITIONAL_COMMENT",
  pattern: /<!\[[\s\S]*?\]>/,
  line_breaks: !0
}), Fd = C({
  name: "XML",
  pattern: /<\?xml(?:.|\s)*?\?>/
}), vi = C({
  name: "CDATA",
  pattern: /<!\[CDATA\[[\s\S]*?]]>/
}), Pi = C({
  name: "DTD",
  pattern: /<!.*?>/
}), Mi = C({
  name: "SCRIPTLET",
  pattern: /<%(.*?)%>|<\?(.*?)\?>/
}), Dd = C({
  name: "SEA_WS",
  pattern: /[ \t]+/,
  group: G.SKIPPED
}), ki = C({
  name: "LINE_BREAK",
  pattern: /\r?\n/
}), xi = C({
  name: "SCRIPT_OPEN",
  pattern: /<script\b[^>]*>/,
  push_mode: "SCRIPT"
}), bi = C({
  name: "STYLE_OPEN",
  pattern: /<style\b[^>]*>/,
  push_mode: "STYLE"
}), Fn = C({
  name: "TAG_OPEN",
  pattern: /</,
  push_mode: "TAG"
}), wi = C({
  name: "HTML_TEXT",
  pattern: /[^<@]+/,
  line_breaks: !0
}), Dn = C({
  name: "TAG_CLOSE",
  pattern: />/,
  pop_mode: !0
}), Ui = C({
  name: "TAG_SLASH_CLOSE",
  pattern: /\/>/,
  pop_mode: !0
}), Fi = C({
  name: "TAG_SLASH",
  pattern: /\//
}), Di = C({
  name: "TAG_EQUALS",
  pattern: /=/,
  push_mode: "ATTVALUE"
}), Ft = C({
  name: "TAG_NAME",
  pattern: /[:@a-zA-Z\u2070-\u218F\u2C00-\u2FEF\u3001-\uD7FF\uF900-\uFDCF\uFDF0-\uFFFD][:\.\-@\u00B7\u0300-\u036F\u203F-\u2040.0-9a-zA-Z]*/
}), $d = C({
  name: "TAG_WHITESPACE",
  pattern: /[ \t\r\n]+/,
  group: G.SKIPPED
}), $i = C({
  name: "SCRIPT_BODY",
  pattern: /[\s\S]*?<\/script>/,
  pop_mode: !0
}), Gi = C({
  name: "STYLE_BODY",
  pattern: /[\s\S]*?<\/style>/,
  pop_mode: !0
}), Bi = C({
  name: "ATTVALUE_VALUE",
  pattern: / *("[^<"]*"|'[^<']*'|[-_./+,?=:#;0-9a-zA-Z]+ ?|#[0-9a-fA-F]+|[0-9]+%?)/,
  pop_mode: !0
}), $n = C({
  name: "EDGE_COMMENT",
  pattern: /{{--[\s\S]*?--}}/,
  line_breaks: !0
}), Gn = C({
  name: "EDGE_MUSTACHE",
  pattern: /{{[\s\S]*?}}\s*/
}), Bn = C({
  name: "EDGE_SAFE_MUSTACHE",
  pattern: /{{{[\s\S]*?}}}\s*/
}), Hi = C({
  name: "EDGE_ESCAPED_MUSTACHE",
  pattern: /@{{[\s\S]*?}}\s*/
}), Ki = C({
  name: "EDGE_TAG",
  pattern: /@(?:!?\w+(?:\.\w+)*)\s*(?:\((?:[^)(]+|\((?:[^)(]+|\([^)(]*\))*\))*\))?~?/
}), Wi = C({
  name: "EDGE_TAG_PROP",
  pattern: /@(if|elseif|else|each|flashMessage)\([^)]*\)\s*([\s\S]*?)@end/
}), Hn = {
  defaultMode: "default",
  modes: {
    default: [
      Li,
      yi,
      Fd,
      vi,
      Pi,
      Mi,
      ki,
      Dd,
      xi,
      bi,
      Fn,
      $n,
      Hi,
      Bn,
      Gn,
      Ki,
      wi
    ],
    TAG: [
      Dn,
      Ui,
      Fi,
      Di,
      $n,
      Bn,
      Gn,
      Wi,
      Ft,
      $d
    ],
    SCRIPT: [$i],
    STYLE: [Gi],
    ATTVALUE: [Bi]
  }
}, Gd = () => {
  const t = [];
  for (const e in Hn.modes)
    t.push(...Hn.modes[e]);
  return t;
}, Bd = new G(Hn);
class ji extends Ud {
  constructor() {
    super(Gd());
    const e = this;
    e.RULE("document", () => {
      e.MANY(() => {
        e.SUBRULE(e.content);
      });
    }), e.RULE("content", () => {
      e.OR([
        { ALT: () => e.CONSUME(wi) },
        { ALT: () => e.CONSUME(ki) },
        { ALT: () => e.SUBRULE(e.element) },
        { ALT: () => e.SUBRULE(e.scriptlet) },
        { ALT: () => e.SUBRULE(e.htmlComment) },
        { ALT: () => e.SUBRULE(e.htmlConditionalComment) },
        { ALT: () => e.SUBRULE(e.cdata) },
        { ALT: () => e.SUBRULE(e.dtd) },
        { ALT: () => e.SUBRULE(e.scriptElement) },
        { ALT: () => e.SUBRULE(e.styleElement) },
        { ALT: () => e.SUBRULE(e.edgeComment) },
        { ALT: () => e.SUBRULE(e.edgeSafeMustache) },
        { ALT: () => e.SUBRULE(e.edgeMustache) },
        { ALT: () => e.SUBRULE(e.edgeEscapedMustache) },
        { ALT: () => e.SUBRULE(e.edgeTag) }
      ]);
    }), e.RULE("element", () => {
      e.OR([
        {
          ALT: () => e.SUBRULE(e.openingTag)
        },
        {
          ALT: () => e.SUBRULE(e.closingTag)
        }
      ]);
    }), e.RULE("openingTag", () => {
      e.CONSUME(Fn), e.CONSUME(Ft), e.MANY(() => {
        e.OR([
          { ALT: () => e.SUBRULE(e.edgeTagProp) },
          { ALT: () => e.SUBRULE(e.attribute) },
          { ALT: () => e.SUBRULE(e.edgeSafeMustache) },
          { ALT: () => e.SUBRULE(e.edgeMustache) },
          { ALT: () => e.SUBRULE(e.edgeComment) }
        ]);
      }), e.OR1([
        { ALT: () => e.CONSUME1(Ui) },
        { ALT: () => e.CONSUME1(Dn) }
      ]);
    }), e.RULE("attribute", () => {
      e.CONSUME(Ft), e.OPTION(() => {
        e.CONSUME(Di), e.CONSUME(Bi);
      });
    }), e.RULE("closingTag", () => {
      e.CONSUME(Fn), e.CONSUME(Fi), e.CONSUME(Ft), e.CONSUME(Dn);
    }), e.RULE("scriptlet", () => {
      e.CONSUME(Mi);
    }), e.RULE("htmlComment", () => {
      e.CONSUME(Li);
    }), e.RULE("htmlConditionalComment", () => {
      e.CONSUME(yi);
    }), e.RULE("cdata", () => {
      e.CONSUME(vi);
    }), e.RULE("dtd", () => {
      e.CONSUME(Pi);
    }), e.RULE("scriptElement", () => {
      e.CONSUME(xi), e.CONSUME($i);
    }), e.RULE("styleElement", () => {
      e.CONSUME(bi), e.CONSUME(Gi);
    }), e.RULE("edgeComment", () => {
      e.CONSUME($n);
    }), e.RULE("edgeMustache", () => {
      e.CONSUME(Gn);
    }), e.RULE("edgeSafeMustache", () => {
      e.CONSUME(Bn);
    }), e.RULE("edgeEscapedMustache", () => {
      e.CONSUME(Hi);
    }), e.RULE("edgeTag", () => {
      e.CONSUME(Ki);
    }), e.RULE("edgeTagProp", () => {
      e.CONSUME(Wi);
    }), this.performSelfAnalysis();
  }
}
const Hd = new ji(), Kd = Hd.getBaseCstVisitorConstructor();
class Wd extends Kd {
  #t = /* @__PURE__ */ new Set([
    "area",
    "base",
    "br",
    "col",
    "embed",
    "hr",
    "img",
    "input",
    "link",
    "meta",
    "param",
    "source",
    "track",
    "wbr",
    // Below are for SVGs
    "animateMotion",
    "animateTransform",
    "animate",
    "circle",
    "ellipse",
    "feGaussianBlur",
    "feDropShadow",
    "feOffset",
    "feBlend",
    "feColorMatrix",
    "feComposite",
    "feDisplacementMap",
    "feFlood",
    "feImage",
    "feMergeNode",
    "feMorphology",
    "fePointLight",
    "feSpotLight",
    "feTile",
    "feTurbulence",
    "image",
    "line",
    "mpath",
    "path",
    "polygon",
    "polyline",
    "rect",
    "set",
    "stop",
    "use",
    "view"
  ]);
  constructor() {
    super(), this.validateVisitor();
  }
  #e(e) {
    return this.#t.has(e);
  }
  document(e) {
    const n = e.content[0].startOffset, r = e.content[e.content.length - 1].endOffset;
    return {
      type: "document",
      children: e.content.map((s) => this.visit(s)),
      start: n,
      end: r
    };
  }
  content(e) {
    if (e.HTML_TEXT)
      return {
        type: "htmlText",
        value: e.HTML_TEXT[0].image,
        start: e.HTML_TEXT[0].startOffset,
        end: e.HTML_TEXT[0].endOffset
      };
    if (e.LINE_BREAK)
      return {
        type: "linebreak",
        value: e.LINE_BREAK[0].image,
        start: e.LINE_BREAK[0].startOffset,
        end: e.LINE_BREAK[0].endOffset
      };
    const n = e.element || e.scriptlet || e.htmlComment || e.htmlConditionalComment || e.cdata || e.dtd || e.scriptElement || e.styleElement || e.edgeComment || e.edgeMustache || e.edgeSafeMustache || e.edgeEscapedMustache || e.edgeTag || e.htmlText;
    return this.visit(n);
  }
  element(e) {
    return e.openingTag ? this.visit(e.openingTag) : e.selfClosingTag ? this.visit(e.selfClosingTag) : this.visit(e.closingTag);
  }
  openingTag(e) {
    const n = e.TAG_NAME[0].image, r = e.edgeSafeMustache ? e.edgeSafeMustache.map((l) => this.visit(l)) : [], s = e.edgeMustache ? e.edgeMustache.map((l) => this.visit(l)) : [], i = e.edgeTagProp ? e.edgeTagProp.map((l) => this.visit(l)) : [], a = e.attribute ? e.attribute.map((l) => this.visit(l)) : [], o = e.edgeComment ? e.edgeComment.map((l) => this.visit(l)) : [], c = e.TAG_NAME[0].startOffset, u = e.TAG_NAME[0].endOffset;
    return this.#e(n) || e.TAG_SLASH_CLOSE ? {
      type: "voidTag",
      tagName: n,
      edgeSafeMustaches: r,
      edgeMustaches: s,
      edgeTagProps: i,
      attributes: a,
      comments: o,
      start: c,
      end: u
    } : {
      type: "openingTag",
      tagName: n,
      edgeSafeMustaches: r,
      edgeMustaches: s,
      edgeTagProps: i,
      attributes: a,
      comments: o,
      start: c,
      end: u
    };
  }
  attribute(e) {
    const n = e.TAG_NAME[0].image, r = e.ATTVALUE_VALUE?.[0]?.image?.trim(), s = e.TAG_NAME[0].startOffset, i = e.ATTVALUE_VALUE ? e.ATTVALUE_VALUE[0].endOffset : e.TAG_NAME[0].endOffset;
    return { type: "attribute", attributeName: n, attributeValue: r, start: s, end: i };
  }
  closingTag(e) {
    const n = e.TAG_NAME[0].image, r = e.TAG_OPEN[0].startOffset, s = e.TAG_CLOSE[0].endOffset;
    return this.#e(n) ? {
      type: "doNotPrint"
    } : { type: "closingTag", tagName: n, start: r, end: s };
  }
  scriptlet(e) {
    const n = e.SCRIPTLET[0].image, r = e.SCRIPTLET[0].startOffset, s = e.SCRIPTLET[0].endOffset;
    return { type: "scriptlet", value: n, start: r, end: s };
  }
  htmlComment(e) {
    const n = e.HTML_COMMENT[0].image, r = e.HTML_COMMENT[0].startOffset, s = e.HTML_COMMENT[0].endOffset;
    return { type: "htmlComment", value: n, start: r, end: s };
  }
  htmlConditionalComment(e) {
    const n = e.HTML_CONDITIONAL_COMMENT[0].image, r = e.HTML_CONDITIONAL_COMMENT[0].startOffset, s = e.HTML_CONDITIONAL_COMMENT[0].endOffset;
    return { type: "htmlConditionalComment", value: n, start: r, end: s };
  }
  cdata(e) {
    const n = e.CDATA[0].image, r = e.CDATA[0].startOffset, s = e.CDATA[0].endOffset;
    return { type: "cdata", value: n, start: r, end: s };
  }
  dtd(e) {
    const n = e.DTD[0].image, r = e.DTD[0].startOffset, s = e.DTD[0].endOffset;
    return { type: "dtd", value: n, start: r, end: s };
  }
  scriptElement(e) {
    const n = e.SCRIPT_OPEN[0].image + e.SCRIPT_BODY[0].image, r = e.SCRIPT_OPEN[0].startOffset, s = e.SCRIPT_BODY[0].endOffset;
    return { type: "scriptElement", value: n, start: r, end: s };
  }
  styleElement(e) {
    const n = e.STYLE_OPEN[0].image + e.STYLE_BODY[0].image, r = e.STYLE_OPEN[0].startOffset, s = e.STYLE_BODY[0].endOffset;
    return { type: "styleElement", value: n, start: r, end: s };
  }
  edgeComment(e) {
    const n = e.EDGE_COMMENT[0].image, r = e.EDGE_COMMENT[0].startOffset, s = e.EDGE_COMMENT[0].endOffset;
    return { type: "edgeComment", value: n, start: r, end: s };
  }
  edgeMustache(e) {
    const n = e.EDGE_MUSTACHE[0].image, r = e.EDGE_MUSTACHE[0].startOffset, s = e.EDGE_MUSTACHE[0].endOffset;
    return { type: "edgeMustache", value: n, start: r, end: s };
  }
  edgeSafeMustache(e) {
    const n = e.EDGE_SAFE_MUSTACHE[0].image, r = e.EDGE_SAFE_MUSTACHE[0].startOffset, s = e.EDGE_SAFE_MUSTACHE[0].endOffset;
    return { type: "edgeSafeMustache", value: n, start: r, end: s };
  }
  edgeEscapedMustache(e) {
    const n = e.EDGE_ESCAPED_MUSTACHE[0].image, r = e.EDGE_ESCAPED_MUSTACHE[0].startOffset, s = e.EDGE_ESCAPED_MUSTACHE[0].endOffset;
    return { type: "edgeEscapedMustache", value: n, start: r, end: s };
  }
  edgeTag(e) {
    const n = e.EDGE_TAG[0].image, r = e.EDGE_TAG[0].startOffset, s = e.EDGE_TAG[0].endOffset;
    return { type: "edgeTag", value: n, start: r, end: s };
  }
  edgeTagProp(e) {
    const n = e.EDGE_TAG_PROP[0].image, r = e.EDGE_TAG_PROP[0].startOffset, s = e.EDGE_TAG_PROP[0].endOffset;
    return { type: "edgeTagProp", value: n, start: r, end: s };
  }
}
const Pt = new ji(), jd = new Wd();
function zd(t) {
  const e = Bd.tokenize(t);
  Pt.input = e.tokens;
  const n = Pt.document();
  if (Pt.errors.length > 0)
    throw new Error(
      `Parsing Errors Detected: ${JSON.stringify(Pt.errors)}`
    );
  return jd.visit(n);
}
export {
  zd as default
};
